<!doctype html>
<html class="no-js" lang="en">
  <head><meta charset="utf-8"/>
    <meta name="viewport" content="width=device-width,initial-scale=1"/>
    <meta name="color-scheme" content="light dark"><meta name="generator" content="Docutils 0.18.1: http://docutils.sourceforge.net/" />
<link rel="index" title="Index" href="genindex.html" /><link rel="search" title="Search" href="search.html" /><link rel="next" title="Structural and Copy Number variants" href="balsamic_sv_cnv.html" /><link rel="prev" title="Command line arguments" href="cli_package.html" />

    <!-- Generated with Sphinx 6.2.1 and Furo 2023.07.26 -->
        <title>Calling and filtering variants - BALSAMIC documentation</title>
      <link rel="stylesheet" type="text/css" href="_static/pygments.css" />
    <link rel="stylesheet" type="text/css" href="_static/styles/furo.css?digest=369552022d0b975c8e74270ce6eabe0fb7978f24" />
    <link rel="stylesheet" type="text/css" href="_static/styles/furo-extensions.css?digest=30d1aed668e5c3a91c3e3bf6a60b675221979f0e" />
    
    


<style>
  body {
    --color-code-background: #f8f8f8;
  --color-code-foreground: black;
  
  }
  @media not print {
    body[data-theme="dark"] {
      --color-code-background: #202020;
  --color-code-foreground: #d0d0d0;
  
    }
    @media (prefers-color-scheme: dark) {
      body:not([data-theme="light"]) {
        --color-code-background: #202020;
  --color-code-foreground: #d0d0d0;
  
      }
    }
  }
</style></head>
  <body>
    
    <script>
      document.body.dataset.theme = localStorage.getItem("theme") || "auto";
    </script>
    

<svg xmlns="http://www.w3.org/2000/svg" style="display: none;">
  <symbol id="svg-toc" viewBox="0 0 24 24">
    <title>Contents</title>
    <svg stroke="currentColor" fill="currentColor" stroke-width="0" viewBox="0 0 1024 1024">
      <path d="M408 442h480c4.4 0 8-3.6 8-8v-56c0-4.4-3.6-8-8-8H408c-4.4 0-8 3.6-8 8v56c0 4.4 3.6 8 8 8zm-8 204c0 4.4 3.6 8 8 8h480c4.4 0 8-3.6 8-8v-56c0-4.4-3.6-8-8-8H408c-4.4 0-8 3.6-8 8v56zm504-486H120c-4.4 0-8 3.6-8 8v56c0 4.4 3.6 8 8 8h784c4.4 0 8-3.6 8-8v-56c0-4.4-3.6-8-8-8zm0 632H120c-4.4 0-8 3.6-8 8v56c0 4.4 3.6 8 8 8h784c4.4 0 8-3.6 8-8v-56c0-4.4-3.6-8-8-8zM115.4 518.9L271.7 642c5.8 4.6 14.4.5 14.4-6.9V388.9c0-7.4-8.5-11.5-14.4-6.9L115.4 505.1a8.74 8.74 0 0 0 0 13.8z"/>
    </svg>
  </symbol>
  <symbol id="svg-menu" viewBox="0 0 24 24">
    <title>Menu</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="2" stroke-linecap="round" stroke-linejoin="round" class="feather-menu">
      <line x1="3" y1="12" x2="21" y2="12"></line>
      <line x1="3" y1="6" x2="21" y2="6"></line>
      <line x1="3" y1="18" x2="21" y2="18"></line>
    </svg>
  </symbol>
  <symbol id="svg-arrow-right" viewBox="0 0 24 24">
    <title>Expand</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="2" stroke-linecap="round" stroke-linejoin="round" class="feather-chevron-right">
      <polyline points="9 18 15 12 9 6"></polyline>
    </svg>
  </symbol>
  <symbol id="svg-sun" viewBox="0 0 24 24">
    <title>Light mode</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="1.5" stroke-linecap="round" stroke-linejoin="round" class="feather-sun">
      <circle cx="12" cy="12" r="5"></circle>
      <line x1="12" y1="1" x2="12" y2="3"></line>
      <line x1="12" y1="21" x2="12" y2="23"></line>
      <line x1="4.22" y1="4.22" x2="5.64" y2="5.64"></line>
      <line x1="18.36" y1="18.36" x2="19.78" y2="19.78"></line>
      <line x1="1" y1="12" x2="3" y2="12"></line>
      <line x1="21" y1="12" x2="23" y2="12"></line>
      <line x1="4.22" y1="19.78" x2="5.64" y2="18.36"></line>
      <line x1="18.36" y1="5.64" x2="19.78" y2="4.22"></line>
    </svg>
  </symbol>
  <symbol id="svg-moon" viewBox="0 0 24 24">
    <title>Dark mode</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="1.5" stroke-linecap="round" stroke-linejoin="round" class="icon-tabler-moon">
      <path stroke="none" d="M0 0h24v24H0z" fill="none" />
      <path d="M12 3c.132 0 .263 0 .393 0a7.5 7.5 0 0 0 7.92 12.446a9 9 0 1 1 -8.313 -12.454z" />
    </svg>
  </symbol>
  <symbol id="svg-sun-half" viewBox="0 0 24 24">
    <title>Auto light/dark mode</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="1.5" stroke-linecap="round" stroke-linejoin="round" class="icon-tabler-shadow">
      <path stroke="none" d="M0 0h24v24H0z" fill="none"/>
      <circle cx="12" cy="12" r="9" />
      <path d="M13 12h5" />
      <path d="M13 15h4" />
      <path d="M13 18h1" />
      <path d="M13 9h4" />
      <path d="M13 6h1" />
    </svg>
  </symbol>
</svg>

<input type="checkbox" class="sidebar-toggle" name="__navigation" id="__navigation">
<input type="checkbox" class="sidebar-toggle" name="__toc" id="__toc">
<label class="overlay sidebar-overlay" for="__navigation">
  <div class="visually-hidden">Hide navigation sidebar</div>
</label>
<label class="overlay toc-overlay" for="__toc">
  <div class="visually-hidden">Hide table of contents sidebar</div>
</label>



<div class="page">
  <header class="mobile-header">
    <div class="header-left">
      <label class="nav-overlay-icon" for="__navigation">
        <div class="visually-hidden">Toggle site navigation sidebar</div>
        <i class="icon"><svg><use href="#svg-menu"></use></svg></i>
      </label>
    </div>
    <div class="header-center">
      <a href="index.html"><div class="brand">BALSAMIC  documentation</div></a>
    </div>
    <div class="header-right">
      <div class="theme-toggle-container theme-toggle-header">
        <button class="theme-toggle">
          <div class="visually-hidden">Toggle Light / Dark / Auto color theme</div>
          <svg class="theme-icon-when-auto"><use href="#svg-sun-half"></use></svg>
          <svg class="theme-icon-when-dark"><use href="#svg-moon"></use></svg>
          <svg class="theme-icon-when-light"><use href="#svg-sun"></use></svg>
        </button>
      </div>
      <label class="toc-overlay-icon toc-header-icon" for="__toc">
        <div class="visually-hidden">Toggle table of contents sidebar</div>
        <i class="icon"><svg><use href="#svg-toc"></use></svg></i>
      </label>
    </div>
  </header>
  <aside class="sidebar-drawer">
    <div class="sidebar-container">
      
      <div class="sidebar-sticky"><a class="sidebar-brand" href="index.html">
  
  <div class="sidebar-logo-container">
    <img class="sidebar-logo" src="_static/balsamic_logo.png" alt="Logo"/>
  </div>
  
  <span class="sidebar-brand-text">BALSAMIC  documentation</span>
  
</a><form class="sidebar-search-container" method="get" action="search.html" role="search">
  <input class="sidebar-search" placeholder="Search" name="q" aria-label="Search">
  <input type="hidden" name="check_keywords" value="yes">
  <input type="hidden" name="area" value="default">
</form>
<div id="searchbox"></div><div class="sidebar-scroll"><div class="sidebar-tree">
  <p class="caption" role="heading"><span class="caption-text">Installation and Usage</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="install.html">Installation</a></li>
<li class="toctree-l1"><a class="reference internal" href="user_guide.html">Short tutorial</a></li>
<li class="toctree-l1"><a class="reference internal" href="cli_package.html">Command line arguments</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">Analysis and Interpretation</span></p>
<ul class="current">
<li class="toctree-l1 current current-page"><a class="current reference internal" href="#">Calling and filtering variants</a></li>
<li class="toctree-l1"><a class="reference internal" href="balsamic_sv_cnv.html">Structural and Copy Number variants</a></li>
<li class="toctree-l1"><a class="reference internal" href="balsamic_annotation.html">Annotation resources</a></li>
<li class="toctree-l1"><a class="reference internal" href="balsamic_pon.html">Panel of Normals (PON)</a></li>
<li class="toctree-l1"><a class="reference internal" href="balsamic_pon.html#cnvkit-pon">CNVkit PON</a></li>
<li class="toctree-l1"><a class="reference internal" href="balsamic_pon.html#gens-pon">GENS PON</a></li>
<li class="toctree-l1"><a class="reference internal" href="balsamic_methods.html">Method description</a></li>
<li class="toctree-l1"><a class="reference internal" href="history.html">Changelog</a></li>
<li class="toctree-l1"><a class="reference internal" href="bioinfo_softwares.html">Tools and software</a></li>
<li class="toctree-l1"><a class="reference internal" href="resources.html">References and other resources</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">Development guide</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="README.html">Documentation Guidline</a></li>
<li class="toctree-l1"><a class="reference internal" href="programming_guide.html">Coding etiquettes</a></li>
<li class="toctree-l1"><a class="reference internal" href="semver.html">Semantic versioning</a></li>
<li class="toctree-l1"><a class="reference internal" href="FAQs.html">Frequently Asked Questions (FAQs)</a></li>
</ul>

</div>
</div>

      </div>
      
    </div>
  </aside>
  <div class="main">
    <div class="content">
      <div class="article-container">
        <a href="#" class="back-to-top muted-link">
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24">
            <path d="M13 20h-2V8l-5.5 5.5-1.42-1.42L12 4.16l7.92 7.92-1.42 1.42L13 8v12z"></path>
          </svg>
          <span>Back to top</span>
        </a>
        <div class="content-icon-container">
          
<div class="theme-toggle-container theme-toggle-content">
            <button class="theme-toggle">
              <div class="visually-hidden">Toggle Light / Dark / Auto color theme</div>
              <svg class="theme-icon-when-auto"><use href="#svg-sun-half"></use></svg>
              <svg class="theme-icon-when-dark"><use href="#svg-moon"></use></svg>
              <svg class="theme-icon-when-light"><use href="#svg-sun"></use></svg>
            </button>
          </div>
          <label class="toc-overlay-icon toc-content-icon" for="__toc">
            <div class="visually-hidden">Toggle table of contents sidebar</div>
            <i class="icon"><svg><use href="#svg-toc"></use></svg></i>
          </label>
        </div>
        <article role="main">
          <section id="calling-and-filtering-variants">
<h1>Calling and filtering variants<a class="headerlink" href="#calling-and-filtering-variants" title="Permalink to this heading">#</a></h1>
<p>In BALSAMIC, various bioinfo tools are integrated for reporting somatic and germline variants summarized in the table below. The choice of these tools differs between the type of analysis; <cite>Whole Genome Sequencing (WGS)</cite>, or <cite>Target Genome Analysis (TGA)</cite> and <cite>Target Genome Analysis (TGA) with UMI-analysis activated</cite>.</p>
<div class="table-wrapper colwidths-given docutils container" id="id3">
<table class="docutils align-default" id="id3">
<caption><span class="caption-text">SNV and small-Indel callers</span><a class="headerlink" href="#id3" title="Permalink to this table">#</a></caption>
<colgroup>
<col style="width: 19%" />
<col style="width: 24%" />
<col style="width: 22%" />
<col style="width: 18%" />
<col style="width: 18%" />
</colgroup>
<thead>
<tr class="row-odd"><th class="head"><p>Variant caller</p></th>
<th class="head"><p>Sequencing type</p></th>
<th class="head"><p>Analysis type</p></th>
<th class="head"><p>Somatic/Germline</p></th>
<th class="head"><p>Variant type</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p>DNAscope</p></td>
<td><p>TGA, WGS</p></td>
<td><p>tumor-normal, tumor-only</p></td>
<td><p>germline</p></td>
<td><p>SNV, InDel</p></td>
</tr>
<tr class="row-odd"><td><p>TNscope</p></td>
<td><p>WGS, TGA (with UMIs activated)</p></td>
<td><p>tumor-normal, tumor-only</p></td>
<td><p>somatic</p></td>
<td><p>SNV, InDel</p></td>
</tr>
<tr class="row-even"><td><p>VarDict</p></td>
<td><p>TGA</p></td>
<td><p>tumor-normal, tumor-only</p></td>
<td><p>somatic</p></td>
<td><p>SNV, InDel</p></td>
</tr>
</tbody>
</table>
</div>
<p>Various filters (Pre-call and Post-call filtering) are applied at different levels to report high-confidence variant calls.</p>
<p><strong>Pre-call filtering</strong> is where the variant-calling tool decides not to add a variant to the VCF file if the default filters of the variant-caller did not pass the filter criteria. The set of default filters differs between the various variant-calling algorithms.</p>
<p>To know more about the pre-call filters used by the variant callers, please have a look at the VCF header of the particular variant-calling results.
For example:</p>
<figure class="align-default" id="id4">
<a class="reference internal image-reference" href="_images/vcf_filters.png"><img alt="_images/vcf_filters.png" src="_images/vcf_filters.png" style="width: 500px;" /></a>
<figcaption>
<p><span class="caption-text">Pre-call filters applied by the <cite>Vardict</cite> variant-caller is listed in the VCF header.</span><a class="headerlink" href="#id4" title="Permalink to this image">#</a></p>
</figcaption>
</figure>
<p>In the VCF file, the <cite>FILTER</cite> status is <cite>PASS</cite> if this position has passed all filters, i.e., a call is made at this position. Contrary,
if the site has not passed any of the filters, a semicolon-separated list of those failed filter(s) will be appended to the <cite>FILTER</cite> column instead of <cite>PASS</cite>. E.g., <cite>p8;pSTD</cite> might
indicate that at this site, the mean position in reads is less than 8, and the position in reads has a standard deviation of 0.</p>
<p><strong>Note:</strong>
<strong>In BALSAMIC, this VCF file is named as `*.&lt;research/clinical&gt;.vcf.gz` (eg: `SNV.somatic.&lt;CASE_ID&gt;.vardict.&lt;research/clinical&gt;.vcf.gz`)</strong></p>
<figure class="align-default" id="id5">
<a class="reference internal image-reference" href="_images/filter_status.png"><img alt="_images/filter_status.png" src="_images/filter_status.png" style="width: 500px;" /></a>
<figcaption>
<p><span class="caption-text">Vardict Variant calls with different ‘FILTER’ status underlined in white line (<cite>NM4.5</cite>, <cite>PASS</cite>, <cite>p8;pSTD</cite>)</span><a class="headerlink" href="#id5" title="Permalink to this image">#</a></p>
</figcaption>
</figure>
<p><strong>Post-call filtering</strong> is where a variant is further filtered with quality, depth, VAF, etc., with more stringent thresholds.</p>
<p>For <cite>Post-call filtering</cite>, in BALSAMIC we have applied various filtering criteria (<a class="reference internal" href="#vardict-filtering">Vardict_filtering</a>, <a class="reference internal" href="#tnscope-filtering-tumor-normal">TNscope filtering (Tumor_normal)</a> ) depending on the analysis-type (TGS/WGS) and sample-type(tumor-only/tumor-normal).</p>
<p><strong>Note:</strong>
<strong>In BALSAMIC, this VCF file is named as `*.&lt;research/clinical&gt;.filtered.vcf.gz` (eg: `SNV.somatic.&lt;CASE_ID&gt;.vardict.&lt;research/clinical&gt;.filtered.vcf.gz`)</strong></p>
<p>Only those variants that fulfill the pre-call and post-call filters are scored as <cite>PASS</cite> in the <cite>STATUS</cite> column of the VCF file. We filter those <cite>PASS</cite> variants and deliver a final list of variants to the customer either via <cite>Scout</cite> or <cite>Caesar</cite></p>
<p><strong>Note:</strong>
<strong>In BALSAMIC, this VCF file is named as `*.&lt;research/clinical&gt;.filtered.pass.vcf.gz` (eg: `SNV.somatic.&lt;CASE_ID&gt;.vardict.&lt;research/clinical&gt;.filtered.pass.vcf.gz`)</strong></p>
<div class="table-wrapper colwidths-given docutils container" id="id6">
<table class="docutils align-default" id="id6">
<caption><span class="caption-text">Description of VCF files</span><a class="headerlink" href="#id6" title="Permalink to this table">#</a></caption>
<colgroup>
<col style="width: 30%" />
<col style="width: 50%" />
<col style="width: 20%" />
</colgroup>
<thead>
<tr class="row-odd"><th class="head"><p>VCF file name</p></th>
<th class="head"><p>Description</p></th>
<th class="head"><p>Delivered to the customer</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p>.vcf.gz</p></td>
<td><p>Unannotated VCF file with pre-call filters included in the STATUS column</p></td>
<td><p>Yes (Caesar)</p></td>
</tr>
<tr class="row-odd"><td><p>.&lt;research/clinical&gt;.vcf.gz</p></td>
<td><p>Annotated VCF file with pre-call filters included in the STATUS column</p></td>
<td><p>No</p></td>
</tr>
<tr class="row-even"><td><p>.&lt;research/clinical&gt;.filtered.pass.vcf.gz</p></td>
<td><p>Annotated and filtered VCF file by excluding all filters that did not meet the pre and post-call filter criteria. Includes only variants with the <cite>PASS</cite> STATUS</p></td>
<td><p>Yes (Caesar and Scout)</p></td>
</tr>
</tbody>
</table>
</div>
<section id="targeted-genome-analysis">
<h2><strong>Targeted Genome Analysis</strong><a class="headerlink" href="#targeted-genome-analysis" title="Permalink to this heading">#</a></h2>
<section id="somatic-callers-for-reporting-snvs-indels">
<h3>Somatic Callers for reporting SNVs/INDELS<a class="headerlink" href="#somatic-callers-for-reporting-snvs-indels" title="Permalink to this heading">#</a></h3>
<section id="vardict">
<h4><strong>Vardict</strong><a class="headerlink" href="#vardict" title="Permalink to this heading">#</a></h4>
<p><a class="reference external" href="https://github.com/AstraZeneca-NGS/VarDict">Vardict</a> is a sensitive variant caller used for both tumor-only and tumor-normal variant calling.
The results of <cite>Vardict</cite> variant calling are further post-filtered based on several criteria (<a class="reference internal" href="#vardict-filtering">Vardict_filtering</a>) to retrieve high-confidence variant calls.
These high-confidence variant calls are the final list of variants uploaded to Scout or available in the delivered VCF file in Caesar.</p>
<p>There are two slightly different post-processing filters used depending on if the flag “–exome” is supplied at the start, which is designed to accomodate targeted analyses of lower depths.</p>
<section id="vardict-filtering">
<h5><strong>Vardict_filtering</strong><a class="headerlink" href="#vardict-filtering" title="Permalink to this heading">#</a></h5>
<p>Following is the set of criteria applied for filtering vardict results. It is used for both tumor-normal and tumor-only samples.</p>
<p><strong>Post-call Quality Filters for panels</strong></p>
<p><em>Mean Mapping Quality (MQ)</em>: Refers to the root mean square (RMS) mapping quality of all the reads spanning the given variant site.</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">MQ</span> <span class="o">&gt;=</span> <span class="mi">30</span>
</pre></div>
</div>
<p><em>Total Depth (DP)</em>: Refers to the overall read depth supporting the called variant.</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">DP</span> <span class="o">&gt;=</span> <span class="mi">100</span>
</pre></div>
</div>
<p><em>Variant depth (VD)</em>: Total reads supporting the ALT allele</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">VD</span> <span class="o">&gt;=</span> <span class="mi">5</span>
</pre></div>
</div>
<p><em>Allelic Frequency (AF)</em>: Fraction of the reads supporting the alternate allele</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">Minimum</span> <span class="n">AF</span> <span class="o">&gt;=</span> <span class="mf">0.007</span>
</pre></div>
</div>
<p><strong>Post-call Quality Filters for exomes</strong></p>
<p><em>Mean Mapping Quality (MQ)</em>: Refers to the root mean square (RMS) mapping quality of all the reads spanning the given variant site.</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">MQ</span> <span class="o">&gt;=</span> <span class="mi">30</span>
</pre></div>
</div>
<p><em>Total Depth (DP)</em>: Refers to the overall read depth supporting the called variant.</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">DP</span> <span class="o">&gt;=</span> <span class="mi">20</span>
</pre></div>
</div>
<p><em>Variant depth (VD)</em>: Total reads supporting the ALT allele</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">VD</span> <span class="o">&gt;=</span> <span class="mi">5</span>
</pre></div>
</div>
<p><em>Allelic Frequency (AF)</em>: Fraction of the reads supporting the alternate allele</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">Minimum</span> <span class="n">AF</span> <span class="o">&gt;=</span> <span class="mf">0.007</span>
</pre></div>
</div>
<p><strong>Note:</strong>
<strong>Additionally, the variant is excluded for tumor-normal cases if marked as ‘germline’ in the `STATUS` column of the VCF file.</strong></p>
<p><strong>Attention:</strong>
<strong>BALSAMIC &lt;= v8.2.7 uses minimum AF 1% (0.01). From Balsamic v8.2.8, minimum VAF is changed to 0.7% (0.007)</strong></p>
<p><strong>Post-call Observation database Filters</strong></p>
<p><em>GNOMADAF_POPMAX</em>: Maximum Allele Frequency across populations</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">GNOMADAF_popmax</span> <span class="o">&lt;=</span> <span class="mf">0.005</span>  <span class="p">(</span><span class="ow">or</span><span class="p">)</span> <span class="n">GNOMADAF_popmax</span> <span class="o">==</span> <span class="s2">&quot;.&quot;</span>
</pre></div>
</div>
<p><em>SWEGENAF</em>: SweGen Allele Frequency</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">SWEGENAF</span> <span class="o">&lt;=</span> <span class="mf">0.01</span>  <span class="p">(</span><span class="ow">or</span><span class="p">)</span> <span class="n">SWEGENAF</span> <span class="o">==</span> <span class="s2">&quot;.&quot;</span>
</pre></div>
</div>
<p><em>Frq</em>: Frequency of observation of the variants from normal <cite>Clinical</cite> samples</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">Frq</span> <span class="o">&lt;=</span> <span class="mf">0.01</span>  <span class="p">(</span><span class="ow">or</span><span class="p">)</span> <span class="n">Frq</span> <span class="o">==</span> <span class="s2">&quot;.&quot;</span>
</pre></div>
</div>
</section>
</section>
</section>
<section id="target-genome-analysis-with-umi-s-into-account">
<h3><strong>Target Genome Analysis with UMI’s into account</strong><a class="headerlink" href="#target-genome-analysis-with-umi-s-into-account" title="Permalink to this heading">#</a></h3>
<section id="sentieon-s-tnscope">
<h4><strong>Sentieon’s TNscope</strong><a class="headerlink" href="#sentieon-s-tnscope" title="Permalink to this heading">#</a></h4>
<p><a class="reference external" href="https://balsamic.readthedocs.io/en/latest/FAQs.html">UMI workflow</a> performs the variant calling of SNVs/INDELS using the <cite>TNscope</cite> algorithm from UMI consensus-called reads.
The following filter applies for both tumor-normal and tumor-only samples.</p>
<p><strong>Pre-call Filters</strong></p>
<p><em>minreads</em>: Filtering of consensus called reads based on the minimum reads supporting each UMI tag group</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">minreads</span> <span class="o">=</span> <span class="mi">3</span><span class="p">,</span><span class="mi">1</span><span class="p">,</span><span class="mi">1</span>
</pre></div>
</div>
<p>It means that at least <cite>3</cite> UMI tag groups should be ideally considered from both DNA strands, where a minimum of at least <cite>1</cite> UMI tag group should exist in each of the single-stranded consensus reads.</p>
<p><em>min_init_tumor_lod</em>: Initial Log odds for the that the variant exists in the tumor.</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">min_init_tumor_lod</span> <span class="o">=</span> <span class="mf">0.5</span>
</pre></div>
</div>
<p><em>min_tumor_lod</em>: Minimum log odds in the final call of variant in the tumor.</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">min_tumor_lod</span> <span class="o">=</span> <span class="mf">4.0</span>
</pre></div>
</div>
<p><em>min_tumor_allele_frac</em>: Set the minimum tumor AF to be considered as potential variant site.</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">min_tumor_allele_frac</span> <span class="o">=</span> <span class="mf">0.0005</span>
</pre></div>
</div>
<p><em>interval_padding</em>:  Adding an extra 100bp to each end of the target region in the bed file before variant calling.</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">interval_padding</span> <span class="o">=</span> <span class="mi">100</span>
</pre></div>
</div>
<p><strong>Post-call Quality Filters</strong></p>
<p><em>alt_allele_in_normal</em>: Default filter set by TNscope was considered too stringent in filtering tumor in normal and is removed.</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">bcftools</span> <span class="n">annotate</span> <span class="o">-</span><span class="n">x</span> <span class="n">FILTER</span><span class="o">/</span><span class="n">alt_allele_in_normal</span>
</pre></div>
</div>
<p><em>Relative tumor AF in normal</em>: Allows for maximum Tumor-In-Normal-Contamination of 30%.</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">excludes</span> <span class="n">variant</span> <span class="k">if</span><span class="p">:</span> <span class="n">AF</span><span class="p">(</span><span class="n">normal</span><span class="p">)</span> <span class="o">/</span> <span class="n">AF</span><span class="p">(</span><span class="n">tumor</span><span class="p">)</span> <span class="o">&gt;</span> <span class="mf">0.3</span>
</pre></div>
</div>
<p><strong>Post-call Observation database Filters</strong></p>
<p><em>GNOMADAF_POPMAX</em>: Maximum Allele Frequency across populations</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">GNOMADAF_popmax</span> <span class="o">&lt;=</span> <span class="mf">0.02</span> <span class="p">(</span><span class="ow">or</span><span class="p">)</span> <span class="n">GNOMADAF_popmax</span> <span class="o">==</span> <span class="s2">&quot;.&quot;</span>
</pre></div>
</div>
<p><em>SWEGENAF</em>: SweGen Allele Frequency</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">SWEGENAF</span> <span class="o">&lt;=</span> <span class="mf">0.01</span>  <span class="p">(</span><span class="ow">or</span><span class="p">)</span> <span class="n">SWEGENAF</span> <span class="o">==</span> <span class="s2">&quot;.&quot;</span>
</pre></div>
</div>
<p><em>Frq</em>: Frequency of observation of the variants from normal <cite>Clinical</cite> samples</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">Frq</span> <span class="o">&lt;=</span> <span class="mf">0.01</span>  <span class="p">(</span><span class="ow">or</span><span class="p">)</span> <span class="n">Frq</span> <span class="o">==</span> <span class="s2">&quot;.&quot;</span>
</pre></div>
</div>
<p>The variants scored as <cite>PASS</cite> or <cite>triallelic_sites</cite> are included in the final vcf file (<cite>SNV.somatic.&lt;CASE_ID&gt;.tnscope.&lt;research/clinical&gt;.filtered.pass.vcf.gz</cite>).</p>
</section>
</section>
<section id="whole-genome-sequencing-wgs">
<h3><strong>Whole Genome Sequencing (WGS)</strong><a class="headerlink" href="#whole-genome-sequencing-wgs" title="Permalink to this heading">#</a></h3>
<section id="id2">
<h4><strong>Sentieon’s TNscope</strong><a class="headerlink" href="#id2" title="Permalink to this heading">#</a></h4>
<p>BALSAMIC utilizes the <cite>TNscope</cite> algorithm for calling somatic SNVs and INDELS in WGS samples.
The <a class="reference external" href="https://www.biorxiv.org/content/10.1101/250647v1.abstract">TNscope</a> algorithm performs the somatic variant calling on the tumor-normal or the tumor-only samples.</p>
<section id="tnscope-filtering-tumor-normal">
<h5><strong>TNscope filtering (Tumor_normal)</strong><a class="headerlink" href="#tnscope-filtering-tumor-normal" title="Permalink to this heading">#</a></h5>
<p><strong>Pre-call Filters</strong></p>
<p><em>Apply TNscope trained MachineLearning Model</em>: Sets MLrejected on variants with ML_PROB below 0.32.</p>
<dl class="simple">
<dt>::</dt><dd><p>ML model: SentieonTNscopeModel_GiAB_HighAF_LowFP-201711.05.model is applied</p>
</dd>
</dl>
<p><em>min_init_tumor_lod</em>: Initial Log odds for the that the variant exists in the tumor.</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">min_init_tumor_lod</span> <span class="o">=</span> <span class="mi">1</span>
</pre></div>
</div>
<p><em>min_tumor_lod</em>: Minimum log odds in the final call of variant in the tumor.</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">min_tumor_lod</span> <span class="o">=</span> <span class="mi">8</span>
</pre></div>
</div>
<p><em>min_init_normal_lod</em>: Initial Log odds for the that the variant exists in the normal.</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">min_init_normal_lod</span> <span class="o">=</span> <span class="mf">0.5</span>
</pre></div>
</div>
<p><em>min_normal_lod</em>: Minimum log odds in the final call of variant in the normal.</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">min_normal_lod</span> <span class="o">=</span> <span class="mi">1</span>
</pre></div>
</div>
<p><strong>Post-call Quality Filters</strong></p>
<p><em>Total Depth (DP)</em>: Refers to the overall read depth from all target samples supporting the variant call</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">DP</span><span class="p">(</span><span class="n">tumor</span><span class="p">)</span> <span class="o">&gt;=</span> <span class="mi">10</span> <span class="p">(</span><span class="ow">or</span><span class="p">)</span> <span class="n">DP</span><span class="p">(</span><span class="n">normal</span><span class="p">)</span> <span class="o">&gt;=</span> <span class="mi">10</span>
</pre></div>
</div>
<p><em>Allelic Depth (AD)</em>: Total reads supporting the ALT allele in the tumor sample</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">AD</span><span class="p">(</span><span class="n">tumor</span><span class="p">)</span> <span class="o">&gt;=</span> <span class="mi">3</span>
</pre></div>
</div>
<p><em>Allelic Frequency (AF)</em>: Fraction of the reads supporting the alternate allele</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">Minimum</span> <span class="n">AF</span><span class="p">(</span><span class="n">tumor</span><span class="p">)</span> <span class="o">&gt;=</span> <span class="mf">0.05</span>
</pre></div>
</div>
<p><em>alt_allele_in_normal</em>: Default filter set by TNscope was considered too stringent in filtering tumor in normal and is removed.</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">bcftools</span> <span class="n">annotate</span> <span class="o">-</span><span class="n">x</span> <span class="n">FILTER</span><span class="o">/</span><span class="n">alt_allele_in_normal</span>
</pre></div>
</div>
<p><em>Relative tumor AF in normal</em>: Allows for maximum Tumor-In-Normal-Contamination of 30%.</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">excludes</span> <span class="n">variant</span> <span class="k">if</span><span class="p">:</span> <span class="n">AF</span><span class="p">(</span><span class="n">normal</span><span class="p">)</span> <span class="o">/</span> <span class="n">AF</span><span class="p">(</span><span class="n">tumor</span><span class="p">)</span> <span class="o">&gt;</span> <span class="mf">0.3</span>
</pre></div>
</div>
<p><strong>Post-call Observation database Filters</strong></p>
<p><em>GNOMADAF_POPMAX</em>: Maximum Allele Frequency across populations</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">GNOMADAF_popmax</span> <span class="o">&lt;=</span> <span class="mf">0.001</span> <span class="p">(</span><span class="ow">or</span><span class="p">)</span> <span class="n">GNOMADAF_popmax</span> <span class="o">==</span> <span class="s2">&quot;.&quot;</span>
</pre></div>
</div>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">SWEGENAF</span> <span class="o">&lt;=</span> <span class="mf">0.01</span>  <span class="p">(</span><span class="ow">or</span><span class="p">)</span> <span class="n">SWEGENAF</span> <span class="o">==</span> <span class="s2">&quot;.&quot;</span>
</pre></div>
</div>
<p><em>Frq</em>: Frequency of observation of the variants from normal <cite>Clinical</cite> samples</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">Frq</span> <span class="o">&lt;=</span> <span class="mf">0.01</span>  <span class="p">(</span><span class="ow">or</span><span class="p">)</span> <span class="n">Frq</span> <span class="o">==</span> <span class="s2">&quot;.&quot;</span>
</pre></div>
</div>
<p>The variants scored as <cite>PASS</cite> or <cite>triallelic_sites</cite> are included in the final vcf file (<cite>SNV.somatic.&lt;CASE_ID&gt;.tnscope.&lt;research/clinical&gt;.filtered.pass.vcf.gz</cite>).</p>
</section>
<section id="tnscope-filtering-tumor-only">
<h5><strong>TNscope filtering (tumor_only)</strong><a class="headerlink" href="#tnscope-filtering-tumor-only" title="Permalink to this heading">#</a></h5>
<p><strong>Pre-call Filters</strong></p>
<p><em>min_init_tumor_lod</em>: Initial Log odds for the that the variant exists in the tumor.</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">min_init_tumor_lod</span> <span class="o">=</span> <span class="mi">1</span>
</pre></div>
</div>
<p><em>min_tumor_lod</em>: Minimum log odds in the final call of variant in the tumor.</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">min_tumor_lod</span> <span class="o">=</span> <span class="mi">8</span>
</pre></div>
</div>
<p>The somatic variants in TNscope raw VCF file (<cite>SNV.somatic.&lt;CASE_ID&gt;.tnscope.all.vcf.gz</cite>) are filtered out for the genomic regions that are not reliable (eg: centromeric regions, non-chromosome contigs) to enhance the computation time. This WGS interval region file is collected from gatk_bundles <a class="reference external" href="gs://gatk-legacy-bundles/b37/wgs_calling_regions.v1.interval_list">gs://gatk-legacy-bundles/b37/wgs_calling_regions.v1.interval_list</a>.</p>
<p><strong>Post-call Quality Filters</strong></p>
<p><em>Total Depth (DP)</em>: Refers to the overall read depth supporting the variant call</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">DP</span><span class="p">(</span><span class="n">tumor</span><span class="p">)</span> <span class="o">&gt;=</span> <span class="mi">10</span>
</pre></div>
</div>
<p><em>Allelic Depth (AD)</em>: Total reads supporting the ALT allele in the tumor sample</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">AD</span><span class="p">(</span><span class="n">tumor</span><span class="p">)</span> <span class="o">&gt;</span> <span class="mi">3</span>
</pre></div>
</div>
<p><em>Allelic Frequency (AF)</em>: Fraction of the reads supporting the alternate allele</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">Minimum</span> <span class="n">AF</span><span class="p">(</span><span class="n">tumor</span><span class="p">)</span> <span class="o">&gt;</span> <span class="mf">0.05</span>
</pre></div>
</div>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">SUM</span><span class="p">(</span><span class="n">QSS</span><span class="p">)</span><span class="o">/</span><span class="n">SUM</span><span class="p">(</span><span class="n">AD</span><span class="p">)</span> <span class="o">&gt;=</span> <span class="mi">20</span>
</pre></div>
</div>
<p><em>Read Counts</em>: Count of reads in a given (F1R2, F2R1) pair orientation supporting the alternate allele and reference alleles</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">ALT_F1R2</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">,</span> <span class="n">ALT_F2R1</span> <span class="o">&gt;</span> <span class="mi">0</span>
<span class="n">REF_F1R2</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">,</span> <span class="n">REF_F2R1</span> <span class="o">&gt;</span> <span class="mi">0</span>
</pre></div>
</div>
<p><em>SOR</em>: Symmetric Odds Ratio of 2x2 contingency table to detect strand bias</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">SOR</span> <span class="o">&lt;</span> <span class="mi">3</span>
</pre></div>
</div>
<p><strong>Post-call Observation database Filters</strong></p>
<p><em>GNOMADAF_POPMAX</em>: Maximum Allele Frequency across populations</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">GNOMADAF_popmax</span> <span class="o">&lt;=</span> <span class="mf">0.001</span> <span class="p">(</span><span class="ow">or</span><span class="p">)</span> <span class="n">GNOMADAF_popmax</span> <span class="o">==</span> <span class="s2">&quot;.&quot;</span>
</pre></div>
</div>
<p><em>Normalized base quality scores</em>:  The sum of base quality scores for each allele (QSS) is divided by the allelic depth of alt and ref alleles (AD)</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">SWEGENAF</span> <span class="o">&lt;=</span> <span class="mf">0.01</span>  <span class="p">(</span><span class="ow">or</span><span class="p">)</span> <span class="n">SWEGENAF</span> <span class="o">==</span> <span class="s2">&quot;.&quot;</span>
</pre></div>
</div>
<p><em>Frq</em>: Frequency of observation of the variants from normal <cite>Clinical</cite> samples</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">Frq</span> <span class="o">&lt;=</span> <span class="mf">0.01</span>  <span class="p">(</span><span class="ow">or</span><span class="p">)</span> <span class="n">Frq</span> <span class="o">==</span> <span class="s2">&quot;.&quot;</span>
</pre></div>
</div>
<p>The variants scored as <cite>PASS</cite> or <cite>triallelic_sites</cite> are included in the final vcf file (<cite>SNV.somatic.&lt;CASE_ID&gt;.tnscope.&lt;research/clinical&gt;.filtered.pass.vcf.gz</cite>).</p>
<p><strong>Attention:</strong>
<strong>BALSAMIC &lt;= v8.2.10 uses GNOMAD_popmax &lt;= 0.005. From Balsamic v9.0.0, this settings is changed to 0.02, to reduce the stringency.</strong>
<strong>BALSAMIC &gt;= v11.0.0 removes unmapped reads from the bam and cram files for all the workflows.</strong>
<strong>BALSAMIC &gt;= v13.0.0 keeps unmapped reads in bam and cram files for all the workflows.</strong></p>
</section>
</section>
</section>
</section>
</section>

        </article>
      </div>
      <footer>
        
        <div class="related-pages">
          <a class="next-page" href="balsamic_sv_cnv.html">
              <div class="page-info">
                <div class="context">
                  <span>Next</span>
                </div>
                <div class="title">Structural and Copy Number variants</div>
              </div>
              <svg class="furo-related-icon"><use href="#svg-arrow-right"></use></svg>
            </a>
          <a class="prev-page" href="cli_package.html">
              <svg class="furo-related-icon"><use href="#svg-arrow-right"></use></svg>
              <div class="page-info">
                <div class="context">
                  <span>Previous</span>
                </div>
                
                <div class="title">Command line arguments</div>
                
              </div>
            </a>
        </div>
        <div class="bottom-of-page">
          <div class="left-details">
            <div class="copyright">
                Copyright &#169; 2023, Clinical Genomics
            </div>
            Made with <a href="https://www.sphinx-doc.org/">Sphinx</a> and <a class="muted-link" href="https://pradyunsg.me">@pradyunsg</a>'s
            
            <a href="https://github.com/pradyunsg/furo">Furo</a>
            
          </div>
          <div class="right-details">
            
          </div>
        </div>
        
      </footer>
    </div>
    <aside class="toc-drawer">
      
      
      <div class="toc-sticky toc-scroll">
        <div class="toc-title-container">
          <span class="toc-title">
            On this page
          </span>
        </div>
        <div class="toc-tree-container">
          <div class="toc-tree">
            <ul>
<li><a class="reference internal" href="#">Calling and filtering variants</a><ul>
<li><a class="reference internal" href="#targeted-genome-analysis"><strong>Targeted Genome Analysis</strong></a><ul>
<li><a class="reference internal" href="#somatic-callers-for-reporting-snvs-indels">Somatic Callers for reporting SNVs/INDELS</a><ul>
<li><a class="reference internal" href="#vardict"><strong>Vardict</strong></a><ul>
<li><a class="reference internal" href="#vardict-filtering"><strong>Vardict_filtering</strong></a></li>
</ul>
</li>
</ul>
</li>
<li><a class="reference internal" href="#target-genome-analysis-with-umi-s-into-account"><strong>Target Genome Analysis with UMI’s into account</strong></a><ul>
<li><a class="reference internal" href="#sentieon-s-tnscope"><strong>Sentieon’s TNscope</strong></a></li>
</ul>
</li>
<li><a class="reference internal" href="#whole-genome-sequencing-wgs"><strong>Whole Genome Sequencing (WGS)</strong></a><ul>
<li><a class="reference internal" href="#id2"><strong>Sentieon’s TNscope</strong></a><ul>
<li><a class="reference internal" href="#tnscope-filtering-tumor-normal"><strong>TNscope filtering (Tumor_normal)</strong></a></li>
<li><a class="reference internal" href="#tnscope-filtering-tumor-only"><strong>TNscope filtering (tumor_only)</strong></a></li>
</ul>
</li>
</ul>
</li>
</ul>
</li>
</ul>
</li>
</ul>

          </div>
        </div>
      </div>
      
      
    </aside>
  </div>
</div><script data-url_root="./" id="documentation_options" src="_static/documentation_options.js"></script>
    <script src="_static/doctools.js"></script>
    <script src="_static/sphinx_highlight.js"></script>
    <script src="_static/scripts/furo.js"></script>
    </body>
</html>