<!doctype html>
<html class="no-js" lang="en">
  <head><meta charset="utf-8"/>
    <meta name="viewport" content="width=device-width,initial-scale=1"/>
    <meta name="color-scheme" content="light dark"><meta name="generator" content="Docutils 0.18.1: http://docutils.sourceforge.net/" />
<link rel="index" title="Index" href="genindex.html" /><link rel="search" title="Search" href="search.html" /><link rel="next" title="Method description" href="balsamic_methods.html" /><link rel="prev" title="Annotation resources" href="balsamic_annotation.html" />

    <!-- Generated with Sphinx 6.2.1 and Furo 2023.07.26 -->
        <title>Panel of Normals (PON) - BALSAMIC documentation</title>
      <link rel="stylesheet" type="text/css" href="_static/pygments.css" />
    <link rel="stylesheet" type="text/css" href="_static/styles/furo.css?digest=369552022d0b975c8e74270ce6eabe0fb7978f24" />
    <link rel="stylesheet" type="text/css" href="_static/styles/furo-extensions.css?digest=30d1aed668e5c3a91c3e3bf6a60b675221979f0e" />
    
    


<style>
  body {
    --color-code-background: #f8f8f8;
  --color-code-foreground: black;
  
  }
  @media not print {
    body[data-theme="dark"] {
      --color-code-background: #202020;
  --color-code-foreground: #d0d0d0;
  
    }
    @media (prefers-color-scheme: dark) {
      body:not([data-theme="light"]) {
        --color-code-background: #202020;
  --color-code-foreground: #d0d0d0;
  
      }
    }
  }
</style></head>
  <body>
    
    <script>
      document.body.dataset.theme = localStorage.getItem("theme") || "auto";
    </script>
    

<svg xmlns="http://www.w3.org/2000/svg" style="display: none;">
  <symbol id="svg-toc" viewBox="0 0 24 24">
    <title>Contents</title>
    <svg stroke="currentColor" fill="currentColor" stroke-width="0" viewBox="0 0 1024 1024">
      <path d="M408 442h480c4.4 0 8-3.6 8-8v-56c0-4.4-3.6-8-8-8H408c-4.4 0-8 3.6-8 8v56c0 4.4 3.6 8 8 8zm-8 204c0 4.4 3.6 8 8 8h480c4.4 0 8-3.6 8-8v-56c0-4.4-3.6-8-8-8H408c-4.4 0-8 3.6-8 8v56zm504-486H120c-4.4 0-8 3.6-8 8v56c0 4.4 3.6 8 8 8h784c4.4 0 8-3.6 8-8v-56c0-4.4-3.6-8-8-8zm0 632H120c-4.4 0-8 3.6-8 8v56c0 4.4 3.6 8 8 8h784c4.4 0 8-3.6 8-8v-56c0-4.4-3.6-8-8-8zM115.4 518.9L271.7 642c5.8 4.6 14.4.5 14.4-6.9V388.9c0-7.4-8.5-11.5-14.4-6.9L115.4 505.1a8.74 8.74 0 0 0 0 13.8z"/>
    </svg>
  </symbol>
  <symbol id="svg-menu" viewBox="0 0 24 24">
    <title>Menu</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="2" stroke-linecap="round" stroke-linejoin="round" class="feather-menu">
      <line x1="3" y1="12" x2="21" y2="12"></line>
      <line x1="3" y1="6" x2="21" y2="6"></line>
      <line x1="3" y1="18" x2="21" y2="18"></line>
    </svg>
  </symbol>
  <symbol id="svg-arrow-right" viewBox="0 0 24 24">
    <title>Expand</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="2" stroke-linecap="round" stroke-linejoin="round" class="feather-chevron-right">
      <polyline points="9 18 15 12 9 6"></polyline>
    </svg>
  </symbol>
  <symbol id="svg-sun" viewBox="0 0 24 24">
    <title>Light mode</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="1.5" stroke-linecap="round" stroke-linejoin="round" class="feather-sun">
      <circle cx="12" cy="12" r="5"></circle>
      <line x1="12" y1="1" x2="12" y2="3"></line>
      <line x1="12" y1="21" x2="12" y2="23"></line>
      <line x1="4.22" y1="4.22" x2="5.64" y2="5.64"></line>
      <line x1="18.36" y1="18.36" x2="19.78" y2="19.78"></line>
      <line x1="1" y1="12" x2="3" y2="12"></line>
      <line x1="21" y1="12" x2="23" y2="12"></line>
      <line x1="4.22" y1="19.78" x2="5.64" y2="18.36"></line>
      <line x1="18.36" y1="5.64" x2="19.78" y2="4.22"></line>
    </svg>
  </symbol>
  <symbol id="svg-moon" viewBox="0 0 24 24">
    <title>Dark mode</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="1.5" stroke-linecap="round" stroke-linejoin="round" class="icon-tabler-moon">
      <path stroke="none" d="M0 0h24v24H0z" fill="none" />
      <path d="M12 3c.132 0 .263 0 .393 0a7.5 7.5 0 0 0 7.92 12.446a9 9 0 1 1 -8.313 -12.454z" />
    </svg>
  </symbol>
  <symbol id="svg-sun-half" viewBox="0 0 24 24">
    <title>Auto light/dark mode</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="1.5" stroke-linecap="round" stroke-linejoin="round" class="icon-tabler-shadow">
      <path stroke="none" d="M0 0h24v24H0z" fill="none"/>
      <circle cx="12" cy="12" r="9" />
      <path d="M13 12h5" />
      <path d="M13 15h4" />
      <path d="M13 18h1" />
      <path d="M13 9h4" />
      <path d="M13 6h1" />
    </svg>
  </symbol>
</svg>

<input type="checkbox" class="sidebar-toggle" name="__navigation" id="__navigation">
<input type="checkbox" class="sidebar-toggle" name="__toc" id="__toc">
<label class="overlay sidebar-overlay" for="__navigation">
  <div class="visually-hidden">Hide navigation sidebar</div>
</label>
<label class="overlay toc-overlay" for="__toc">
  <div class="visually-hidden">Hide table of contents sidebar</div>
</label>



<div class="page">
  <header class="mobile-header">
    <div class="header-left">
      <label class="nav-overlay-icon" for="__navigation">
        <div class="visually-hidden">Toggle site navigation sidebar</div>
        <i class="icon"><svg><use href="#svg-menu"></use></svg></i>
      </label>
    </div>
    <div class="header-center">
      <a href="index.html"><div class="brand">BALSAMIC  documentation</div></a>
    </div>
    <div class="header-right">
      <div class="theme-toggle-container theme-toggle-header">
        <button class="theme-toggle">
          <div class="visually-hidden">Toggle Light / Dark / Auto color theme</div>
          <svg class="theme-icon-when-auto"><use href="#svg-sun-half"></use></svg>
          <svg class="theme-icon-when-dark"><use href="#svg-moon"></use></svg>
          <svg class="theme-icon-when-light"><use href="#svg-sun"></use></svg>
        </button>
      </div>
      <label class="toc-overlay-icon toc-header-icon no-toc" for="__toc">
        <div class="visually-hidden">Toggle table of contents sidebar</div>
        <i class="icon"><svg><use href="#svg-toc"></use></svg></i>
      </label>
    </div>
  </header>
  <aside class="sidebar-drawer">
    <div class="sidebar-container">
      
      <div class="sidebar-sticky"><a class="sidebar-brand" href="index.html">
  
  <div class="sidebar-logo-container">
    <img class="sidebar-logo" src="_static/balsamic_logo.png" alt="Logo"/>
  </div>
  
  <span class="sidebar-brand-text">BALSAMIC  documentation</span>
  
</a><form class="sidebar-search-container" method="get" action="search.html" role="search">
  <input class="sidebar-search" placeholder="Search" name="q" aria-label="Search">
  <input type="hidden" name="check_keywords" value="yes">
  <input type="hidden" name="area" value="default">
</form>
<div id="searchbox"></div><div class="sidebar-scroll"><div class="sidebar-tree">
  <p class="caption" role="heading"><span class="caption-text">Installation and Usage</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="install.html">Installation</a></li>
<li class="toctree-l1"><a class="reference internal" href="user_guide.html">Short tutorial</a></li>
<li class="toctree-l1"><a class="reference internal" href="cli_package.html">Command line arguments</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">Analysis and Interpretation</span></p>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="balsamic_filters.html">Calling and filtering variants</a></li>
<li class="toctree-l1"><a class="reference internal" href="balsamic_sv_cnv.html">Structural and Copy Number variants</a></li>
<li class="toctree-l1"><a class="reference internal" href="balsamic_annotation.html">Annotation resources</a></li>
<li class="toctree-l1 current current-page"><a class="current reference internal" href="#">Panel of Normals (PON)</a></li>
<li class="toctree-l1"><a class="reference internal" href="#cnvkit-pon">CNVkit PON</a></li>
<li class="toctree-l1"><a class="reference internal" href="#gens-pon">GENS PON</a></li>
<li class="toctree-l1"><a class="reference internal" href="balsamic_methods.html">Method description</a></li>
<li class="toctree-l1"><a class="reference internal" href="history.html">Changelog</a></li>
<li class="toctree-l1"><a class="reference internal" href="bioinfo_softwares.html">Tools and software</a></li>
<li class="toctree-l1"><a class="reference internal" href="resources.html">References and other resources</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">Development guide</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="README.html">Documentation Guidline</a></li>
<li class="toctree-l1"><a class="reference internal" href="programming_guide.html">Coding etiquettes</a></li>
<li class="toctree-l1"><a class="reference internal" href="semver.html">Semantic versioning</a></li>
<li class="toctree-l1"><a class="reference internal" href="FAQs.html">Frequently Asked Questions (FAQs)</a></li>
</ul>

</div>
</div>

      </div>
      
    </div>
  </aside>
  <div class="main">
    <div class="content">
      <div class="article-container">
        <a href="#" class="back-to-top muted-link">
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24">
            <path d="M13 20h-2V8l-5.5 5.5-1.42-1.42L12 4.16l7.92 7.92-1.42 1.42L13 8v12z"></path>
          </svg>
          <span>Back to top</span>
        </a>
        <div class="content-icon-container">
          
<div class="theme-toggle-container theme-toggle-content">
            <button class="theme-toggle">
              <div class="visually-hidden">Toggle Light / Dark / Auto color theme</div>
              <svg class="theme-icon-when-auto"><use href="#svg-sun-half"></use></svg>
              <svg class="theme-icon-when-dark"><use href="#svg-moon"></use></svg>
              <svg class="theme-icon-when-light"><use href="#svg-sun"></use></svg>
            </button>
          </div>
          <label class="toc-overlay-icon toc-content-icon no-toc" for="__toc">
            <div class="visually-hidden">Toggle table of contents sidebar</div>
            <i class="icon"><svg><use href="#svg-toc"></use></svg></i>
          </label>
        </div>
        <article role="main">
          <section id="panel-of-normals-pon">
<h1>Panel of Normals (PON)<a class="headerlink" href="#panel-of-normals-pon" title="Permalink to this heading">#</a></h1>
<p>Currently two PON-methods are implemented in BALSAMIC to correct for biases and normalise coverage values:</p>
<ul class="simple">
<li><p>For producing more accurate CNV variant-calls using <code class="docutils literal notranslate"><span class="pre">CNVkit</span></code> for TGA cases.</p></li>
<li><p>To produce normalised CN-profiles for WGS cases visualised in <code class="docutils literal notranslate"><span class="pre">GENS</span></code>.</p></li>
</ul>
</section>
<section id="cnvkit-pon">
<h1>CNVkit PON<a class="headerlink" href="#cnvkit-pon" title="Permalink to this heading">#</a></h1>
<p>BALSAMIC provides a functionality to generate a Panel of Normals (PON) for more accurate copy-number filtering of false positives and that can be used as an input for the <code class="docutils literal notranslate"><span class="pre">CNVkit</span></code> variant caller.</p>
<p>For a more detailed PON use case, please refer to the following documentation:</p>
<ul class="simple">
<li><p><a class="reference external" href="https://cnvkit.readthedocs.io/en/stable/pipeline.html#paired-or-pooled-normals">CNVkit</a></p></li>
<li><p><a class="reference external" href="https://support.illumina.com/content/dam/illumina-support/help/Illumina_DRAGEN_Bio_IT_Platform_v3_7_1000000141465/Content/SW/Informatics/Dragen/GPipelineVarCalNorm_fDG.htm">Illumina DRAGEN</a></p></li>
<li><p><a class="reference external" href="https://gatk.broadinstitute.org/hc/en-us/articles/360035890631-Panel-of-Normals-PON-">GATK</a></p></li>
</ul>
<section id="pon-generation">
<h2>PON Generation<a class="headerlink" href="#pon-generation" title="Permalink to this heading">#</a></h2>
<p>When creating a new PON reference file, the next steps have to be followed:</p>
<ol class="arabic simple">
<li><p>Identify the samples to be included in the PON and add their <code class="docutils literal notranslate"><span class="pre">fastq</span></code> files to the <code class="docutils literal notranslate"><span class="pre">fastq</span></code> directory</p></li>
</ol>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>One needs to fetch normal samples coming from the same origin, tissue or blood</p>
</div>
<ol class="arabic simple" start="2">
<li><p>Generate the <code class="docutils literal notranslate"><span class="pre">&lt;CASE_ID&gt;_PON.json</span></code> configuration file:</p></li>
</ol>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">balsamic</span> <span class="n">config</span> <span class="n">pon</span> <span class="o">--</span><span class="n">pon</span><span class="o">-</span><span class="n">workflow</span> <span class="n">CNVkit</span> <span class="o">--</span><span class="n">case</span><span class="o">-</span><span class="nb">id</span> <span class="o">&lt;</span><span class="n">CASE_ID</span><span class="o">&gt;</span> <span class="o">--</span><span class="n">balsamic</span><span class="o">-</span><span class="n">cache</span> <span class="o">&lt;/</span><span class="n">path</span><span class="o">/</span><span class="n">reference_cache</span><span class="o">/&gt;</span> <span class="o">--</span><span class="n">analysis</span><span class="o">-</span><span class="nb">dir</span> <span class="o">&lt;/</span><span class="n">path</span><span class="o">/</span><span class="n">analysis</span><span class="o">/&gt;</span> <span class="o">--</span><span class="n">fastq</span><span class="o">-</span><span class="n">path</span> <span class="o">&lt;/</span><span class="n">path</span><span class="o">/</span><span class="n">fastq</span><span class="o">/&gt;</span> <span class="o">--</span><span class="n">panel</span><span class="o">-</span><span class="n">bed</span> <span class="o">&lt;/</span><span class="n">path</span><span class="o">/</span><span class="n">panel</span><span class="o">.</span><span class="n">bed</span><span class="o">&gt;</span>
</pre></div>
</div>
<ol class="arabic simple" start="3">
<li><p>Run the BALSAMIC PON workflow:</p></li>
</ol>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">balsamic</span> <span class="n">run</span> <span class="n">analysis</span> <span class="o">-</span><span class="n">s</span> <span class="o">&lt;/</span><span class="n">path</span><span class="o">/</span><span class="n">analysis</span><span class="o">/&lt;</span><span class="n">CASE_ID</span><span class="o">&gt;/&lt;</span><span class="n">CASE_ID</span><span class="o">&gt;</span><span class="n">_PON</span><span class="o">.</span><span class="n">json</span> <span class="o">-</span><span class="n">r</span>
</pre></div>
</div>
<ol class="arabic simple" start="4">
<li><p>Check for the PON reference finish and output files:</p></li>
</ol>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="o">/</span><span class="n">path</span><span class="o">/</span><span class="n">analysis</span><span class="o">/</span><span class="n">analysis_PON_finish</span>
<span class="o">/</span><span class="n">path</span><span class="o">/</span><span class="n">analysis</span><span class="o">/</span><span class="n">cnv</span><span class="o">/&lt;</span><span class="n">panel_name</span><span class="o">&gt;</span><span class="n">_CNVkit_PON_reference_</span><span class="o">&lt;</span><span class="n">version</span><span class="o">&gt;.</span><span class="n">cnn</span>
</pre></div>
</div>
</section>
<section id="using-the-pon-during-analysis">
<h2>Using the PON during analysis<a class="headerlink" href="#using-the-pon-during-analysis" title="Permalink to this heading">#</a></h2>
<p>BALSAMIC can use a PON reference file if its provided while running CNVkit analysis:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">balsamic</span> <span class="n">config</span> <span class="n">case</span> <span class="o">--</span><span class="n">case</span><span class="o">-</span><span class="nb">id</span> <span class="o">&lt;</span><span class="n">CASE_ID</span><span class="o">&gt;</span> <span class="o">--</span><span class="n">pon</span><span class="o">-</span><span class="n">cnn</span> <span class="o">/</span><span class="n">path</span><span class="o">/</span><span class="n">analysis</span><span class="o">/</span><span class="n">cnv</span><span class="o">/&lt;</span><span class="n">panel_name</span><span class="o">&gt;</span><span class="n">_CNVkit_PON_reference_</span><span class="o">&lt;</span><span class="n">version</span><span class="o">&gt;.</span><span class="n">cnn</span> <span class="o">--</span><span class="n">balsamic</span><span class="o">-</span><span class="n">cache</span> <span class="o">&lt;/</span><span class="n">path</span><span class="o">/</span><span class="n">reference_cache</span><span class="o">/&gt;</span> <span class="o">--</span><span class="n">analysis</span><span class="o">-</span><span class="nb">dir</span> <span class="o">&lt;/</span><span class="n">path</span><span class="o">/</span><span class="n">analysis</span><span class="o">/&gt;</span> <span class="o">--</span><span class="n">panel</span><span class="o">-</span><span class="n">bed</span> <span class="o">&lt;/</span><span class="n">path</span><span class="o">/</span><span class="n">panel</span><span class="o">.</span><span class="n">bed</span><span class="o">&gt;</span> <span class="o">--</span><span class="n">tumor</span><span class="o">-</span><span class="n">path</span> <span class="o">&lt;/</span><span class="n">path</span><span class="o">/</span><span class="n">tumor</span><span class="o">.</span><span class="n">fastq</span><span class="o">&gt;</span>
</pre></div>
</div>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>In the absence of a PON reference file, CNVkit is capable of generating a flat reference (tumor-only) or normal reference (tumor-normal) file on its own to correct for GC content and regional coverage</p>
</div>
</section>
</section>
<section id="gens-pon">
<h1>GENS PON<a class="headerlink" href="#gens-pon" title="Permalink to this heading">#</a></h1>
<p>In order to produce an accurate CN-profile to visualise in GENS you need to create 2 PONs one for each gender (see instructions below).</p>
<p>The original instructions for how to create this PON, and which has been implemented in this BALSAMIC workflow can be found on the Clinical-Genomics-Lund GENS-repository:</p>
<ul class="simple">
<li><p><a class="reference external" href="https://github.com/Clinical-Genomics-Lund/gens">Clinical-Genomics-Lund-GENS</a></p></li>
</ul>
<p>To create the PON using the GENS PON creation workflow you can follow the guide below.</p>
<section id="id1">
<h2>PON Generation<a class="headerlink" href="#id1" title="Permalink to this heading">#</a></h2>
<p>To create a GENS PON using the BALSAMIC workflow you need to follow these steps:</p>
<ol class="arabic simple">
<li><p>Create a genome-interval file.</p></li>
</ol>
<p><strong>Note:</strong></p>
<blockquote>
<div><p>These are the genome bins within which the coverage will be calculated, and consequently is the lowest resolution of viewing the CN-profile.</p>
</div></blockquote>
<p>This is the setting we used:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">gatk</span> <span class="n">PreprocessIntervals</span> <span class="o">--</span><span class="n">reference</span> <span class="p">[</span><span class="n">ref</span><span class="p">]</span> <span class="o">--</span><span class="nb">bin</span><span class="o">-</span><span class="n">length</span> <span class="mi">100</span> <span class="o">--</span><span class="n">interval</span><span class="o">-</span><span class="n">merging</span><span class="o">-</span><span class="n">rule</span> <span class="n">OVERLAPPING_ONLY</span> <span class="o">-</span><span class="n">O</span> <span class="n">human_g1k_v37_gens_targets_preprocessed_100bp</span><span class="o">.</span><span class="n">interval_list</span>
</pre></div>
</div>
<ol class="arabic simple" start="2">
<li><p>Identify the samples to be included in the PON and add or link their <code class="docutils literal notranslate"><span class="pre">fastq</span></code> files to the <code class="docutils literal notranslate"><span class="pre">fastq</span></code> directory</p></li>
</ol>
<p><strong>Note:</strong></p>
<blockquote>
<div><p>It is recommended to include approximately 100 samples of the same gender, using the same library preparation and sequencing method as your intended analysis-samples.</p>
</div></blockquote>
<ol class="arabic simple" start="2">
<li><p>Generate the <code class="docutils literal notranslate"><span class="pre">&lt;CASE_ID&gt;_PON.json</span></code> configuration file:</p></li>
</ol>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">balsamic</span> <span class="n">config</span> <span class="n">pon</span> <span class="o">--</span><span class="n">pon</span><span class="o">-</span><span class="n">creation</span><span class="o">-</span><span class="nb">type</span> <span class="o">&lt;</span><span class="p">[</span><span class="n">GENS_female</span><span class="p">,</span><span class="n">GENS_male</span><span class="p">]</span><span class="o">&gt;</span> <span class="o">--</span><span class="n">genome</span><span class="o">-</span><span class="n">interval</span> <span class="o">&lt;</span><span class="p">[</span><span class="n">path</span><span class="o">-</span><span class="n">to</span><span class="o">-</span><span class="n">file</span><span class="o">-</span><span class="n">from</span><span class="o">-</span><span class="n">step1</span><span class="p">]</span><span class="o">&gt;</span> <span class="o">--</span><span class="n">case</span><span class="o">-</span><span class="nb">id</span> <span class="o">&lt;</span><span class="n">CASE_ID</span><span class="o">&gt;</span> <span class="o">--</span><span class="n">balsamic</span><span class="o">-</span><span class="n">cache</span> <span class="o">&lt;/</span><span class="n">path</span><span class="o">/</span><span class="n">reference_cache</span><span class="o">/&gt;</span> <span class="o">--</span><span class="n">analysis</span><span class="o">-</span><span class="nb">dir</span> <span class="o">&lt;/</span><span class="n">path</span><span class="o">/</span><span class="n">analysis</span><span class="o">/&gt;</span> <span class="o">--</span><span class="n">fastq</span><span class="o">-</span><span class="n">path</span> <span class="o">&lt;/</span><span class="n">path</span><span class="o">/</span><span class="n">fastq</span><span class="o">/&gt;</span> <span class="o">--</span><span class="n">panel</span><span class="o">-</span><span class="n">bed</span> <span class="o">&lt;/</span><span class="n">path</span><span class="o">/</span><span class="n">panel</span><span class="o">.</span><span class="n">bed</span><span class="o">&gt;</span>
</pre></div>
</div>
<ol class="arabic simple" start="3">
<li><p>Run the BALSAMIC PON workflow:</p></li>
</ol>
<dl class="simple">
<dt><strong>Note:</strong></dt><dd><p>If you are following these instructions using 100 WGS samples, you require access to compute-nodes with a lot of memory (one of our jobs crashed at 117GB).</p>
</dd>
</dl>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">balsamic</span> <span class="n">run</span> <span class="n">analysis</span> <span class="o">-</span><span class="n">s</span> <span class="o">&lt;/</span><span class="n">path</span><span class="o">/</span><span class="n">analysis</span><span class="o">/&lt;</span><span class="n">CASE_ID</span><span class="o">&gt;/&lt;</span><span class="n">CASE_ID</span><span class="o">&gt;</span><span class="n">_PON</span><span class="o">.</span><span class="n">json</span> <span class="o">-</span><span class="n">r</span>
</pre></div>
</div>
<p>This workflow runs trimming and alignment for all samples to be included in the PON. Calculates coverages in bins using <code class="docutils literal notranslate"><span class="pre">GATK</span> <span class="pre">CollectReadCounts</span></code> then creates the PON using all read-counts with the tool <code class="docutils literal notranslate"><span class="pre">GATK</span> <span class="pre">CreateReadCountPanelOfNormals</span></code>.</p>
<ol class="arabic simple" start="4">
<li><p>Check for the PON output files:</p></li>
</ol>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="o">/</span><span class="n">path</span><span class="o">/</span><span class="n">analysis</span><span class="o">/</span><span class="n">analysis_PON_finish</span>
<span class="o">/</span><span class="n">path</span><span class="o">/</span><span class="n">analysis</span><span class="o">/</span><span class="n">cnv</span><span class="o">/</span><span class="n">gens_pon_100bp</span><span class="o">.&lt;</span><span class="n">GENDER</span><span class="o">&gt;.&lt;</span><span class="n">VERSION</span><span class="o">&gt;.</span><span class="n">hdf5</span>
</pre></div>
</div>
</section>
<section id="id2">
<h2>Using the PON during analysis<a class="headerlink" href="#id2" title="Permalink to this heading">#</a></h2>
<p>This PON is a required input in order to produce the final output-files to be loaded into the GENS platform.</p>
<p>How to run a case using this PON and to activate GENS for your WGS analysis you are referred to this page:</p>
<p><a class="reference external" href="https://balsamic.readthedocs.io/en/latest/balsamic_sv_cnv.html">Using GENS for WGS</a>.</p>
</section>
</section>

        </article>
      </div>
      <footer>
        
        <div class="related-pages">
          <a class="next-page" href="balsamic_methods.html">
              <div class="page-info">
                <div class="context">
                  <span>Next</span>
                </div>
                <div class="title">Method description</div>
              </div>
              <svg class="furo-related-icon"><use href="#svg-arrow-right"></use></svg>
            </a>
          <a class="prev-page" href="balsamic_annotation.html">
              <svg class="furo-related-icon"><use href="#svg-arrow-right"></use></svg>
              <div class="page-info">
                <div class="context">
                  <span>Previous</span>
                </div>
                
                <div class="title">Annotation resources</div>
                
              </div>
            </a>
        </div>
        <div class="bottom-of-page">
          <div class="left-details">
            <div class="copyright">
                Copyright &#169; 2023, Clinical Genomics
            </div>
            Made with <a href="https://www.sphinx-doc.org/">Sphinx</a> and <a class="muted-link" href="https://pradyunsg.me">@pradyunsg</a>'s
            
            <a href="https://github.com/pradyunsg/furo">Furo</a>
            
          </div>
          <div class="right-details">
            
          </div>
        </div>
        
      </footer>
    </div>
    <aside class="toc-drawer no-toc">
      
      
      
    </aside>
  </div>
</div><script data-url_root="./" id="documentation_options" src="_static/documentation_options.js"></script>
    <script src="_static/doctools.js"></script>
    <script src="_static/sphinx_highlight.js"></script>
    <script src="_static/scripts/furo.js"></script>
    </body>
</html>