<!doctype html>
<html class="no-js" lang="en">
  <head><meta charset="utf-8"/>
    <meta name="viewport" content="width=device-width,initial-scale=1"/>
    <meta name="color-scheme" content="light dark"><meta name="generator" content="Docutils 0.18.1: http://docutils.sourceforge.net/" />
<link rel="index" title="Index" href="genindex.html" /><link rel="search" title="Search" href="search.html" /><link rel="next" title="Panel of Normals (PON)" href="balsamic_pon.html" /><link rel="prev" title="Structural and Copy Number variants" href="balsamic_sv_cnv.html" />

    <!-- Generated with Sphinx 6.2.1 and Furo 2023.07.26 -->
        <title>Annotation resources - BALSAMIC documentation</title>
      <link rel="stylesheet" type="text/css" href="_static/pygments.css" />
    <link rel="stylesheet" type="text/css" href="_static/styles/furo.css?digest=369552022d0b975c8e74270ce6eabe0fb7978f24" />
    <link rel="stylesheet" type="text/css" href="_static/styles/furo-extensions.css?digest=30d1aed668e5c3a91c3e3bf6a60b675221979f0e" />
    
    


<style>
  body {
    --color-code-background: #f8f8f8;
  --color-code-foreground: black;
  
  }
  @media not print {
    body[data-theme="dark"] {
      --color-code-background: #202020;
  --color-code-foreground: #d0d0d0;
  
    }
    @media (prefers-color-scheme: dark) {
      body:not([data-theme="light"]) {
        --color-code-background: #202020;
  --color-code-foreground: #d0d0d0;
  
      }
    }
  }
</style></head>
  <body>
    
    <script>
      document.body.dataset.theme = localStorage.getItem("theme") || "auto";
    </script>
    

<svg xmlns="http://www.w3.org/2000/svg" style="display: none;">
  <symbol id="svg-toc" viewBox="0 0 24 24">
    <title>Contents</title>
    <svg stroke="currentColor" fill="currentColor" stroke-width="0" viewBox="0 0 1024 1024">
      <path d="M408 442h480c4.4 0 8-3.6 8-8v-56c0-4.4-3.6-8-8-8H408c-4.4 0-8 3.6-8 8v56c0 4.4 3.6 8 8 8zm-8 204c0 4.4 3.6 8 8 8h480c4.4 0 8-3.6 8-8v-56c0-4.4-3.6-8-8-8H408c-4.4 0-8 3.6-8 8v56zm504-486H120c-4.4 0-8 3.6-8 8v56c0 4.4 3.6 8 8 8h784c4.4 0 8-3.6 8-8v-56c0-4.4-3.6-8-8-8zm0 632H120c-4.4 0-8 3.6-8 8v56c0 4.4 3.6 8 8 8h784c4.4 0 8-3.6 8-8v-56c0-4.4-3.6-8-8-8zM115.4 518.9L271.7 642c5.8 4.6 14.4.5 14.4-6.9V388.9c0-7.4-8.5-11.5-14.4-6.9L115.4 505.1a8.74 8.74 0 0 0 0 13.8z"/>
    </svg>
  </symbol>
  <symbol id="svg-menu" viewBox="0 0 24 24">
    <title>Menu</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="2" stroke-linecap="round" stroke-linejoin="round" class="feather-menu">
      <line x1="3" y1="12" x2="21" y2="12"></line>
      <line x1="3" y1="6" x2="21" y2="6"></line>
      <line x1="3" y1="18" x2="21" y2="18"></line>
    </svg>
  </symbol>
  <symbol id="svg-arrow-right" viewBox="0 0 24 24">
    <title>Expand</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="2" stroke-linecap="round" stroke-linejoin="round" class="feather-chevron-right">
      <polyline points="9 18 15 12 9 6"></polyline>
    </svg>
  </symbol>
  <symbol id="svg-sun" viewBox="0 0 24 24">
    <title>Light mode</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="1.5" stroke-linecap="round" stroke-linejoin="round" class="feather-sun">
      <circle cx="12" cy="12" r="5"></circle>
      <line x1="12" y1="1" x2="12" y2="3"></line>
      <line x1="12" y1="21" x2="12" y2="23"></line>
      <line x1="4.22" y1="4.22" x2="5.64" y2="5.64"></line>
      <line x1="18.36" y1="18.36" x2="19.78" y2="19.78"></line>
      <line x1="1" y1="12" x2="3" y2="12"></line>
      <line x1="21" y1="12" x2="23" y2="12"></line>
      <line x1="4.22" y1="19.78" x2="5.64" y2="18.36"></line>
      <line x1="18.36" y1="5.64" x2="19.78" y2="4.22"></line>
    </svg>
  </symbol>
  <symbol id="svg-moon" viewBox="0 0 24 24">
    <title>Dark mode</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="1.5" stroke-linecap="round" stroke-linejoin="round" class="icon-tabler-moon">
      <path stroke="none" d="M0 0h24v24H0z" fill="none" />
      <path d="M12 3c.132 0 .263 0 .393 0a7.5 7.5 0 0 0 7.92 12.446a9 9 0 1 1 -8.313 -12.454z" />
    </svg>
  </symbol>
  <symbol id="svg-sun-half" viewBox="0 0 24 24">
    <title>Auto light/dark mode</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="1.5" stroke-linecap="round" stroke-linejoin="round" class="icon-tabler-shadow">
      <path stroke="none" d="M0 0h24v24H0z" fill="none"/>
      <circle cx="12" cy="12" r="9" />
      <path d="M13 12h5" />
      <path d="M13 15h4" />
      <path d="M13 18h1" />
      <path d="M13 9h4" />
      <path d="M13 6h1" />
    </svg>
  </symbol>
</svg>

<input type="checkbox" class="sidebar-toggle" name="__navigation" id="__navigation">
<input type="checkbox" class="sidebar-toggle" name="__toc" id="__toc">
<label class="overlay sidebar-overlay" for="__navigation">
  <div class="visually-hidden">Hide navigation sidebar</div>
</label>
<label class="overlay toc-overlay" for="__toc">
  <div class="visually-hidden">Hide table of contents sidebar</div>
</label>



<div class="page">
  <header class="mobile-header">
    <div class="header-left">
      <label class="nav-overlay-icon" for="__navigation">
        <div class="visually-hidden">Toggle site navigation sidebar</div>
        <i class="icon"><svg><use href="#svg-menu"></use></svg></i>
      </label>
    </div>
    <div class="header-center">
      <a href="index.html"><div class="brand">BALSAMIC  documentation</div></a>
    </div>
    <div class="header-right">
      <div class="theme-toggle-container theme-toggle-header">
        <button class="theme-toggle">
          <div class="visually-hidden">Toggle Light / Dark / Auto color theme</div>
          <svg class="theme-icon-when-auto"><use href="#svg-sun-half"></use></svg>
          <svg class="theme-icon-when-dark"><use href="#svg-moon"></use></svg>
          <svg class="theme-icon-when-light"><use href="#svg-sun"></use></svg>
        </button>
      </div>
      <label class="toc-overlay-icon toc-header-icon" for="__toc">
        <div class="visually-hidden">Toggle table of contents sidebar</div>
        <i class="icon"><svg><use href="#svg-toc"></use></svg></i>
      </label>
    </div>
  </header>
  <aside class="sidebar-drawer">
    <div class="sidebar-container">
      
      <div class="sidebar-sticky"><a class="sidebar-brand" href="index.html">
  
  <div class="sidebar-logo-container">
    <img class="sidebar-logo" src="_static/balsamic_logo.png" alt="Logo"/>
  </div>
  
  <span class="sidebar-brand-text">BALSAMIC  documentation</span>
  
</a><form class="sidebar-search-container" method="get" action="search.html" role="search">
  <input class="sidebar-search" placeholder="Search" name="q" aria-label="Search">
  <input type="hidden" name="check_keywords" value="yes">
  <input type="hidden" name="area" value="default">
</form>
<div id="searchbox"></div><div class="sidebar-scroll"><div class="sidebar-tree">
  <p class="caption" role="heading"><span class="caption-text">Installation and Usage</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="install.html">Installation</a></li>
<li class="toctree-l1"><a class="reference internal" href="user_guide.html">Short tutorial</a></li>
<li class="toctree-l1"><a class="reference internal" href="cli_package.html">Command line arguments</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">Analysis and Interpretation</span></p>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="balsamic_filters.html">Calling and filtering variants</a></li>
<li class="toctree-l1"><a class="reference internal" href="balsamic_sv_cnv.html">Structural and Copy Number variants</a></li>
<li class="toctree-l1 current current-page"><a class="current reference internal" href="#">Annotation resources</a></li>
<li class="toctree-l1"><a class="reference internal" href="balsamic_pon.html">Panel of Normals (PON)</a></li>
<li class="toctree-l1"><a class="reference internal" href="balsamic_pon.html#cnvkit-pon">CNVkit PON</a></li>
<li class="toctree-l1"><a class="reference internal" href="balsamic_pon.html#gens-pon">GENS PON</a></li>
<li class="toctree-l1"><a class="reference internal" href="balsamic_methods.html">Method description</a></li>
<li class="toctree-l1"><a class="reference internal" href="history.html">Changelog</a></li>
<li class="toctree-l1"><a class="reference internal" href="bioinfo_softwares.html">Tools and software</a></li>
<li class="toctree-l1"><a class="reference internal" href="resources.html">References and other resources</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">Development guide</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="README.html">Documentation Guidline</a></li>
<li class="toctree-l1"><a class="reference internal" href="programming_guide.html">Coding etiquettes</a></li>
<li class="toctree-l1"><a class="reference internal" href="semver.html">Semantic versioning</a></li>
<li class="toctree-l1"><a class="reference internal" href="FAQs.html">Frequently Asked Questions (FAQs)</a></li>
</ul>

</div>
</div>

      </div>
      
    </div>
  </aside>
  <div class="main">
    <div class="content">
      <div class="article-container">
        <a href="#" class="back-to-top muted-link">
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24">
            <path d="M13 20h-2V8l-5.5 5.5-1.42-1.42L12 4.16l7.92 7.92-1.42 1.42L13 8v12z"></path>
          </svg>
          <span>Back to top</span>
        </a>
        <div class="content-icon-container">
          
<div class="theme-toggle-container theme-toggle-content">
            <button class="theme-toggle">
              <div class="visually-hidden">Toggle Light / Dark / Auto color theme</div>
              <svg class="theme-icon-when-auto"><use href="#svg-sun-half"></use></svg>
              <svg class="theme-icon-when-dark"><use href="#svg-moon"></use></svg>
              <svg class="theme-icon-when-light"><use href="#svg-sun"></use></svg>
            </button>
          </div>
          <label class="toc-overlay-icon toc-content-icon" for="__toc">
            <div class="visually-hidden">Toggle table of contents sidebar</div>
            <i class="icon"><svg><use href="#svg-toc"></use></svg></i>
          </label>
        </div>
        <article role="main">
          <section id="annotation-resources">
<h1>Annotation resources<a class="headerlink" href="#annotation-resources" title="Permalink to this heading">#</a></h1>
<p>BALSAMIC annotates somatic single nucleotide variants (SNVs) using <code class="docutils literal notranslate"><span class="pre">ensembl-vep</span></code> and <code class="docutils literal notranslate"><span class="pre">vcfanno</span></code>. Somatic structural variants (SVs), somatic copy-number variants (CNVs) and germline single nucleotide variants are annotated using only <code class="docutils literal notranslate"><span class="pre">ensembl-vep</span></code>. All SVs and CNVs are merged using <code class="docutils literal notranslate"><span class="pre">SVDB</span></code> before annotating for <cite>Target Genome Analysis (TGA)</cite> or <cite>Whole Genome Sequencing (WGS)</cite> analyses.</p>
<section id="gnomad">
<h2>gnomAD<a class="headerlink" href="#gnomad" title="Permalink to this heading">#</a></h2>
<p><cite>BALSAMIC</cite> adds the following annotation from <cite>gnomAD</cite> database using <code class="docutils literal notranslate"><span class="pre">vcfanno</span></code>.</p>
<div class="table-wrapper colwidths-given docutils container" id="id1">
<table class="docutils align-default" id="id1">
<caption><span class="caption-text">gnomAD annotations</span><a class="headerlink" href="#id1" title="Permalink to this table">#</a></caption>
<colgroup>
<col style="width: 50%" />
<col style="width: 50%" />
</colgroup>
<thead>
<tr class="row-odd"><th class="head"><p>VCF tag</p></th>
<th class="head"><p>description</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p>GNOMADAF_popmax</p></td>
<td><p>maximum allele frequency across populations</p></td>
</tr>
<tr class="row-odd"><td><p>GNOMADAF</p></td>
<td><p>fraction of the reads supporting the alternate allele, allelic frequency</p></td>
</tr>
</tbody>
</table>
</div>
</section>
<section id="clinvar">
<h2>ClinVar<a class="headerlink" href="#clinvar" title="Permalink to this heading">#</a></h2>
<p><cite>BALSAMIC</cite> adds the following annotation from <cite>ClinVar</cite> database using <code class="docutils literal notranslate"><span class="pre">vcfanno</span></code>.</p>
<div class="table-wrapper colwidths-given docutils container" id="id2">
<table class="docutils align-default" id="id2">
<caption><span class="caption-text">ClinVar annotations</span><a class="headerlink" href="#id2" title="Permalink to this table">#</a></caption>
<colgroup>
<col style="width: 50%" />
<col style="width: 50%" />
</colgroup>
<thead>
<tr class="row-odd"><th class="head"><p>VCF tag</p></th>
<th class="head"><p>description</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p>CLNACC</p></td>
<td><p>Variant Accession and Versions</p></td>
</tr>
<tr class="row-odd"><td><p>CLNREVSTAT</p></td>
<td><p>ClinVar review status for the Variation ID</p></td>
</tr>
<tr class="row-even"><td><p>CLNSIG</p></td>
<td><p>Clinical significance for this single variant</p></td>
</tr>
<tr class="row-odd"><td><p>CLNVCSO</p></td>
<td><p>Sequence Ontology id for variant type</p></td>
</tr>
<tr class="row-even"><td><p>CLNVC</p></td>
<td><p>Variant type</p></td>
</tr>
<tr class="row-odd"><td><p>ORIGIN</p></td>
<td><p>Allele origin</p></td>
</tr>
</tbody>
</table>
</div>
<p>The values for <cite>ORIGIN</cite> are described below:</p>
<div class="table-wrapper colwidths-given docutils container" id="id3">
<table class="docutils align-default" id="id3">
<caption><span class="caption-text">ClinVar ORIGIN</span><a class="headerlink" href="#id3" title="Permalink to this table">#</a></caption>
<colgroup>
<col style="width: 50%" />
<col style="width: 50%" />
</colgroup>
<thead>
<tr class="row-odd"><th class="head"><p>Value</p></th>
<th class="head"><p>Annotation</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p>0</p></td>
<td><p>unknown</p></td>
</tr>
<tr class="row-odd"><td><p>1</p></td>
<td><p>germline</p></td>
</tr>
<tr class="row-even"><td><p>2</p></td>
<td><p>somatic</p></td>
</tr>
<tr class="row-odd"><td><p>4</p></td>
<td><p>inherited</p></td>
</tr>
<tr class="row-even"><td><p>8</p></td>
<td><p>paternal</p></td>
</tr>
<tr class="row-odd"><td><p>16</p></td>
<td><p>maternal</p></td>
</tr>
<tr class="row-even"><td><p>32</p></td>
<td><p><em>de-novo</em></p></td>
</tr>
<tr class="row-odd"><td><p>64</p></td>
<td><p>biparental</p></td>
</tr>
<tr class="row-even"><td><p>128</p></td>
<td><p>uniparental</p></td>
</tr>
<tr class="row-odd"><td><p>256</p></td>
<td><p>not-tested</p></td>
</tr>
<tr class="row-even"><td><p>512</p></td>
<td><p>tested-inconclusive</p></td>
</tr>
<tr class="row-odd"><td><p>1073741824</p></td>
<td><p>other</p></td>
</tr>
</tbody>
</table>
</div>
</section>
<section id="cosmic">
<h2>COSMIC<a class="headerlink" href="#cosmic" title="Permalink to this heading">#</a></h2>
<p><cite>BALSAMIC</cite> uses <cite>ensembl-vep</cite> to add the following annotation from <cite>COSMIC</cite> database.</p>
<div class="table-wrapper colwidths-given docutils container" id="id4">
<table class="docutils align-default" id="id4">
<caption><span class="caption-text">COSMIC annotations</span><a class="headerlink" href="#id4" title="Permalink to this table">#</a></caption>
<colgroup>
<col style="width: 50%" />
<col style="width: 50%" />
</colgroup>
<thead>
<tr class="row-odd"><th class="head"><p>VCF tag</p></th>
<th class="head"><p>description</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p>COSMIC_CDS</p></td>
<td><p>CDS annotation</p></td>
</tr>
<tr class="row-odd"><td><p>COSMIC_GENE</p></td>
<td><p>gene name</p></td>
</tr>
<tr class="row-even"><td><p>COSMIC_STRAND</p></td>
<td><p>strand</p></td>
</tr>
<tr class="row-odd"><td><p>COSMIC_CNT</p></td>
<td><p>number of samples with this mutation in the <cite>COSMIC</cite> database</p></td>
</tr>
<tr class="row-even"><td><p>COSMIC_AA</p></td>
<td><p>peptide annotation</p></td>
</tr>
</tbody>
</table>
</div>
</section>
<section id="cadd">
<h2>CADD<a class="headerlink" href="#cadd" title="Permalink to this heading">#</a></h2>
<p><cite>BALSAMIC</cite> adds the following annotation for SNVs from <cite>CADD</cite> database using <code class="docutils literal notranslate"><span class="pre">vcfanno</span></code>.</p>
<div class="table-wrapper colwidths-given docutils container" id="id5">
<table class="docutils align-default" id="id5">
<caption><span class="caption-text">CADD annotations</span><a class="headerlink" href="#id5" title="Permalink to this table">#</a></caption>
<colgroup>
<col style="width: 50%" />
<col style="width: 50%" />
</colgroup>
<thead>
<tr class="row-odd"><th class="head"><p>VCF tag</p></th>
<th class="head"><p>description</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p>CADD</p></td>
<td><p>Combined Annotation Dependent Depletion</p></td>
</tr>
</tbody>
</table>
</div>
</section>
<section id="loqusdb-somatic-frequencies-cancer-cases">
<h2>LoqusDB somatic frequencies (cancer cases)<a class="headerlink" href="#loqusdb-somatic-frequencies-cancer-cases" title="Permalink to this heading">#</a></h2>
<div class="table-wrapper colwidths-given docutils container" id="id6">
<table class="docutils align-default" id="id6">
<caption><span class="caption-text">LoqusDB Somatic Annotations</span><a class="headerlink" href="#id6" title="Permalink to this table">#</a></caption>
<colgroup>
<col style="width: 20%" />
<col style="width: 60%" />
<col style="width: 20%" />
</colgroup>
<thead>
<tr class="row-odd"><th class="head"><p>VCF tag</p></th>
<th class="head"><p>description</p></th>
<th class="head"><p>variant type</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p>Cancer_Somatic_Frq</p></td>
<td><p>Frequency of observation for somatic mutations</p></td>
<td><p>SNV, SV</p></td>
</tr>
<tr class="row-odd"><td><p>Cancer_Somatic_Obs</p></td>
<td><p>allele counts of the somatic variant</p></td>
<td><p>SNV, SV</p></td>
</tr>
<tr class="row-even"><td><p>Cancer_Somatic_Hom</p></td>
<td><p>allele counts of the homozygous somatic variant</p></td>
<td><p>SNV</p></td>
</tr>
</tbody>
</table>
</div>
</section>
<section id="loqusdb-germline-frequencies-cancer-cases">
<h2>LoqusDB germline frequencies (cancer cases)<a class="headerlink" href="#loqusdb-germline-frequencies-cancer-cases" title="Permalink to this heading">#</a></h2>
<div class="table-wrapper colwidths-given docutils container" id="id7">
<table class="docutils align-default" id="id7">
<caption><span class="caption-text">loqusDB germline SNV annotations</span><a class="headerlink" href="#id7" title="Permalink to this table">#</a></caption>
<colgroup>
<col style="width: 20%" />
<col style="width: 60%" />
<col style="width: 20%" />
</colgroup>
<thead>
<tr class="row-odd"><th class="head"><p>VCF tag</p></th>
<th class="head"><p>description</p></th>
<th class="head"><p>variant type</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p>Cancer_Germline_Frq</p></td>
<td><p>Frequency of observation for germline mutations</p></td>
<td><p>SNV</p></td>
</tr>
<tr class="row-odd"><td><p>Cancer_Germline_Obs</p></td>
<td><p>allele counts of the germline variant</p></td>
<td><p>SNV</p></td>
</tr>
<tr class="row-even"><td><p>Cancer_Germline_Hom</p></td>
<td><p>allele counts of the homozygous germline variant</p></td>
<td><p>SNV</p></td>
</tr>
</tbody>
</table>
</div>
</section>
<section id="loqusdb-germline-frequencies-non-cancer-cases">
<h2>LoqusDB germline frequencies (non-cancer cases)<a class="headerlink" href="#loqusdb-germline-frequencies-non-cancer-cases" title="Permalink to this heading">#</a></h2>
<p><cite>BALSAMIC</cite> adds the following annotation from database of <cite>non-cancer clinical</cite> samples using <code class="docutils literal notranslate"><span class="pre">vcfanno</span></code> for SNVs and SVDB for SVs.</p>
<div class="table-wrapper colwidths-given docutils container" id="id8">
<table class="docutils align-default" id="id8">
<caption><span class="caption-text">loqusDB germline (non-cancer) SNV annotations</span><a class="headerlink" href="#id8" title="Permalink to this table">#</a></caption>
<colgroup>
<col style="width: 20%" />
<col style="width: 60%" />
<col style="width: 20%" />
</colgroup>
<thead>
<tr class="row-odd"><th class="head"><p>VCF tag</p></th>
<th class="head"><p>description</p></th>
<th class="head"><p>variant type</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p>Frq</p></td>
<td><p>Frequency of observation of the variants from normal <cite>non-cancer clinical</cite> samples</p></td>
<td><p>SNV, SV</p></td>
</tr>
<tr class="row-odd"><td><p>Obs</p></td>
<td><p>allele counts of the variant in normal <cite>non-cancer clinical</cite> samples</p></td>
<td><p>SNV</p></td>
</tr>
<tr class="row-even"><td><p>Hom</p></td>
<td><p>allele counts of the homozygous variant in normal <cite>non-cancer clinical</cite> samples</p></td>
<td><p>SNV</p></td>
</tr>
<tr class="row-odd"><td><p>clin_obs</p></td>
<td><p>allele counts</p></td>
<td><p>SV</p></td>
</tr>
</tbody>
</table>
</div>
</section>
<section id="swegen">
<h2>SWEGEN<a class="headerlink" href="#swegen" title="Permalink to this heading">#</a></h2>
<p><cite>BALSAMIC</cite> adds the following annotation from <cite>SWEGEN</cite> database using <code class="docutils literal notranslate"><span class="pre">vcfanno</span></code> for SNVs and SVDB for SVs.</p>
<div class="table-wrapper colwidths-given docutils container" id="id9">
<table class="docutils align-default" id="id9">
<caption><span class="caption-text">Swegen SNV annotations</span><a class="headerlink" href="#id9" title="Permalink to this table">#</a></caption>
<colgroup>
<col style="width: 20%" />
<col style="width: 60%" />
<col style="width: 20%" />
</colgroup>
<thead>
<tr class="row-odd"><th class="head"><p>VCF tag</p></th>
<th class="head"><p>description</p></th>
<th class="head"><p>variant type</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p>SWEGENAF</p></td>
<td><p>allele frequency from 1000 Swedish genomes project</p></td>
<td><p>SNV, SV</p></td>
</tr>
<tr class="row-odd"><td><p>SWEGENAAC_Hom</p></td>
<td><p>allele counts of homozygous variants</p></td>
<td><p>SNV</p></td>
</tr>
<tr class="row-even"><td><p>SWEGENAAC_Het</p></td>
<td><p>allele counts of heterozygous variants</p></td>
<td><p>SNV</p></td>
</tr>
<tr class="row-odd"><td><p>SWEGENAAC_Hemi</p></td>
<td><p>allele counts of hemizygous variants</p></td>
<td><p>SNV</p></td>
</tr>
<tr class="row-even"><td><p>swegen_obs</p></td>
<td><p>allele count</p></td>
<td><p>SV</p></td>
</tr>
</tbody>
</table>
</div>
</section>
<section id="ensembl-vep-annotations">
<h2>ENSEMBL-VEP annotations<a class="headerlink" href="#ensembl-vep-annotations" title="Permalink to this heading">#</a></h2>
<p>Where relevant, <cite>BALSAMIC</cite> uses <cite>ensembl-vep</cite> to annotate somatic and germline SNVs and somatic SVs/CNVs from <cite>1000genomes (phase3)</cite>, <cite>ClinVar</cite>, <cite>ESP, HGMD-PUBLIC</cite>, <cite>dbSNP</cite>, <cite>gencode</cite>, <cite>gnomAD</cite>, <cite>polyphen</cite>, <cite>refseq</cite>, and <cite>sift</cite> databases.
The following annotations are added by <cite>ensembl-vep</cite>.</p>
<p>VEP has a setting for the maximum size of a structural variant that it will annotate, currently this is set to the size of the size of chromosome 1 (in hg19) (<cite>–max_sv_size 249250621</cite>).</p>
<div class="table-wrapper colwidths-given docutils container" id="id10">
<table class="docutils align-default" id="id10">
<caption><span class="caption-text">ensembl-vep</span><a class="headerlink" href="#id10" title="Permalink to this table">#</a></caption>
<colgroup>
<col style="width: 14%" />
<col style="width: 86%" />
</colgroup>
<thead>
<tr class="row-odd"><th class="head"><p>Annotation</p></th>
<th class="head"><p>description</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p>Allele</p></td>
<td><p>the variant allele used to calculate the consequence</p></td>
</tr>
<tr class="row-odd"><td><p>Gene</p></td>
<td><p>Ensembl stable ID of affected gene</p></td>
</tr>
<tr class="row-even"><td><p>Feature</p></td>
<td><p>Ensembl stable ID of feature</p></td>
</tr>
<tr class="row-odd"><td><p>Feature type</p></td>
<td><p>type of feature. Currently one of Transcript, RegulatoryFeature, MotifFeature.</p></td>
</tr>
<tr class="row-even"><td><p>Consequence</p></td>
<td><p>consequence type of this variant</p></td>
</tr>
<tr class="row-odd"><td><p>Position in cDNA</p></td>
<td><p>relative position of base pair in cDNA sequence</p></td>
</tr>
<tr class="row-even"><td><p>Position in CDS</p></td>
<td><p>relative position of base pair in coding sequence</p></td>
</tr>
<tr class="row-odd"><td><p>Position in protein</p></td>
<td><p>relative position of amino acid in protein</p></td>
</tr>
<tr class="row-even"><td><p>Amino acid change</p></td>
<td><p>only given if the variant affects the protein-coding sequence</p></td>
</tr>
<tr class="row-odd"><td><p>Codon change</p></td>
<td><p>the alternative codons with the variant base in upper case</p></td>
</tr>
<tr class="row-even"><td><p>Co-located variation</p></td>
<td><p>identifier of any existing variants</p></td>
</tr>
<tr class="row-odd"><td><p>VARIANT_CLASS</p></td>
<td><p>Sequence Ontology variant class</p></td>
</tr>
<tr class="row-even"><td><p>SYMBOL</p></td>
<td><p>the gene symbol</p></td>
</tr>
<tr class="row-odd"><td><p>SYMBOL_SOURCE</p></td>
<td><p>the source of the gene symbol</p></td>
</tr>
<tr class="row-even"><td><p>STRAND</p></td>
<td><p>the DNA strand (1 or -1) on which the transcript/feature lies</p></td>
</tr>
<tr class="row-odd"><td><p>ENSP</p></td>
<td><p>the Ensembl protein identifier of the affected transcript</p></td>
</tr>
<tr class="row-even"><td><p>FLAGS</p></td>
<td><div class="line-block">
<div class="line">transcript quality flags:</div>
<div class="line">cds_start_NF: CDS 5’ incomplete</div>
<div class="line">cds_end_NF: CDS 3’ incomplete</div>
</div>
</td>
</tr>
<tr class="row-odd"><td><p>SWISSPROT</p></td>
<td><p>Best match UniProtKB/Swiss-Prot accession of protein product</p></td>
</tr>
<tr class="row-even"><td><p>TREMBL</p></td>
<td><p>Best match UniProtKB/TrEMBL accession of protein product</p></td>
</tr>
<tr class="row-odd"><td><p>UNIPARC</p></td>
<td><p>Best match UniParc accession of protein product</p></td>
</tr>
<tr class="row-even"><td><p>HGVSc</p></td>
<td><p>the HGVS coding sequence name</p></td>
</tr>
<tr class="row-odd"><td><p>HGVSp</p></td>
<td><p>the HGVS protein sequence name</p></td>
</tr>
<tr class="row-even"><td><p>HGVSg</p></td>
<td><p>the HGVS genomic sequence name</p></td>
</tr>
<tr class="row-odd"><td><p>HGVS_OFFSET</p></td>
<td><p>Indicates by how many bases the HGVS notations for this variant have been shifted</p></td>
</tr>
<tr class="row-even"><td><p>SIFT</p></td>
<td><p>the SIFT prediction and/or score, with both given as prediction(score)</p></td>
</tr>
<tr class="row-odd"><td><p>PolyPhen</p></td>
<td><p>the PolyPhen prediction and/or score</p></td>
</tr>
<tr class="row-even"><td><p>MOTIF_NAME</p></td>
<td><p>The source and identifier of a transcription factor binding profile aligned at this position</p></td>
</tr>
<tr class="row-odd"><td><p>MOTIF_POS</p></td>
<td><p>The relative position of the variation in the aligned TFBP</p></td>
</tr>
<tr class="row-even"><td><p>HIGH_INF_POS</p></td>
<td><p>A flag indicating if the variant falls in a high information position of a transcription factor binding profile (TFBP)</p></td>
</tr>
<tr class="row-odd"><td><p>MOTIF_SCORE_CHANGE</p></td>
<td><p>The difference in motif score of the reference and variant sequences for the TFBP</p></td>
</tr>
<tr class="row-even"><td><p>CANONICAL</p></td>
<td><p>a flag indicating if the transcript is denoted as the canonical transcript for this gene</p></td>
</tr>
<tr class="row-odd"><td><p>CCDS</p></td>
<td><p>the CCDS identifer for this transcript, where applicable</p></td>
</tr>
<tr class="row-even"><td><p>INTRON</p></td>
<td><p>the intron number (out of total number)</p></td>
</tr>
<tr class="row-odd"><td><p>EXON</p></td>
<td><p>the exon number (out of total number)</p></td>
</tr>
<tr class="row-even"><td><p>DOMAINS</p></td>
<td><p>the source and identifer of any overlapping protein domains</p></td>
</tr>
<tr class="row-odd"><td><p>DISTANCE</p></td>
<td><p>Shortest distance from variant to transcript</p></td>
</tr>
<tr class="row-even"><td><p>AF</p></td>
<td><p>Frequency of existing variant in 1000 Genomes</p></td>
</tr>
<tr class="row-odd"><td><p>AFR_AF</p></td>
<td><p>Frequency of existing variant in 1000 Genomes combined African population</p></td>
</tr>
<tr class="row-even"><td><p>AMR_AF</p></td>
<td><p>Frequency of existing variant in 1000 Genomes combined American population</p></td>
</tr>
<tr class="row-odd"><td><p>EUR_AF</p></td>
<td><p>Frequency of existing variant in 1000 Genomes combined European population</p></td>
</tr>
<tr class="row-even"><td><p>EAS_AF</p></td>
<td><p>Frequency of existing variant in 1000 Genomes combined East Asian population</p></td>
</tr>
<tr class="row-odd"><td><p>SAS_AF</p></td>
<td><p>Frequency of existing variant in 1000 Genomes combined South Asian population</p></td>
</tr>
<tr class="row-even"><td><p>AA_AF</p></td>
<td><p>Frequency of existing variant in NHLBI-ESP African American population</p></td>
</tr>
<tr class="row-odd"><td><p>EA_AF</p></td>
<td><p>Frequency of existing variant in NHLBI-ESP European American population</p></td>
</tr>
<tr class="row-even"><td><p>gnomAD_AF</p></td>
<td><p>Frequency of existing variant in gnomAD exomes combined population</p></td>
</tr>
<tr class="row-odd"><td><p>gnomAD_AFR_AF</p></td>
<td><p>Frequency of existing variant in gnomAD exomes African/American population</p></td>
</tr>
<tr class="row-even"><td><p>gnomAD_AMR_AF</p></td>
<td><p>Frequency of existing variant in gnomAD exomes American population</p></td>
</tr>
<tr class="row-odd"><td><p>gnomAD_ASJ_AF</p></td>
<td><p>Frequency of existing variant in gnomAD exomes Ashkenazi Jewish population</p></td>
</tr>
<tr class="row-even"><td><p>gnomAD_EAS_AF</p></td>
<td><p>Frequency of existing variant in gnomAD exomes East Asian population</p></td>
</tr>
<tr class="row-odd"><td><p>gnomAD_FIN_AF</p></td>
<td><p>Frequency of existing variant in gnomAD exomes Finnish population</p></td>
</tr>
<tr class="row-even"><td><p>gnomAD_NFE_AF</p></td>
<td><p>Frequency of existing variant in gnomAD exomes Non-Finnish European population</p></td>
</tr>
<tr class="row-odd"><td><p>gnomAD_OTH_AF</p></td>
<td><p>Frequency of existing variant in gnomAD exomes combined other combined populations</p></td>
</tr>
<tr class="row-even"><td><p>gnomAD_SAS_AF</p></td>
<td><p>Frequency of existing variant in gnomAD exomes South Asian population</p></td>
</tr>
<tr class="row-odd"><td><p>MAX_AF</p></td>
<td><p>Maximum observed allele frequency in 1000 Genomes, ESP and gnomAD</p></td>
</tr>
<tr class="row-even"><td><p>MAX_AF_POPS</p></td>
<td><p>Populations in which maximum allele frequency was observed</p></td>
</tr>
<tr class="row-odd"><td><p>CLIN_SIG</p></td>
<td><p>ClinVar clinical significance of the dbSNP variant</p></td>
</tr>
<tr class="row-even"><td><p>BIOTYPE</p></td>
<td><p>Biotype of transcript or regulatory feature</p></td>
</tr>
<tr class="row-odd"><td><p>APPRIS</p></td>
<td><p>Annotates alternatively spliced transcripts as primary or alternate based on a range of computational methods. NB: not available for GRCh37</p></td>
</tr>
<tr class="row-even"><td><p>TSL</p></td>
<td><p>Transcript support level. NB: not available for GRCh37</p></td>
</tr>
<tr class="row-odd"><td><p>PUBMED</p></td>
<td><p>Pubmed ID(s) of publications that cite existing variant</p></td>
</tr>
<tr class="row-even"><td><p>SOMATIC</p></td>
<td><p>Somatic status of existing variant(s); multiple values correspond to multiple values in the Existing_variation field</p></td>
</tr>
<tr class="row-odd"><td><p>PHENO</p></td>
<td><p>Indicates if existing variant is associated with a phenotype, disease or trait; multiple values correspond to multiple values in the Existing_variation field</p></td>
</tr>
<tr class="row-even"><td><p>GENE_PHENO</p></td>
<td><p>Indicates if overlapped gene is associated with a phenotype, disease or trait</p></td>
</tr>
<tr class="row-odd"><td><p>BAM_EDIT</p></td>
<td><p>Indicates success or failure of edit using BAM file</p></td>
</tr>
<tr class="row-even"><td><p>GIVEN_REF</p></td>
<td><p>Reference allele from input</p></td>
</tr>
<tr class="row-odd"><td><p>REFSEQ_MATCH</p></td>
<td><div class="line-block">
<div class="line">the RefSeq transcript match status; contains a number of flags indicating whether this RefSeq transcript matches the underlying reference sequence and/or an Ensembl transcript (more information):</div>
</div>
<ul class="simple">
<li><p>rseq_3p_mismatch: signifies a mismatch between the RefSeq transcript and the underlying primary genome assembly sequence. Specifically, there is a mismatch in the 3’ UTR of the RefSeq model with respect to the primary genome assembly (e.g. GRCh37/GRCh38).</p></li>
<li><p>rseq_5p_mismatch: signifies a mismatch between the RefSeq transcript and the underlying primary genome assembly sequence. Specifically, there is a mismatch in the 5’ UTR of the RefSeq model with respect to the primary genome assembly.</p></li>
<li><p>rseq_cds_mismatch: signifies a mismatch between the RefSeq transcript and the underlying primary genome assembly sequence. Specifically, there is a mismatch in the CDS of the RefSeq model with respect to the primary genome assembly.</p></li>
<li><p>rseq_ens_match_cds: signifies that for the RefSeq transcript there is an overlapping Ensembl model that is identical across the CDS region only. A CDS match is defined as follows: the CDS and peptide sequences are identical and the genomic coordinates of every translatable exon match. Useful related attributes are: rseq_ens_match_wt and rseq_ens_no_match.</p></li>
<li><p>rseq_ens_match_wt: signifies that for the RefSeq transcript there is an overlapping Ensembl model that is identical across the whole transcript. A whole transcript match is defined as follows: 1) In the case that both models are coding, the transcript, CDS and peptide sequences are all identical and the genomic coordinates of every exon match. 2) In the case that both transcripts are non-coding the transcript sequences and the genomic coordinates of every exon are identical. No comparison is made between a coding and a non-coding transcript. Useful related attributes are: rseq_ens_match_cds and rseq_ens_no_match.</p></li>
<li><p>rseq_ens_no_match: signifies that for the RefSeq transcript there is no overlapping Ensembl model that is identical across either the whole transcript or the CDS. This is caused by differences between the transcript, CDS or peptide sequences or between the exon genomic coordinates. Useful related attributes are: rseq_ens_match_wt and rseq_ens_match_cds.</p></li>
<li><p>rseq_mrna_match: signifies an exact match between the RefSeq transcript and the underlying primary genome assembly sequence (based on a match between the transcript stable id and an accession in the RefSeq mRNA file). An exact match occurs when the underlying genomic sequence of the model can be perfectly aligned to the mRNA sequence post polyA clipping.</p></li>
<li><p>rseq_mrna_nonmatch: signifies a non-match between the RefSeq transcript and the underlying primary genome assembly sequence. A non-match is deemed to have occurred if the underlying genomic sequence does not have a perfect alignment to the mRNA sequence post polyA clipping. It can also signify that no comparison was possible as the model stable id may not have had a corresponding entry in the RefSeq mRNA file (sometimes happens when accessions are retired or changed). When a non-match occurs one or several of the following transcript attributes will also be present to provide more detail on the nature of the non-match: rseq_5p_mismatch, rseq_cds_mismatch, rseq_3p_mismatch, rseq_nctran_mismatch, rseq_no_comparison</p></li>
<li><p>rseq_nctran_mismatch: signifies a mismatch between the RefSeq transcript and the underlying primary genome assembly sequence. This is a comparison between the entire underlying genomic sequence of the RefSeq model to the mRNA in the case of RefSeq models that are non-coding.</p></li>
<li><p>rseq_no_comparison: signifies that no alignment was carried out between the underlying primary genome assembly sequence and a corresponding RefSeq mRNA. The reason for this is generally that no corresponding, unversioned accession was found in the RefSeq mRNA file for the transcript stable id. This sometimes happens when accessions are retired or replaced. A second possibility is that the sequences were too long and problematic to align (though this is rare).</p></li>
</ul>
</td>
</tr>
<tr class="row-even"><td><p>CHECK_REF</p></td>
<td><p>Reports variants where the input reference does not match the expected reference</p></td>
</tr>
<tr class="row-odd"><td><p>HGNC_ID</p></td>
<td><p>A unique ID provided by the HGNC for each gene with an approved symbol</p></td>
</tr>
<tr class="row-even"><td><p>MANE</p></td>
<td><p>indicating if the transcript is the MANE Select or MANE Plus Clinical transcript for the gene.</p></td>
</tr>
<tr class="row-odd"><td><p>miRNA</p></td>
<td><p>Reports where the variant lies in the miRNA secondary structure.</p></td>
</tr>
</tbody>
</table>
</div>
</section>
</section>

        </article>
      </div>
      <footer>
        
        <div class="related-pages">
          <a class="next-page" href="balsamic_pon.html">
              <div class="page-info">
                <div class="context">
                  <span>Next</span>
                </div>
                <div class="title">Panel of Normals (PON)</div>
              </div>
              <svg class="furo-related-icon"><use href="#svg-arrow-right"></use></svg>
            </a>
          <a class="prev-page" href="balsamic_sv_cnv.html">
              <svg class="furo-related-icon"><use href="#svg-arrow-right"></use></svg>
              <div class="page-info">
                <div class="context">
                  <span>Previous</span>
                </div>
                
                <div class="title">Structural and Copy Number variants</div>
                
              </div>
            </a>
        </div>
        <div class="bottom-of-page">
          <div class="left-details">
            <div class="copyright">
                Copyright &#169; 2023, Clinical Genomics
            </div>
            Made with <a href="https://www.sphinx-doc.org/">Sphinx</a> and <a class="muted-link" href="https://pradyunsg.me">@pradyunsg</a>'s
            
            <a href="https://github.com/pradyunsg/furo">Furo</a>
            
          </div>
          <div class="right-details">
            
          </div>
        </div>
        
      </footer>
    </div>
    <aside class="toc-drawer">
      
      
      <div class="toc-sticky toc-scroll">
        <div class="toc-title-container">
          <span class="toc-title">
            On this page
          </span>
        </div>
        <div class="toc-tree-container">
          <div class="toc-tree">
            <ul>
<li><a class="reference internal" href="#">Annotation resources</a><ul>
<li><a class="reference internal" href="#gnomad">gnomAD</a></li>
<li><a class="reference internal" href="#clinvar">ClinVar</a></li>
<li><a class="reference internal" href="#cosmic">COSMIC</a></li>
<li><a class="reference internal" href="#cadd">CADD</a></li>
<li><a class="reference internal" href="#loqusdb-somatic-frequencies-cancer-cases">LoqusDB somatic frequencies (cancer cases)</a></li>
<li><a class="reference internal" href="#loqusdb-germline-frequencies-cancer-cases">LoqusDB germline frequencies (cancer cases)</a></li>
<li><a class="reference internal" href="#loqusdb-germline-frequencies-non-cancer-cases">LoqusDB germline frequencies (non-cancer cases)</a></li>
<li><a class="reference internal" href="#swegen">SWEGEN</a></li>
<li><a class="reference internal" href="#ensembl-vep-annotations">ENSEMBL-VEP annotations</a></li>
</ul>
</li>
</ul>

          </div>
        </div>
      </div>
      
      
    </aside>
  </div>
</div><script data-url_root="./" id="documentation_options" src="_static/documentation_options.js"></script>
    <script src="_static/doctools.js"></script>
    <script src="_static/sphinx_highlight.js"></script>
    <script src="_static/scripts/furo.js"></script>
    </body>
</html>