<!doctype html>
<html class="no-js" lang="en">
  <head><meta charset="utf-8"/>
    <meta name="viewport" content="width=device-width,initial-scale=1"/>
    <meta name="color-scheme" content="light dark"><meta name="generator" content="Docutils 0.18.1: http://docutils.sourceforge.net/" />
<link rel="index" title="Index" href="genindex.html" /><link rel="search" title="Search" href="search.html" /><link rel="next" title="Tools and software" href="bioinfo_softwares.html" /><link rel="prev" title="Method description" href="balsamic_methods.html" />

    <!-- Generated with Sphinx 6.2.1 and Furo 2023.07.26 -->
        <title>Changelog - BALSAMIC documentation</title>
      <link rel="stylesheet" type="text/css" href="_static/pygments.css" />
    <link rel="stylesheet" type="text/css" href="_static/styles/furo.css?digest=369552022d0b975c8e74270ce6eabe0fb7978f24" />
    <link rel="stylesheet" type="text/css" href="_static/styles/furo-extensions.css?digest=30d1aed668e5c3a91c3e3bf6a60b675221979f0e" />
    
    


<style>
  body {
    --color-code-background: #f8f8f8;
  --color-code-foreground: black;
  
  }
  @media not print {
    body[data-theme="dark"] {
      --color-code-background: #202020;
  --color-code-foreground: #d0d0d0;
  
    }
    @media (prefers-color-scheme: dark) {
      body:not([data-theme="light"]) {
        --color-code-background: #202020;
  --color-code-foreground: #d0d0d0;
  
      }
    }
  }
</style></head>
  <body>
    
    <script>
      document.body.dataset.theme = localStorage.getItem("theme") || "auto";
    </script>
    

<svg xmlns="http://www.w3.org/2000/svg" style="display: none;">
  <symbol id="svg-toc" viewBox="0 0 24 24">
    <title>Contents</title>
    <svg stroke="currentColor" fill="currentColor" stroke-width="0" viewBox="0 0 1024 1024">
      <path d="M408 442h480c4.4 0 8-3.6 8-8v-56c0-4.4-3.6-8-8-8H408c-4.4 0-8 3.6-8 8v56c0 4.4 3.6 8 8 8zm-8 204c0 4.4 3.6 8 8 8h480c4.4 0 8-3.6 8-8v-56c0-4.4-3.6-8-8-8H408c-4.4 0-8 3.6-8 8v56zm504-486H120c-4.4 0-8 3.6-8 8v56c0 4.4 3.6 8 8 8h784c4.4 0 8-3.6 8-8v-56c0-4.4-3.6-8-8-8zm0 632H120c-4.4 0-8 3.6-8 8v56c0 4.4 3.6 8 8 8h784c4.4 0 8-3.6 8-8v-56c0-4.4-3.6-8-8-8zM115.4 518.9L271.7 642c5.8 4.6 14.4.5 14.4-6.9V388.9c0-7.4-8.5-11.5-14.4-6.9L115.4 505.1a8.74 8.74 0 0 0 0 13.8z"/>
    </svg>
  </symbol>
  <symbol id="svg-menu" viewBox="0 0 24 24">
    <title>Menu</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="2" stroke-linecap="round" stroke-linejoin="round" class="feather-menu">
      <line x1="3" y1="12" x2="21" y2="12"></line>
      <line x1="3" y1="6" x2="21" y2="6"></line>
      <line x1="3" y1="18" x2="21" y2="18"></line>
    </svg>
  </symbol>
  <symbol id="svg-arrow-right" viewBox="0 0 24 24">
    <title>Expand</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="2" stroke-linecap="round" stroke-linejoin="round" class="feather-chevron-right">
      <polyline points="9 18 15 12 9 6"></polyline>
    </svg>
  </symbol>
  <symbol id="svg-sun" viewBox="0 0 24 24">
    <title>Light mode</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="1.5" stroke-linecap="round" stroke-linejoin="round" class="feather-sun">
      <circle cx="12" cy="12" r="5"></circle>
      <line x1="12" y1="1" x2="12" y2="3"></line>
      <line x1="12" y1="21" x2="12" y2="23"></line>
      <line x1="4.22" y1="4.22" x2="5.64" y2="5.64"></line>
      <line x1="18.36" y1="18.36" x2="19.78" y2="19.78"></line>
      <line x1="1" y1="12" x2="3" y2="12"></line>
      <line x1="21" y1="12" x2="23" y2="12"></line>
      <line x1="4.22" y1="19.78" x2="5.64" y2="18.36"></line>
      <line x1="18.36" y1="5.64" x2="19.78" y2="4.22"></line>
    </svg>
  </symbol>
  <symbol id="svg-moon" viewBox="0 0 24 24">
    <title>Dark mode</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="1.5" stroke-linecap="round" stroke-linejoin="round" class="icon-tabler-moon">
      <path stroke="none" d="M0 0h24v24H0z" fill="none" />
      <path d="M12 3c.132 0 .263 0 .393 0a7.5 7.5 0 0 0 7.92 12.446a9 9 0 1 1 -8.313 -12.454z" />
    </svg>
  </symbol>
  <symbol id="svg-sun-half" viewBox="0 0 24 24">
    <title>Auto light/dark mode</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="1.5" stroke-linecap="round" stroke-linejoin="round" class="icon-tabler-shadow">
      <path stroke="none" d="M0 0h24v24H0z" fill="none"/>
      <circle cx="12" cy="12" r="9" />
      <path d="M13 12h5" />
      <path d="M13 15h4" />
      <path d="M13 18h1" />
      <path d="M13 9h4" />
      <path d="M13 6h1" />
    </svg>
  </symbol>
</svg>

<input type="checkbox" class="sidebar-toggle" name="__navigation" id="__navigation">
<input type="checkbox" class="sidebar-toggle" name="__toc" id="__toc">
<label class="overlay sidebar-overlay" for="__navigation">
  <div class="visually-hidden">Hide navigation sidebar</div>
</label>
<label class="overlay toc-overlay" for="__toc">
  <div class="visually-hidden">Hide table of contents sidebar</div>
</label>



<div class="page">
  <header class="mobile-header">
    <div class="header-left">
      <label class="nav-overlay-icon" for="__navigation">
        <div class="visually-hidden">Toggle site navigation sidebar</div>
        <i class="icon"><svg><use href="#svg-menu"></use></svg></i>
      </label>
    </div>
    <div class="header-center">
      <a href="index.html"><div class="brand">BALSAMIC  documentation</div></a>
    </div>
    <div class="header-right">
      <div class="theme-toggle-container theme-toggle-header">
        <button class="theme-toggle">
          <div class="visually-hidden">Toggle Light / Dark / Auto color theme</div>
          <svg class="theme-icon-when-auto"><use href="#svg-sun-half"></use></svg>
          <svg class="theme-icon-when-dark"><use href="#svg-moon"></use></svg>
          <svg class="theme-icon-when-light"><use href="#svg-sun"></use></svg>
        </button>
      </div>
      <label class="toc-overlay-icon toc-header-icon" for="__toc">
        <div class="visually-hidden">Toggle table of contents sidebar</div>
        <i class="icon"><svg><use href="#svg-toc"></use></svg></i>
      </label>
    </div>
  </header>
  <aside class="sidebar-drawer">
    <div class="sidebar-container">
      
      <div class="sidebar-sticky"><a class="sidebar-brand" href="index.html">
  
  <div class="sidebar-logo-container">
    <img class="sidebar-logo" src="_static/balsamic_logo.png" alt="Logo"/>
  </div>
  
  <span class="sidebar-brand-text">BALSAMIC  documentation</span>
  
</a><form class="sidebar-search-container" method="get" action="search.html" role="search">
  <input class="sidebar-search" placeholder="Search" name="q" aria-label="Search">
  <input type="hidden" name="check_keywords" value="yes">
  <input type="hidden" name="area" value="default">
</form>
<div id="searchbox"></div><div class="sidebar-scroll"><div class="sidebar-tree">
  <p class="caption" role="heading"><span class="caption-text">Installation and Usage</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="install.html">Installation</a></li>
<li class="toctree-l1"><a class="reference internal" href="user_guide.html">Short tutorial</a></li>
<li class="toctree-l1"><a class="reference internal" href="cli_package.html">Command line arguments</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">Analysis and Interpretation</span></p>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="balsamic_filters.html">Calling and filtering variants</a></li>
<li class="toctree-l1"><a class="reference internal" href="balsamic_sv_cnv.html">Structural and Copy Number variants</a></li>
<li class="toctree-l1"><a class="reference internal" href="balsamic_annotation.html">Annotation resources</a></li>
<li class="toctree-l1"><a class="reference internal" href="balsamic_pon.html">Panel of Normals (PON)</a></li>
<li class="toctree-l1"><a class="reference internal" href="balsamic_pon.html#cnvkit-pon">CNVkit PON</a></li>
<li class="toctree-l1"><a class="reference internal" href="balsamic_pon.html#gens-pon">GENS PON</a></li>
<li class="toctree-l1"><a class="reference internal" href="balsamic_methods.html">Method description</a></li>
<li class="toctree-l1 current current-page"><a class="current reference internal" href="#">Changelog</a></li>
<li class="toctree-l1"><a class="reference internal" href="bioinfo_softwares.html">Tools and software</a></li>
<li class="toctree-l1"><a class="reference internal" href="resources.html">References and other resources</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">Development guide</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="README.html">Documentation Guidline</a></li>
<li class="toctree-l1"><a class="reference internal" href="programming_guide.html">Coding etiquettes</a></li>
<li class="toctree-l1"><a class="reference internal" href="semver.html">Semantic versioning</a></li>
<li class="toctree-l1"><a class="reference internal" href="FAQs.html">Frequently Asked Questions (FAQs)</a></li>
</ul>

</div>
</div>

      </div>
      
    </div>
  </aside>
  <div class="main">
    <div class="content">
      <div class="article-container">
        <a href="#" class="back-to-top muted-link">
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24">
            <path d="M13 20h-2V8l-5.5 5.5-1.42-1.42L12 4.16l7.92 7.92-1.42 1.42L13 8v12z"></path>
          </svg>
          <span>Back to top</span>
        </a>
        <div class="content-icon-container">
          
<div class="theme-toggle-container theme-toggle-content">
            <button class="theme-toggle">
              <div class="visually-hidden">Toggle Light / Dark / Auto color theme</div>
              <svg class="theme-icon-when-auto"><use href="#svg-sun-half"></use></svg>
              <svg class="theme-icon-when-dark"><use href="#svg-moon"></use></svg>
              <svg class="theme-icon-when-light"><use href="#svg-sun"></use></svg>
            </button>
          </div>
          <label class="toc-overlay-icon toc-content-icon" for="__toc">
            <div class="visually-hidden">Toggle table of contents sidebar</div>
            <i class="icon"><svg><use href="#svg-toc"></use></svg></i>
          </label>
        </div>
        <article role="main">
          <section id="changelog">
<h1>Changelog<a class="headerlink" href="#changelog" title="Permalink to this heading">#</a></h1>
<section id="id1">
<h2>[13.0.0]<a class="headerlink" href="#id1" title="Permalink to this heading">#</a></h2>
<section id="added">
<h3>Added:<a class="headerlink" href="#added" title="Permalink to this heading">#</a></h3>
<ul class="simple">
<li><p>Fastq concatenation <a class="reference external" href="https://github.com/Clinical-Genomics/BALSAMIC/pull/1069">https://github.com/Clinical-Genomics/BALSAMIC/pull/1069</a></p></li>
<li><p><cite>CADD</cite> SNV references <a class="reference external" href="https://github.com/Clinical-Genomics/BALSAMIC/pull/1126">https://github.com/Clinical-Genomics/BALSAMIC/pull/1126</a></p></li>
<li><p><cite>CADD</cite> SNV annotation <a class="reference external" href="https://github.com/Clinical-Genomics/BALSAMIC/pull/1150">https://github.com/Clinical-Genomics/BALSAMIC/pull/1150</a></p></li>
<li><p>Samtools <cite>stats</cite>, <cite>flagstat</cite>, <cite>idxstat</cite> to WGS workflow <a class="reference external" href="https://github.com/Clinical-Genomics/BALSAMIC/pull/1176">https://github.com/Clinical-Genomics/BALSAMIC/pull/1176</a></p></li>
<li><p>Functionality for dynamically assigning fastq-info to sample dict in config from input fastq-dir <a class="reference external" href="https://github.com/Clinical-Genomics/BALSAMIC/pull/1176">https://github.com/Clinical-Genomics/BALSAMIC/pull/1176</a></p></li>
<li><p>Annotate SNVs with cancer germline SNV observations from Loqusdb <a class="reference external" href="https://github.com/Clinical-Genomics/BALSAMIC/pull/1178">https://github.com/Clinical-Genomics/BALSAMIC/pull/1178</a></p></li>
<li><p>Annotate SNVs with somatic SNV observations from Loqusdb <a class="reference external" href="https://github.com/Clinical-Genomics/BALSAMIC/pull/1187">https://github.com/Clinical-Genomics/BALSAMIC/pull/1187</a></p></li>
<li><p>Tests for Annotation with Cancer germline, somatic and clinical observations, and swegen frequencies <a class="reference external" href="https://github/Clinical-Genomics/BALSAMIC/pull/1190">https://github/Clinical-Genomics/BALSAMIC/pull/1190</a></p></li>
<li><p>Annotate SVs with somatic SV observations from Loqusdb <a class="reference external" href="https://github.com/Clinical-Genomics/BALSAMIC/pull/1194">https://github.com/Clinical-Genomics/BALSAMIC/pull/1194</a></p></li>
<li><p>Support singularity bind paths with different destination directories <a class="reference external" href="https://github/Clinical-Genomics/BALSAMIC/pull/1211">https://github/Clinical-Genomics/BALSAMIC/pull/1211</a></p></li>
<li><p>Added <cite>–rerun-trigger mtime</cite> option to Snakemake command <a class="reference external" href="https://github.com/Clinical-Genomics/BALSAMIC/pull/1217">https://github.com/Clinical-Genomics/BALSAMIC/pull/1217</a></p></li>
<li><p><cite>CADD</cite> container <a class="reference external" href="https://github.com/Clinical-Genomics/BALSAMIC/pull/1222">https://github.com/Clinical-Genomics/BALSAMIC/pull/1222</a></p></li>
<li><p>Container ettiquette to ReadtheDocs <a class="reference external" href="https://github.com/Clinical-Genomics/BALSAMIC/pull/1232">https://github.com/Clinical-Genomics/BALSAMIC/pull/1232</a></p></li>
<li><p><cite>htslib</cite> (samtools, bcftools tabix) container <a class="reference external" href="https://github.com/Clinical-Genomics/BALSAMIC/pull/1234">https://github.com/Clinical-Genomics/BALSAMIC/pull/1234</a></p></li>
<li><p>Release version support for cache generation <a class="reference external" href="https://github.com/Clinical-Genomics/BALSAMIC/pull/1231">https://github.com/Clinical-Genomics/BALSAMIC/pull/1231</a></p></li>
<li><p><cite>CADD</cite> scores for INDELs <a class="reference external" href="https://github.com/Clinical-Genomics/BALSAMIC/pull/1238">https://github.com/Clinical-Genomics/BALSAMIC/pull/1238</a></p></li>
<li><p><cite>CADD</cite> reference to tests <a class="reference external" href="https://githuc.com/Clinical-Genomics/BALSAMIC/pull/1241">https://githuc.com/Clinical-Genomics/BALSAMIC/pull/1241</a></p></li>
<li><p>Add cache version option to config case <a class="reference external" href="https://github.com/Clinical-Genomics/BALSAMIC/pull/1244">https://github.com/Clinical-Genomics/BALSAMIC/pull/1244</a></p></li>
<li><p><cite>cnvkit</cite> container <a class="reference external" href="https://github.com/Clinical-Genomics/BALSAMIC/pull/1252">https://github.com/Clinical-Genomics/BALSAMIC/pull/1252</a></p></li>
<li><p><cite>PureCN</cite> container <a class="reference external" href="https://github.com/Clinical-Genomics/BALSAMIC/pull/1255">https://github.com/Clinical-Genomics/BALSAMIC/pull/1255</a></p></li>
<li><p><cite>GATK</cite> container <a class="reference external" href="https://github.com/Clinical-Genomics/BALSAMIC/pull/1266">https://github.com/Clinical-Genomics/BALSAMIC/pull/1266</a></p></li>
<li><p>Resolved FASTQ paths to sample dictionary (balsamic logging) <a class="reference external" href="https://github.com/Clinical-Genomics/BALSAMIC/pull/1275">https://github.com/Clinical-Genomics/BALSAMIC/pull/1275</a></p></li>
<li><p>Picard HsMetrics and CollectGcBiasMetrics for WGS <a class="reference external" href="https://github.com/Clinical-Genomics/BALSAMIC/pull/1288">https://github.com/Clinical-Genomics/BALSAMIC/pull/1288</a></p></li>
<li><p><cite>LOH</cite> to TGA workflow <a class="reference external" href="https://Clinical-Genomics/BALSAMIC/pull/1278">https://Clinical-Genomics/BALSAMIC/pull/1278</a></p></li>
<li><p>CNVs from PureCN to TGA workflow <a class="reference external" href="https://Clinical-Genomics/BALSAMIC/pull/1278">https://Clinical-Genomics/BALSAMIC/pull/1278</a></p></li>
<li><p>Command-line arguments and rules for creation of GENS files <a class="reference external" href="https://github.com/Clinical-Genomics/BALSAMIC/pull/1279">https://github.com/Clinical-Genomics/BALSAMIC/pull/1279</a></p></li>
<li><p>Somatic and germline Loqusdb annotation to ReadtheDocs <a class="reference external" href="https://github.com/Clinical-Genomics/BALSAMIC/pull/1317">https://github.com/Clinical-Genomics/BALSAMIC/pull/1317</a></p></li>
</ul>
</section>
<section id="changed">
<h3>Changed:<a class="headerlink" href="#changed" title="Permalink to this heading">#</a></h3>
<ul class="simple">
<li><p>Changed CN header field in cnvpytor in cnvpytor_tumor_only to be Float instead of Integer <a class="reference external" href="https://github.com/Clinical-Genomics/BALSAMIC/pull/1182">https://github.com/Clinical-Genomics/BALSAMIC/pull/1182</a></p></li>
<li><p>Changed samples in case_config.json from being a dict to a list of dicts  <a class="reference external" href="https://github.com/Clinical-Genomics/BALSAMIC/pull/1176">https://github.com/Clinical-Genomics/BALSAMIC/pull/1176</a></p></li>
<li><p>Updated snakemake version to 7.25.0 <a class="reference external" href="https://github.com/Clinical-Genomics/BALSAMIC/pull/1099">https://github.com/Clinical-Genomics/BALSAMIC/pull/1099</a></p></li>
<li><p>Updated cryptography version to 41.0.1 <a class="reference external" href="https://github.com/Clinical-Genomics/BALSAMIC/pull/1173">https://github.com/Clinical-Genomics/BALSAMIC/pull/1173</a></p></li>
<li><p>Refactor bam and fastq inputs in snakemake to call pydantic model functions <a class="reference external" href="https://github.com/Clinical-Genomics/BALSAMIC/pull/1176">https://github.com/Clinical-Genomics/BALSAMIC/pull/1176</a></p></li>
<li><p>Standardised alignment workflows to WGS-workflow <a class="reference external" href="https://github.com/Clinical-Genomics/BALSAMIC/pull/1176">https://github.com/Clinical-Genomics/BALSAMIC/pull/1176</a></p></li>
<li><p>Implemented parallel trimming and alignment in all workflows per lane <a class="reference external" href="https://github.com/Clinical-Genomics/BALSAMIC/pull/1176">https://github.com/Clinical-Genomics/BALSAMIC/pull/1176</a></p></li>
<li><p>All bam-QC tools take the final dedup.realign bamfile as input <a class="reference external" href="https://github.com/Clinical-Genomics/BALSAMIC/pull/1176">https://github.com/Clinical-Genomics/BALSAMIC/pull/1176</a></p></li>
<li><p>Validation of pydantic models done both during config and run <a class="reference external" href="https://github.com/Clinical-Genomics/BALSAMIC/pull/1176">https://github.com/Clinical-Genomics/BALSAMIC/pull/1176</a></p></li>
<li><p>Refactored fastp rules, and changed order of UMI-trimming and quality trimming <a class="reference external" href="https://github.com/Clinical-Genomics/BALSAMIC/pull/1176">https://github.com/Clinical-Genomics/BALSAMIC/pull/1176</a></p></li>
<li><p>Fix pydantic version (&lt;2.0) <a class="reference external" href="https://github.com/Clinical-Genomics/BALSAMIC/pull/1191">https://github.com/Clinical-Genomics/BALSAMIC/pull/1191</a></p></li>
<li><p>Refactor constants <a class="reference external" href="https://github.com/Clinical-Genomics/BALSAMIC/pull/1174">https://github.com/Clinical-Genomics/BALSAMIC/pull/1174</a></p></li>
<li><p>Move models to their own folder <a class="reference external" href="https://github.com/Clinical-Genomics/BALSAMIC/pull/1176">https://github.com/Clinical-Genomics/BALSAMIC/pull/1176</a></p></li>
<li><p>Balsamic init workflow refactoring <a class="reference external" href="https://github.com/Clinical-Genomics/BALSAMIC/pull/1188">https://github.com/Clinical-Genomics/BALSAMIC/pull/1188</a></p></li>
<li><p>Updated cryptography version to 41.0.2 <a class="reference external" href="https://github.com/Clinical-Genomics/BALSAMIC/pull/1205">https://github.com/Clinical-Genomics/BALSAMIC/pull/1205</a></p></li>
<li><p>Refactor snakemake executable command generation <a class="reference external" href="https://github/Clinical-Genomics/BALSAMIC/pull/1211">https://github/Clinical-Genomics/BALSAMIC/pull/1211</a></p></li>
<li><p>Updated Python version to 3.11 and its dependencies <a class="reference external" href="https://github.com/Clinical-Genomics/BALSAMIC/pull/1216">https://github.com/Clinical-Genomics/BALSAMIC/pull/1216</a></p></li>
<li><p>Tools versions in doc <a class="reference external" href="https:/github.com/Clinical-Genomics/BALSAMIC/pull/1239">https:/github.com/Clinical-Genomics/BALSAMIC/pull/1239</a></p></li>
<li><p>Reuse common Balsamic CLI options <a class="reference external" href="https://github.com/Clinical-Genomics/BALSAMIC/pull/1242">https://github.com/Clinical-Genomics/BALSAMIC/pull/1242</a></p></li>
<li><p>Update <cite>reference.json</cite> file to use relative paths <a class="reference external" href="https://github.com/Clinical-Genomics/BALSAMIC/pull/1251">https://github.com/Clinical-Genomics/BALSAMIC/pull/1251</a></p></li>
<li><p>Update pydantic to v2 while maintaining support for v1 models <a class="reference external" href="https://github.com/Clinical-Genomics/BALSAMIC/pull/1253">https://github.com/Clinical-Genomics/BALSAMIC/pull/1253</a></p></li>
<li><p><cite>PCT_PF_READS_IMPROPER_PAIRS</cite> QC threshold lowered to 5% <a class="reference external" href="https://github.com/Clinical-Genomics/BALSAMIC/issues/1265">https://github.com/Clinical-Genomics/BALSAMIC/issues/1265</a></p></li>
<li><p>Migrate Metrics models to pydantic v2 <a class="reference external" href="https://github.com/Clinical-Genomics/BALSAMIC/pull/1270">https://github.com/Clinical-Genomics/BALSAMIC/pull/1270</a></p></li>
<li><p>Migrate Snakemake models to pydantic v2 <a class="reference external" href="https://github.com/Clinical-Genomics/BALSAMIC/pull/1268">https://github.com/Clinical-Genomics/BALSAMIC/pull/1268</a></p></li>
<li><p>Migrate Cache models to pydantic v2 <a class="reference external" href="https://github.com/Clinical-Genomics/BALSAMIC/pull/1274">https://github.com/Clinical-Genomics/BALSAMIC/pull/1274</a></p></li>
<li><p>Made BALSAMIC compatible with multiple PON creation workflows <a class="reference external" href="https://github.com/Clinical-Genomics/BALSAMIC/pull/1279">https://github.com/Clinical-Genomics/BALSAMIC/pull/1279</a></p></li>
<li><p>Use StrEnum from python enum <a class="reference external" href="https://github.com/Clinical-Genomics/BALSAMIC/pull/1303">https://github.com/Clinical-Genomics/BALSAMIC/pull/1303</a></p></li>
<li><p>Renamed final cram bamfile to format <cite>&lt;tumor/normal&gt;.&lt;LIMS_ID&gt;.cram</cite> <a class="reference external" href="https://github.com/Clinical-Genomics/BALSAMIC/pull/1307">https://github.com/Clinical-Genomics/BALSAMIC/pull/1307</a></p></li>
<li><p>Updated snakemake version to 7.32.4 <a class="reference external" href="https://github.com/Clinical-Genomics/BALSAMIC/pull/1308">https://github.com/Clinical-Genomics/BALSAMIC/pull/1308</a></p></li>
<li><p>Migrate analysis models to pydantic v2 <a class="reference external" href="https://github.com/Clinical-Genomics/BALSAMIC/pull/1306">https://github.com/Clinical-Genomics/BALSAMIC/pull/1306</a></p></li>
<li><p>Split analysis model into config and params models <a class="reference external" href="https://github.com/Clinical-Genomics/BALSAMIC/pull/1306">https://github.com/Clinical-Genomics/BALSAMIC/pull/1306</a></p></li>
<li><p>Renamed name in sample column of final clincial vcfs <a class="reference external" href="https://github.com/Clinical-Genomics/BALSAMIC/pull/1310">https://github.com/Clinical-Genomics/BALSAMIC/pull/1310</a></p></li>
<li><p>Update Gens HK tags <a class="reference external" href="https://github.com/Clinical-Genomics/BALSAMIC/pull/1319">https://github.com/Clinical-Genomics/BALSAMIC/pull/1319</a></p></li>
</ul>
</section>
<section id="fixed">
<h3>Fixed:<a class="headerlink" href="#fixed" title="Permalink to this heading">#</a></h3>
<ul class="simple">
<li><p>vcf2cytosure container <a class="reference external" href="https://github.com/Clinical-Genomics/BALSAMIC/pull/1159">https://github.com/Clinical-Genomics/BALSAMIC/pull/1159</a></p></li>
<li><p>Link external fastqs to case folder &amp; create case directory <a class="reference external" href="https://github.com/Clinical-Genomics/BALSAMIC/pull/1195">https://github.com/Clinical-Genomics/BALSAMIC/pull/1195</a></p></li>
<li><p>vcf2cytosure container missing constants <a class="reference external" href="https://github.com/Clinical-Genomics/BALSAMIC/pull/1198">https://github.com/Clinical-Genomics/BALSAMIC/pull/1198</a></p></li>
<li><p>Bash commands in vep_somatic_clinical_snv <a class="reference external" href="https://github.com/Clinical-Genomics/BALSAMIC/pull/1200">https://github.com/Clinical-Genomics/BALSAMIC/pull/1200</a></p></li>
<li><p>Fix SVDB annotation intermediate rule <a class="reference external" href="https://github.com/Clinical-Genomics/BALSAMIC/pull/1218">https://github.com/Clinical-Genomics/BALSAMIC/pull/1218</a></p></li>
<li><p>Broken documentation links <a class="reference external" href="https://github.com/Clinical-Genomics/BALSAMIC/pull/1226">https://github.com/Clinical-Genomics/BALSAMIC/pull/1226</a></p></li>
<li><p>Updated contributors in main README <a class="reference external" href="https://github.com/Clinical-Genomics/BALSAMIC/pull/1237">https://github.com/Clinical-Genomics/BALSAMIC/pull/1237</a></p></li>
<li><p>CNVpytor container <a class="reference external" href="https://github.com/Clinical-Genomics/BALSAMIC/pull/1246">https://github.com/Clinical-Genomics/BALSAMIC/pull/1246</a></p></li>
<li><p>Restored balsamic container in UMI concatenation rule <a class="reference external" href="https://github.com/Clinical-Genomics/BALSAMIC/pull/1261">https://github.com/Clinical-Genomics/BALSAMIC/pull/1261</a></p></li>
<li><p>CNVpytor container, fixing numpy version <a class="reference external" href="https://github.com/Clinical-Genomics/BALSAMIC/pull/1273">https://github.com/Clinical-Genomics/BALSAMIC/pull/1273</a></p></li>
<li><p>QC workflow store <a class="reference external" href="https://github.com/Clinical-Genomics/BALSAMIC/pull/1295">https://github.com/Clinical-Genomics/BALSAMIC/pull/1295</a></p></li>
<li><p>MultiQC rule missing input files <a class="reference external" href="https://github.com/Clinical-Genomics/BALSAMIC/pull/1321">https://github.com/Clinical-Genomics/BALSAMIC/pull/1321</a></p></li>
<li><p><cite>gens_preprocessing</cite> rule missing python directive <a class="reference external" href="https://github.com/Clinical-Genomics/BALSAMIC/pull/1322">https://github.com/Clinical-Genomics/BALSAMIC/pull/1322</a></p></li>
<li><p>CADD annotations container path and code smells <a class="reference external" href="https://github.com/Clinical-Genomics/BALSAMIC/pull/1323">https://github.com/Clinical-Genomics/BALSAMIC/pull/1323</a></p></li>
</ul>
</section>
<section id="removed">
<h3>Removed:<a class="headerlink" href="#removed" title="Permalink to this heading">#</a></h3>
<ul class="simple">
<li><p>Config folder <a class="reference external" href="https://github.com/Clinical-Genomics/BALSAMIC/pull/1175">https://github.com/Clinical-Genomics/BALSAMIC/pull/1175</a></p></li>
<li><p>Quality trimming of fastqs for UMI workflow <a class="reference external" href="https://github.com/Clinical-Genomics/BALSAMIC/pull/1176">https://github.com/Clinical-Genomics/BALSAMIC/pull/1176</a></p></li>
<li><p>Balsamic container <a class="reference external" href="https://github.com/Clinical-Genomics/BALSAMIC/pull/1230">https://github.com/Clinical-Genomics/BALSAMIC/pull/1230</a></p></li>
<li><p>Plugin CLI <a class="reference external" href="https://github.com/Clinical-Genomics/BALSAMIC/pull/1245">https://github.com/Clinical-Genomics/BALSAMIC/pull/1245</a></p></li>
<li><p>Realignment step for TGA workflow <a class="reference external" href="https://github.com/Clinical-Genomics/BALSAMIC/pull/1272">https://github.com/Clinical-Genomics/BALSAMIC/pull/1272</a></p></li>
<li><p>Archived/outdated workflows and scripts <a class="reference external" href="https://github.com/Clinical-Genomics/BALSAMIC/pull/1296">https://github.com/Clinical-Genomics/BALSAMIC/pull/1296</a></p></li>
<li><p>Sed command to convert CNVpytor integer to float, deprecated by updated CNVpytor version <a class="reference external" href="https://github.com/Clinical-Genomics/BALSAMIC/pull/1310">https://github.com/Clinical-Genomics/BALSAMIC/pull/1310</a></p></li>
</ul>
</section>
</section>
<section id="id2">
<h2>[12.0.2]<a class="headerlink" href="#id2" title="Permalink to this heading">#</a></h2>
<section id="id3">
<h3>Fixed:<a class="headerlink" href="#id3" title="Permalink to this heading">#</a></h3>
<ul class="simple">
<li><p>Missing <cite>Number</cite> in VCF header for SVs <a class="reference external" href="https://github.com/Clinical-Genomics/BALSAMIC/pull/1203">https://github.com/Clinical-Genomics/BALSAMIC/pull/1203</a></p></li>
</ul>
</section>
<section id="id4">
<h3>Changed:<a class="headerlink" href="#id4" title="Permalink to this heading">#</a></h3>
<ul class="simple">
<li><p>Fix cyvcf2 to version 0.30.22 <a class="reference external" href="https://github.com/Clinical-Genomics/BALSAMIC/pull/1206">https://github.com/Clinical-Genomics/BALSAMIC/pull/1206</a></p></li>
<li><p>Fix pydantic version (&lt;2.0) <a class="reference external" href="https://github.com/Clinical-Genomics/BALSAMIC/pull/1206">https://github.com/Clinical-Genomics/BALSAMIC/pull/1206</a></p></li>
<li><p>Update varcall-cnvkit container versions <a class="reference external" href="https://github.com/Clinical-Genomics/BALSAMIC/pull/1207">https://github.com/Clinical-Genomics/BALSAMIC/pull/1207</a></p></li>
</ul>
</section>
</section>
<section id="id5">
<h2>[12.0.1]<a class="headerlink" href="#id5" title="Permalink to this heading">#</a></h2>
<section id="id6">
<h3>Added:<a class="headerlink" href="#id6" title="Permalink to this heading">#</a></h3>
<ul class="simple">
<li><p>WGS QC criteria for <cite>PCT_PF_READS_IMPROPER_PAIRS</cite> (condition: &lt;= 0.1) <a class="reference external" href="https://github.com/Clinical-Genomics/BALSAMIC/pull/1164">https://github.com/Clinical-Genomics/BALSAMIC/pull/1164</a></p></li>
</ul>
</section>
<section id="id7">
<h3>Fixed:<a class="headerlink" href="#id7" title="Permalink to this heading">#</a></h3>
<ul class="simple">
<li><p>Logged version of Delly (changing it to v1.0.3)  <a class="reference external" href="https://github.com/Clinical-Genomics/BALSAMIC/pull/1170">https://github.com/Clinical-Genomics/BALSAMIC/pull/1170</a></p></li>
</ul>
</section>
</section>
<section id="id8">
<h2>[12.0.0]<a class="headerlink" href="#id8" title="Permalink to this heading">#</a></h2>
<section id="id9">
<h3>Added:<a class="headerlink" href="#id9" title="Permalink to this heading">#</a></h3>
<ul class="simple">
<li><p>PIP specific missing tools to config <a class="reference external" href="https://github.com/Clinical-Genomics/BALSAMIC/pull/1096">https://github.com/Clinical-Genomics/BALSAMIC/pull/1096</a></p></li>
<li><p>Filtering script to remove normal variants from TIDDIT <a class="reference external" href="https://github.com/Clinical-Genomics/BALSAMIC/pull/1120">https://github.com/Clinical-Genomics/BALSAMIC/pull/1120</a></p></li>
<li><p>Store TMB files in HK <a class="reference external" href="https://github.com/Clinical-Genomics/BALSAMIC/pull/1144">https://github.com/Clinical-Genomics/BALSAMIC/pull/1144</a></p></li>
</ul>
</section>
<section id="id10">
<h3>Changed:<a class="headerlink" href="#id10" title="Permalink to this heading">#</a></h3>
<ul class="simple">
<li><p>Fixed all conda container dependencies <a class="reference external" href="https://github.com/Clinical-Genomics/BALSAMIC/pull/1096">https://github.com/Clinical-Genomics/BALSAMIC/pull/1096</a></p></li>
<li><p>Changed –max_sv_size in VEP params to the size of chr1 for hg19 <a class="reference external" href="https://github.com/Clinical-Genomics/BALSAMIC/pull/1124">https://github.com/Clinical-Genomics/BALSAMIC/pull/1124</a></p></li>
<li><p>Increased time-limit for sambamba_exon_depth and picard_markduplicates to 6 hours <a class="reference external" href="https://github.com/Clinical-Genomics/BALSAMIC/pull/1143">https://github.com/Clinical-Genomics/BALSAMIC/pull/1143</a></p></li>
<li><p>Update cosmicdb to v97 <a class="reference external" href="https://github.com/Clinical-Genomics/BALSAMIC/pull/1147">https://github.com/Clinical-Genomics/BALSAMIC/pull/1147</a></p></li>
<li><p>Updated read the docs with the changes relevant to mention <a class="reference external" href="https://github.com/Clinical-Genomics/BALSAMIC/pull/1153">https://github.com/Clinical-Genomics/BALSAMIC/pull/1153</a></p></li>
</ul>
</section>
<section id="id11">
<h3>Fixed:<a class="headerlink" href="#id11" title="Permalink to this heading">#</a></h3>
<ul class="simple">
<li><p>Update cryptography version (39.0.1) due to security alert <a class="reference external" href="https://github.com/Clinical-Genomics/BALSAMIC/pull/1087">https://github.com/Clinical-Genomics/BALSAMIC/pull/1087</a></p></li>
<li><p>Bump cryptography to v40.0.2 and gsutil to v5.23 <a class="reference external" href="https://github.com/Clinical-Genomics/BALSAMIC/pull/1154">https://github.com/Clinical-Genomics/BALSAMIC/pull/1154</a></p></li>
<li><p>Pytest file saved in balsamic directory <a class="reference external" href="https://github.com/Clinical-Genomics/BALSAMIC/pull/1093">https://github.com/Clinical-Genomics/BALSAMIC/pull/1093</a></p></li>
<li><p>Fix varcall_py3 container bcftools dependency error <a class="reference external" href="https://github.com/Clinical-Genomics/BALSAMIC/pull/1097">https://github.com/Clinical-Genomics/BALSAMIC/pull/1097</a></p></li>
<li><p>AscatNgs container <a class="reference external" href="https://github.com/Clinical-Genomics/BALSAMIC/pull/1155">https://github.com/Clinical-Genomics/BALSAMIC/pull/1155</a></p></li>
</ul>
</section>
</section>
<section id="id12">
<h2>[11.2.0]<a class="headerlink" href="#id12" title="Permalink to this heading">#</a></h2>
<section id="id13">
<h3>Fixed:<a class="headerlink" href="#id13" title="Permalink to this heading">#</a></h3>
<ul class="simple">
<li><p>Number of variants are increased with triallelic_site <a class="reference external" href="https://github.com/Clinical-Genomics/BALSAMIC/pull/1089">https://github.com/Clinical-Genomics/BALSAMIC/pull/1089</a></p></li>
</ul>
</section>
</section>
<section id="id14">
<h2>[11.1.0]<a class="headerlink" href="#id14" title="Permalink to this heading">#</a></h2>
<section id="id15">
<h3>Added:<a class="headerlink" href="#id15" title="Permalink to this heading">#</a></h3>
<ul class="simple">
<li><p>Added somalier integration and relatedness check: <a class="reference external" href="https://github.com/Clinical-Genomics/BALSAMIC/pull/1017">https://github.com/Clinical-Genomics/BALSAMIC/pull/1017</a></p></li>
<li><p>Cluster resources for CNVPytor tumor only <a class="reference external" href="https://github.com/Clinical-Genomics/BALSAMIC/pull/1083">https://github.com/Clinical-Genomics/BALSAMIC/pull/1083</a></p></li>
</ul>
</section>
<section id="id16">
<h3>Changed:<a class="headerlink" href="#id16" title="Permalink to this heading">#</a></h3>
<ul class="simple">
<li><p>Parallelize download of reference files <a class="reference external" href="https://github.com/Clinical-Genomics/BALSAMIC/pull/1065">https://github.com/Clinical-Genomics/BALSAMIC/pull/1065</a></p></li>
<li><p>Parallelize download of container images <a class="reference external" href="https://github.com/Clinical-Genomics/BALSAMIC/pull/1068">https://github.com/Clinical-Genomics/BALSAMIC/pull/1068</a></p></li>
</ul>
</section>
<section id="id17">
<h3>Fixed:<a class="headerlink" href="#id17" title="Permalink to this heading">#</a></h3>
<ul class="simple">
<li><p>triallelic_site in quality filter for SNV <a class="reference external" href="https://github.com/Clinical-Genomics/BALSAMIC/pull/1052">https://github.com/Clinical-Genomics/BALSAMIC/pull/1052</a></p></li>
<li><p>Compression of SNV, research and clinical, VCF files <a class="reference external" href="https://github.com/Clinical-Genomics/BALSAMIC/pull/1060">https://github.com/Clinical-Genomics/BALSAMIC/pull/1060</a></p></li>
<li><p><cite>test_write_json</cite> failing locally <a class="reference external" href="https://github.com/Clinical-Genomics/BALSAMIC/pull/1063">https://github.com/Clinical-Genomics/BALSAMIC/pull/1063</a></p></li>
<li><p>Container build and push via github actions by setting buildx <cite>provenance</cite> flag to false <a class="reference external" href="https://github.com/Clinical-Genomics/BALSAMIC/pull/1071">https://github.com/Clinical-Genomics/BALSAMIC/pull/1071</a></p></li>
<li><p>Added buildx to the submodule workflow <a class="reference external" href="https://github.com/Clinical-Genomics/BALSAMIC/pull/1072">https://github.com/Clinical-Genomics/BALSAMIC/pull/1072</a></p></li>
<li><p>Change user in somalier container to defaultuser <a class="reference external" href="https://github.com/Clinical-Genomics/BALSAMIC/pull/1080">https://github.com/Clinical-Genomics/BALSAMIC/pull/1080</a></p></li>
<li><p>Reference files for hg38 <a class="reference external" href="https://github.com/Clinical-Genomics/BALSAMIC/pull/1081">https://github.com/Clinical-Genomics/BALSAMIC/pull/1081</a></p></li>
</ul>
</section>
</section>
<section id="id18">
<h2>[11.0.2]<a class="headerlink" href="#id18" title="Permalink to this heading">#</a></h2>
<section id="id19">
<h3>Changed:<a class="headerlink" href="#id19" title="Permalink to this heading">#</a></h3>
<ul class="simple">
<li><p>Code owners <a class="reference external" href="https://github.com/Clinical-Genomics/BALSAMIC/pull/1050">https://github.com/Clinical-Genomics/BALSAMIC/pull/1050</a></p></li>
</ul>
</section>
<section id="id20">
<h3>Fixed:<a class="headerlink" href="#id20" title="Permalink to this heading">#</a></h3>
<ul class="simple">
<li><p>MaxDepth in quality filter for SV <a class="reference external" href="https://github.com/Clinical-Genomics/BALSAMIC/pull/1051">https://github.com/Clinical-Genomics/BALSAMIC/pull/1051</a></p></li>
</ul>
</section>
</section>
<section id="id21">
<h2>[11.0.1]<a class="headerlink" href="#id21" title="Permalink to this heading">#</a></h2>
<section id="id22">
<h3>Fixed:<a class="headerlink" href="#id22" title="Permalink to this heading">#</a></h3>
<ul class="simple">
<li><p>Incorrect raw <cite>TNscope</cite> VCF delivered <a class="reference external" href="https://github.com/Clinical-Genomics/BALSAMIC/pull/1042">https://github.com/Clinical-Genomics/BALSAMIC/pull/1042</a></p></li>
</ul>
</section>
</section>
<section id="id23">
<h2>[11.0.0]<a class="headerlink" href="#id23" title="Permalink to this heading">#</a></h2>
<section id="id24">
<h3>Added:<a class="headerlink" href="#id24" title="Permalink to this heading">#</a></h3>
<ul class="simple">
<li><p>Use of PON reference, if exists for CNVkit tumor-normal analysis <a class="reference external" href="https://github.com/Clinical-Genomics/BALSAMIC/pull/982">https://github.com/Clinical-Genomics/BALSAMIC/pull/982</a></p></li>
<li><p>Added PON version to CLI and config.json <a class="reference external" href="https://github.com/Clinical-Genomics/BALSAMIC/pull/983">https://github.com/Clinical-Genomics/BALSAMIC/pull/983</a></p></li>
<li><p><cite>cnvpytor</cite> to varcallpy3 container <a class="reference external" href="https://github.com/Clinical-Genomics/BALSAMIC/pull/991">https://github.com/Clinical-Genomics/BALSAMIC/pull/991</a></p></li>
<li><p><cite>cnvpytor</cite> for tumor only workflow <a class="reference external" href="https://github.com/Clinical-Genomics/BALSAMIC/pull/994">https://github.com/Clinical-Genomics/BALSAMIC/pull/994</a></p></li>
<li><p>R packages to cnvkit container <a class="reference external" href="https://github.com/Clinical-Genomics/BALSAMIC/pull/996">https://github.com/Clinical-Genomics/BALSAMIC/pull/996</a></p></li>
<li><p>Missing R packages to cnvkit container <a class="reference external" href="https://github.com/Clinical-Genomics/BALSAMIC/pull/997">https://github.com/Clinical-Genomics/BALSAMIC/pull/997</a></p></li>
<li><p>add rlang to cnvkit container <a class="reference external" href="https://github.com/Clinical-Genomics/BALSAMIC/pull/998">https://github.com/Clinical-Genomics/BALSAMIC/pull/998</a></p></li>
<li><p>AnnotSV and bedtools to annotate container <a class="reference external" href="https://github.com/Clinical-Genomics/BALSAMIC/pull/1005">https://github.com/Clinical-Genomics/BALSAMIC/pull/1005</a></p></li>
<li><p>cosmicdb to TNscope for tumor only and tumor normal workflows <a class="reference external" href="https://github.com/Clinical-Genomics/BALSAMIC/pull/1006">https://github.com/Clinical-Genomics/BALSAMIC/pull/1006</a></p></li>
<li><p><cite>loqusDB</cite> dump files to the config through the balsamic config case CLI <a class="reference external" href="https://github.com/Clinical-Genomics/BALSAMIC/pull/992">https://github.com/Clinical-Genomics/BALSAMIC/pull/992</a></p></li>
<li><p>Pre-annotation quality filters for SNVs annd added <cite>research</cite> to output files <a class="reference external" href="https://github.com/Clinical-Genomics/BALSAMIC/pull/1007">https://github.com/Clinical-Genomics/BALSAMIC/pull/1007</a></p></li>
<li><p>Annotation of snv_clinical_observations for somatic snv <a class="reference external" href="https://github.com/Clinical-Genomics/BALSAMIC/pull/1012">https://github.com/Clinical-Genomics/BALSAMIC/pull/1012</a></p></li>
<li><p>Annotation of sv_clinical_observations  for somatic sv and SV CNV filter rules <a class="reference external" href="https://github.com/Clinical-Genomics/BALSAMIC/pull/1013">https://github.com/Clinical-Genomics/BALSAMIC/pull/1013</a></p></li>
<li><p>Swegen SNV and SV frequency database for WGS <a class="reference external" href="https://github.com/Clinical-Genomics/BALSAMIC/pull/1014">https://github.com/Clinical-Genomics/BALSAMIC/pull/1014</a></p></li>
<li><p>triallelic_sites and variants with MaxDepth to the VCFs <a class="reference external" href="https://github.com/Clinical-Genomics/BALSAMIC/pull/1021">https://github.com/Clinical-Genomics/BALSAMIC/pull/1021</a></p></li>
<li><p>Clinical VCF for TGA workflow <a class="reference external" href="https://github.com/Clinical-Genomics/BALSAMIC/pull/1024">https://github.com/Clinical-Genomics/BALSAMIC/pull/1024</a></p></li>
<li><p>CNVpytor plots into the CNV PDF report <a class="reference external" href="https://github.com/Clinical-Genomics/BALSAMIC/pull/1023">https://github.com/Clinical-Genomics/BALSAMIC/pull/1023</a></p></li>
<li><p>Research and clinical housekeeper tags <a class="reference external" href="https://github.com/Clinical-Genomics/BALSAMIC/pull/1023">https://github.com/Clinical-Genomics/BALSAMIC/pull/1023</a></p></li>
<li><p>Cluster configuration for rules <a class="reference external" href="https://github.com/Clinical-Genomics/BALSAMIC/pull/1028">https://github.com/Clinical-Genomics/BALSAMIC/pull/1028</a></p></li>
<li><p>Variant filteration using loqusDB and Swegen annotations <a class="reference external" href="https://github.com/Clinical-Genomics/BALSAMIC/pull/1029">https://github.com/Clinical-Genomics/BALSAMIC/pull/1029</a></p></li>
<li><p>Annotation resources to readsthedocs <a class="reference external" href="https://github.com/Clinical-Genomics/BALSAMIC/pull/1031">https://github.com/Clinical-Genomics/BALSAMIC/pull/1031</a></p></li>
<li><p>Delly CNV rules for TGA workflow <a class="reference external" href="https://github.com/Clinical-Genomics/BALSAMIC/pull/103">https://github.com/Clinical-Genomics/BALSAMIC/pull/103</a></p></li>
<li><p>cnvpytor container and removed cnvpytor from varcallpy3 <a class="reference external" href="https://github.com/Clinical-Genomics/BALSAMIC/pull/1037">https://github.com/Clinical-Genomics/BALSAMIC/pull/1037</a></p></li>
</ul>
</section>
<section id="id25">
<h3>Changed:<a class="headerlink" href="#id25" title="Permalink to this heading">#</a></h3>
<ul class="simple">
<li><p>Added version number to the PON reference filename (<cite>.cnn</cite>) <a class="reference external" href="https://github.com/Clinical-Genomics/BALSAMIC/pull/982">https://github.com/Clinical-Genomics/BALSAMIC/pull/982</a></p></li>
<li><p>Update <cite>TIDDIT</cite> to v3.3.0, <cite>SVDB</cite> to v2.6.4, <cite>delly</cite> to v1.1.3, <cite>vcf2cytosure</cite> to v0.8 <a class="reference external" href="https://github.com/Clinical-Genomics/BALSAMIC/pull/987">https://github.com/Clinical-Genomics/BALSAMIC/pull/987</a></p></li>
<li><p>toml config file for vcfanno <a class="reference external" href="https://github.com/Clinical-Genomics/BALSAMIC/pull/1012">https://github.com/Clinical-Genomics/BALSAMIC/pull/1012</a></p></li>
<li><p>Split <cite>vep_germline</cite> rule into <cite>tumor</cite> and <cite>normal</cite> <a class="reference external" href="https://github.com/Clinical-Genomics/BALSAMIC/pull/1018">https://github.com/Clinical-Genomics/BALSAMIC/pull/1018</a></p></li>
<li><p>Extract number of variants from clinical files <a class="reference external" href="https://github.com/Clinical-Genomics/BALSAMIC/pull/1022">https://github.com/Clinical-Genomics/BALSAMIC/pull/1022</a></p></li>
</ul>
</section>
<section id="id26">
<h3>Fixed:<a class="headerlink" href="#id26" title="Permalink to this heading">#</a></h3>
<ul class="simple">
<li><p>Reverted <cite>pandas</cite> version (from <cite>1.3.5</cite> to <cite>1.1.5</cite>) <a class="reference external" href="https://github.com/Clinical-Genomics/BALSAMIC/pull/1018">https://github.com/Clinical-Genomics/BALSAMIC/pull/1018</a></p></li>
<li><p>Mate in realigned bam file <a class="reference external" href="https://github.com/Clinical-Genomics/BALSAMIC/pull/1019">https://github.com/Clinical-Genomics/BALSAMIC/pull/1019</a></p></li>
<li><p>samtools command in merge bam and names in toml for vcfanno <a class="reference external" href="https://github.com/Clinical-Genomics/BALSAMIC/pull/1020">https://github.com/Clinical-Genomics/BALSAMIC/pull/1020</a></p></li>
<li><p>If statement in <cite>vep_somatic_clinical_snv</cite> rule <a class="reference external" href="https://github.com/Clinical-Genomics/BALSAMIC/pull/1022">https://github.com/Clinical-Genomics/BALSAMIC/pull/1022</a></p></li>
<li><p>Invalid flag second of pair validation error <a class="reference external" href="https://github.com/Clinical-Genomics/BALSAMIC/pull/1025">https://github.com/Clinical-Genomics/BALSAMIC/pull/1025</a></p></li>
<li><p>Invalid flag second of pair validation error using picardtools <a class="reference external" href="https://github.com/Clinical-Genomics/BALSAMIC/pull/1027">https://github.com/Clinical-Genomics/BALSAMIC/pull/1027</a></p></li>
<li><p>Samtools command for mergetype tumor <a class="reference external" href="https://github.com/Clinical-Genomics/BALSAMIC/pull/1030">https://github.com/Clinical-Genomics/BALSAMIC/pull/1030</a></p></li>
<li><p><cite>varcall_py3</cite> container building <a class="reference external" href="https://github.com/Clinical-Genomics/BALSAMIC/pull/1036">https://github.com/Clinical-Genomics/BALSAMIC/pull/1036</a></p></li>
<li><p>Picard and fastp commands params and cluster config for umi workflow <a class="reference external" href="https://github.com/Clinical-Genomics/BALSAMIC/pull/1032">https://github.com/Clinical-Genomics/BALSAMIC/pull/1032</a></p></li>
<li><p>Set channels in <cite>varcall_py3</cite> container <a class="reference external" href="https://github.com/Clinical-Genomics/BALSAMIC/pull/1035">https://github.com/Clinical-Genomics/BALSAMIC/pull/1035</a></p></li>
<li><p>Delly command for tumor-normal analysis <a class="reference external" href="https://github.com/Clinical-Genomics/BALSAMIC/pull/1039">https://github.com/Clinical-Genomics/BALSAMIC/pull/1039</a></p></li>
<li><p>tabix command in bcftools_quality_filter_TNscope_umi_tumor_only rule <a class="reference external" href="https://github.com/Clinical-Genomics/BALSAMIC/pull/1040">https://github.com/Clinical-Genomics/BALSAMIC/pull/1040</a></p></li>
</ul>
</section>
<section id="id27">
<h3>Removed:<a class="headerlink" href="#id27" title="Permalink to this heading">#</a></h3>
<ul class="simple">
<li><p>case ID from the PON <cite>.cnn</cite> output file <a class="reference external" href="https://github.com/Clinical-Genomics/BALSAMIC/pull/983">https://github.com/Clinical-Genomics/BALSAMIC/pull/983</a></p></li>
<li><p><cite>TNhaplotyper</cite> for paired WGS analysis <a class="reference external" href="https://github.com/Clinical-Genomics/BALSAMIC/pull/988">https://github.com/Clinical-Genomics/BALSAMIC/pull/988</a></p></li>
<li><p><cite>TNhaplotyper</cite> for tumor only WGS analysis <a class="reference external" href="https://github.com/Clinical-Genomics/BALSAMIC/pull/1006">https://github.com/Clinical-Genomics/BALSAMIC/pull/1006</a></p></li>
<li><p><cite>TNhaplotyper</cite> for TGS <a class="reference external" href="https://github.com/Clinical-Genomics/BALSAMIC/pull/1022">https://github.com/Clinical-Genomics/BALSAMIC/pull/1022</a></p></li>
</ul>
</section>
</section>
<section id="id28">
<h2>[10.0.5]<a class="headerlink" href="#id28" title="Permalink to this heading">#</a></h2>
<section id="id29">
<h3>Changed:<a class="headerlink" href="#id29" title="Permalink to this heading">#</a></h3>
<ul class="simple">
<li><p>Update <cite>vcf2cytosure</cite> version to v0.8 <a class="reference external" href="https://github.com/Clinical-Genomics/BALSAMIC/pull/1010">https://github.com/Clinical-Genomics/BALSAMIC/pull/1010</a></p></li>
<li><p>Update GitHub action images to <cite>ubuntu-20.04</cite> <a class="reference external" href="https://github.com/Clinical-Genomics/BALSAMIC/pull/1010">https://github.com/Clinical-Genomics/BALSAMIC/pull/1010</a></p></li>
<li><p>Update GitHub actions to their latest versions <a class="reference external" href="https://github.com/Clinical-Genomics/BALSAMIC/pull/1010">https://github.com/Clinical-Genomics/BALSAMIC/pull/1010</a></p></li>
</ul>
</section>
</section>
<section id="id30">
<h2>[10.0.4]<a class="headerlink" href="#id30" title="Permalink to this heading">#</a></h2>
<section id="id31">
<h3>Fixed:<a class="headerlink" href="#id31" title="Permalink to this heading">#</a></h3>
<ul class="simple">
<li><p>Increase <cite>sambamba_exon_depth</cite> rule run time <a class="reference external" href="https://github.com/Clinical-Genomics/BALSAMIC/pull/1001">https://github.com/Clinical-Genomics/BALSAMIC/pull/1001</a></p></li>
</ul>
</section>
</section>
<section id="id32">
<h2>[10.0.3]<a class="headerlink" href="#id32" title="Permalink to this heading">#</a></h2>
<section id="id33">
<h3>Fixed:<a class="headerlink" href="#id33" title="Permalink to this heading">#</a></h3>
<ul class="simple">
<li><p>Input VCF files for cnvkit rules, cnvkit command and container <a class="reference external" href="https://github.com/Clinical-Genomics/BALSAMIC/pull/995">https://github.com/Clinical-Genomics/BALSAMIC/pull/995</a></p></li>
</ul>
</section>
</section>
<section id="id34">
<h2>[10.0.2]<a class="headerlink" href="#id34" title="Permalink to this heading">#</a></h2>
<section id="id35">
<h3>Fixed:<a class="headerlink" href="#id35" title="Permalink to this heading">#</a></h3>
<ul class="simple">
<li><p>TIDDIT delivery rule names (undo rule name changes made in Balsamic 10.0.1) <a class="reference external" href="https://github.com/Clinical-Genomics/BALSAMIC/pull/977">https://github.com/Clinical-Genomics/BALSAMIC/pull/977</a></p></li>
<li><p>BALSAMIC readthedocs CLI documentation generation  <a class="reference external" href="https://github.com/Clinical-Genomics/BALSAMIC/issues/965">https://github.com/Clinical-Genomics/BALSAMIC/issues/965</a></p></li>
</ul>
</section>
</section>
<section id="id36">
<h2>[10.0.1]<a class="headerlink" href="#id36" title="Permalink to this heading">#</a></h2>
<section id="id37">
<h3>Fixed:<a class="headerlink" href="#id37" title="Permalink to this heading">#</a></h3>
<ul class="simple">
<li><p>Command and condition for TIDDIT and fixed ReadtheDocs <a class="reference external" href="https://github.com/Clinical-Genomics/BALSAMIC/pull/973">https://github.com/Clinical-Genomics/BALSAMIC/pull/973</a></p></li>
<li><p>ReadtheDocs and updated the header <a class="reference external" href="https://github.com/Clinical-Genomics/BALSAMIC/pull/973">https://github.com/Clinical-Genomics/BALSAMIC/pull/973</a></p></li>
</ul>
</section>
<section id="id38">
<h3>Changed:<a class="headerlink" href="#id38" title="Permalink to this heading">#</a></h3>
<ul class="simple">
<li><p>Time allocation in cluster configuration for SV rules <a class="reference external" href="https://github.com/Clinical-Genomics/BALSAMIC/pull/973">https://github.com/Clinical-Genomics/BALSAMIC/pull/973</a></p></li>
</ul>
</section>
</section>
<section id="id39">
<h2>[10.0.0]<a class="headerlink" href="#id39" title="Permalink to this heading">#</a></h2>
<section id="id40">
<h3>Added:<a class="headerlink" href="#id40" title="Permalink to this heading">#</a></h3>
<ul class="simple">
<li><p>New option <cite>analysis-workflow</cite> to balsamic config case CLI <a class="reference external" href="https://github.com/Clinical-Genomics/BALSAMIC/pull/932">https://github.com/Clinical-Genomics/BALSAMIC/pull/932</a></p></li>
<li><p>New python script to edit INFO tags in <cite>vardict</cite> and <cite>tnscope_umi</cite> VCF files <a class="reference external" href="https://github.com/Clinical-Genomics/BALSAMIC/pull/948">https://github.com/Clinical-Genomics/BALSAMIC/pull/948</a></p></li>
<li><p>Added <cite>cyvcf2</cite> and <cite>click</cite> tools to the <cite>varcallpy3</cite> container <a class="reference external" href="https://github.com/Clinical-Genomics/BALSAMIC/pull/948">https://github.com/Clinical-Genomics/BALSAMIC/pull/948</a></p></li>
<li><p>Delly TIDDIT and vcf2cytosure for WGS <a class="reference external" href="https://github.com/Clinical-Genomics/BALSAMIC/pull/947">https://github.com/Clinical-Genomics/BALSAMIC/pull/947</a></p></li>
<li><p><cite>Delly</cite> <cite>TIDDIT</cite> <cite>vcf2cytosure</cite> and method to process SVs and CNVs for WGS <a class="reference external" href="https://github.com/Clinical-Genomics/BALSAMIC/pull/947">https://github.com/Clinical-Genomics/BALSAMIC/pull/947</a></p></li>
<li><p>SV and CNV analysis and <cite>TIDDIT</cite> to balsamic ReadtheDocs <a class="reference external" href="https://github.com/Clinical-Genomics/BALSAMIC/pull/951">https://github.com/Clinical-Genomics/BALSAMIC/pull/951</a></p></li>
<li><p>Gender to <cite>config.json</cite> <a class="reference external" href="https://github.com/Clinical-Genomics/BALSAMIC/pull/955">https://github.com/Clinical-Genomics/BALSAMIC/pull/955</a></p></li>
<li><p>Provided gender as input for <cite>vcf2cyosure</cite> <a class="reference external" href="https://github.com/Clinical-Genomics/BALSAMIC/pull/955">https://github.com/Clinical-Genomics/BALSAMIC/pull/955</a></p></li>
<li><p>SV CNV doc to balsamic READTHEDOCS <a class="reference external" href="https://github.com/Clinical-Genomics/BALSAMIC/pull/960">https://github.com/Clinical-Genomics/BALSAMIC/pull/960</a></p></li>
<li><p>Germline normal SNV VCF file header renaming to be compatible with genotype uploads <a class="reference external" href="https://github.com/Clinical-Genomics/BALSAMIC/issues/882">https://github.com/Clinical-Genomics/BALSAMIC/issues/882</a></p></li>
<li><p>Add tabix and gzip to vcf2cytosure container <a class="reference external" href="https://github.com/Clinical-Genomics/BALSAMIC/pull/969">https://github.com/Clinical-Genomics/BALSAMIC/pull/969</a></p></li>
</ul>
</section>
<section id="id41">
<h3>Changed:<a class="headerlink" href="#id41" title="Permalink to this heading">#</a></h3>
<ul class="simple">
<li><p>UMI-workflow for panel cases to be run only with <cite>balsamic-umi</cite> flag <a class="reference external" href="https://github.com/Clinical-Genomics/BALSAMIC/issues/896">https://github.com/Clinical-Genomics/BALSAMIC/issues/896</a></p></li>
<li><p>Update <cite>codecov</cite> action version to &#64;v2 <a class="reference external" href="https://github.com/Clinical-Genomics/BALSAMIC/pull/941">https://github.com/Clinical-Genomics/BALSAMIC/pull/941</a></p></li>
<li><p>QC-workflow for panel cases to be run only with <cite>balsamic-qc</cite> <a class="reference external" href="https://github.com/Clinical-Genomics/BALSAMIC/pull/942">https://github.com/Clinical-Genomics/BALSAMIC/pull/942</a></p></li>
<li><p><cite>get_snakefile</cite> function takes the argument <cite>analysis_workflow</cite> to trigger the QC workflow when necessary <a class="reference external" href="https://github.com/Clinical-Genomics/BALSAMIC/pull/942">https://github.com/Clinical-Genomics/BALSAMIC/pull/942</a></p></li>
<li><p><cite>bcftools_counts</cite> input depending on <cite>analysis_workflow</cite> <a class="reference external" href="https://github.com/Clinical-Genomics/BALSAMIC/pull/942">https://github.com/Clinical-Genomics/BALSAMIC/pull/942</a></p></li>
<li><p>UMI output filename <cite>TNscope_umi</cite> is changed to <cite>tnscope_umi</cite> <a class="reference external" href="https://github.com/Clinical-Genomics/BALSAMIC/pull/948">https://github.com/Clinical-Genomics/BALSAMIC/pull/948</a></p></li>
<li><p>Update <cite>delly</cite> to v1.0.3 <a class="reference external" href="https://github.com/Clinical-Genomics/BALSAMIC/pull/950">https://github.com/Clinical-Genomics/BALSAMIC/pull/950</a></p></li>
<li><p>Update versions of <cite>delly</cite> in ReadtheDocs <a class="reference external" href="https://github.com/Clinical-Genomics/BALSAMIC/pull/951">https://github.com/Clinical-Genomics/BALSAMIC/pull/951</a></p></li>
<li><p>Provided gender as input for <cite>ascat</cite> and <cite>cnvkit</cite> <a class="reference external" href="https://github.com/Clinical-Genomics/BALSAMIC/pull/955">https://github.com/Clinical-Genomics/BALSAMIC/pull/955</a></p></li>
<li><p>Update QC criteria for panel and wgs analysis according to <a class="reference external" href="https://github.com/Clinical-Genomics/project-planning/issues/338#issuecomment-1132643330">https://github.com/Clinical-Genomics/project-planning/issues/338#issuecomment-1132643330</a>. <a class="reference external" href="https://github.com/Clinical-Genomics/BALSAMIC/pull/952">https://github.com/Clinical-Genomics/BALSAMIC/pull/952</a></p></li>
<li><p>For uploads to scout, increasing the number of variants failing threshold from 10000 to 50000 <a class="reference external" href="https://github.com/Clinical-Genomics/BALSAMIC/pull/952">https://github.com/Clinical-Genomics/BALSAMIC/pull/952</a></p></li>
</ul>
</section>
<section id="id42">
<h3>Fixed:<a class="headerlink" href="#id42" title="Permalink to this heading">#</a></h3>
<ul class="simple">
<li><p>GENOME_VERSION set to the different genome_version options and replaced with config[“reference”][“genome_version”] <a class="reference external" href="https://github.com/Clinical-Genomics/BALSAMIC/pull/942">https://github.com/Clinical-Genomics/BALSAMIC/pull/942</a></p></li>
<li><p><cite>run_validate.sh</cite> script <a class="reference external" href="https://github.com/Clinical-Genomics/BALSAMIC/pull/952">https://github.com/Clinical-Genomics/BALSAMIC/pull/952</a></p></li>
<li><p>Somatic SV tumor normal rules <a class="reference external" href="https://github.com/Clinical-Genomics/BALSAMIC/pull/959">https://github.com/Clinical-Genomics/BALSAMIC/pull/959</a></p></li>
<li><p>Missing <cite>genderChr</cite> flag for <cite>ascat_tumor_normal</cite> rule <a class="reference external" href="https://github.com/Clinical-Genomics/BALSAMIC/pull/963">https://github.com/Clinical-Genomics/BALSAMIC/pull/963</a></p></li>
<li><p>Command in vcf2cytosure rule and updated ReadtheDocs <a class="reference external" href="https://github.com/Clinical-Genomics/BALSAMIC/pull/966">https://github.com/Clinical-Genomics/BALSAMIC/pull/966</a></p></li>
<li><p>Missing name <cite>analysis_dir</cite> in QC.smk <a class="reference external" href="https://github.com/Clinical-Genomics/BALSAMIC/pull/970">https://github.com/Clinical-Genomics/BALSAMIC/pull/970</a></p></li>
<li><p>Remove <cite>sample_type</cite> wildcard from the <cite>vcfheader_rename_germline</cite> rule and change genotype file name <a class="reference external" href="https://github.com/Clinical-Genomics/BALSAMIC/pull/971">https://github.com/Clinical-Genomics/BALSAMIC/pull/971</a></p></li>
</ul>
</section>
<section id="id43">
<h3>Removed<a class="headerlink" href="#id43" title="Permalink to this heading">#</a></h3>
<ul class="simple">
<li><p>Removed <cite>qc_panel</cite> config in favor of standard config <a class="reference external" href="https://github.com/Clinical-Genomics/BALSAMIC/pull/942">https://github.com/Clinical-Genomics/BALSAMIC/pull/942</a></p></li>
<li><p>Removed cli <cite>–analysis_type</cite> for <cite>balsamic report deliver</cite> command and <cite>balsamic run analysis</cite> <a class="reference external" href="https://github.com/Clinical-Genomics/BALSAMIC/pull/942">https://github.com/Clinical-Genomics/BALSAMIC/pull/942</a></p></li>
<li><p>Removed <cite>analysis_type</cite>: <cite>qc_panel</cite> and replace the trigger for QC workflow by <cite>analysis_workflow</cite>: <cite>balsamic-qc</cite> <a class="reference external" href="https://github.com/Clinical-Genomics/BALSAMIC/pull/942">https://github.com/Clinical-Genomics/BALSAMIC/pull/942</a></p></li>
<li><p>Outdated balsamic report files (<cite>balsamic_report.html</cite> &amp; <cite>balsamic_report.md</cite>) <a class="reference external" href="https://github.com/Clinical-Genomics/BALSAMIC/pull/952">https://github.com/Clinical-Genomics/BALSAMIC/pull/952</a></p></li>
</ul>
</section>
</section>
<section id="id44">
<h2>[9.0.1]<a class="headerlink" href="#id44" title="Permalink to this heading">#</a></h2>
<section id="id45">
<h3>Fixed:<a class="headerlink" href="#id45" title="Permalink to this heading">#</a></h3>
<ul class="simple">
<li><p>Revert <cite>csvkit</cite> tool in align_qc container <a class="reference external" href="https://github.com/Clinical-Genomics/BALSAMIC/pull/928">https://github.com/Clinical-Genomics/BALSAMIC/pull/928</a></p></li>
<li><p>Automatic version update for balsamic methods <a class="reference external" href="https://github.com/Clinical-Genomics/BALSAMIC/pull/930">https://github.com/Clinical-Genomics/BALSAMIC/pull/930</a></p></li>
</ul>
</section>
</section>
<section id="id46">
<h2>[9.0.0]<a class="headerlink" href="#id46" title="Permalink to this heading">#</a></h2>
<section id="id47">
<h3>Added:<a class="headerlink" href="#id47" title="Permalink to this heading">#</a></h3>
<ul class="simple">
<li><p>Snakemake workflow to create canfam3 reference <a class="reference external" href="https://github.com/Clinical-Genomics/BALSAMIC/pull/843">https://github.com/Clinical-Genomics/BALSAMIC/pull/843</a></p></li>
<li><p>Call umi variants using TNscope in bed defined regions <a class="reference external" href="https://github.com/Clinical-Genomics/BALSAMIC/issues/821">https://github.com/Clinical-Genomics/BALSAMIC/issues/821</a></p></li>
<li><p>UMI duplication metrics to report in multiqc_picard_dups.json <a class="reference external" href="https://github.com/Clinical-Genomics/BALSAMIC/issues/844">https://github.com/Clinical-Genomics/BALSAMIC/issues/844</a></p></li>
<li><p>Option to use PON reference in cnv calling for TGA tumor-only cases <a class="reference external" href="https://github.com/Clinical-Genomics/BALSAMIC/pull/851">https://github.com/Clinical-Genomics/BALSAMIC/pull/851</a></p></li>
<li><p>QC default validation conditions (for not defined capture kits) <a class="reference external" href="https://github.com/Clinical-Genomics/BALSAMIC/pull/855">https://github.com/Clinical-Genomics/BALSAMIC/pull/855</a></p></li>
<li><p>SVdb to the varcall_py36 container <a class="reference external" href="https://github.com/Clinical-Genomics/BALSAMIC/pull/872">https://github.com/Clinical-Genomics/BALSAMIC/pull/872</a></p></li>
<li><p>SVdb to WGS workflow <a class="reference external" href="https://github.com/Clinical-Genomics/BALSAMIC/pull/873">https://github.com/Clinical-Genomics/BALSAMIC/pull/873</a></p></li>
<li><p>Docker container for vcf2cytosure <a class="reference external" href="https://github.com/Clinical-Genomics/BALSAMIC/pull/869">https://github.com/Clinical-Genomics/BALSAMIC/pull/869</a></p></li>
<li><p>Snakemake rule for creating <cite>.cgh</cite> files from <cite>CNVkit</cite> outputs <a class="reference external" href="https://github.com/Clinical-Genomics/BALSAMIC/pull/880">https://github.com/Clinical-Genomics/BALSAMIC/pull/880</a></p></li>
<li><p>SVdb to TGA workflow <a class="reference external" href="https://github.com/Clinical-Genomics/BALSAMIC/pull/879">https://github.com/Clinical-Genomics/BALSAMIC/pull/879</a></p></li>
<li><p>SVdb merge SV and CNV <a class="reference external" href="https://github.com/Clinical-Genomics/BALSAMIC/pull/886">https://github.com/Clinical-Genomics/BALSAMIC/pull/886</a></p></li>
<li><p>Readthedocs for BALSAMIC method descriptions <a class="reference external" href="https://github.com/Clinical-Genomics/BALSAMIC/pull/906">https://github.com/Clinical-Genomics/BALSAMIC/pull/906</a></p></li>
<li><p>Readthedocs for BALSAMIC variant filters for WGS somatic callers <a class="reference external" href="https://github.com/Clinical-Genomics/BALSAMIC/pull/906">https://github.com/Clinical-Genomics/BALSAMIC/pull/906</a></p></li>
<li><p>bcftools counts to varcall filter rules <a class="reference external" href="https://github.com/Clinical-Genomics/BALSAMIC/pull/899">https://github.com/Clinical-Genomics/BALSAMIC/pull/899</a></p></li>
<li><p>Additional WGS metrics to be stored in <code class="docutils literal notranslate"><span class="pre">&lt;case&gt;_metrics_deliverables.yaml</span></code> <a class="reference external" href="https://github.com/Clinical-Genomics/BALSAMIC/pull/907">https://github.com/Clinical-Genomics/BALSAMIC/pull/907</a></p></li>
<li><p>ascatNGS copynumber file <a class="reference external" href="https://github.com/Clinical-Genomics/BALSAMIC/pull/914">https://github.com/Clinical-Genomics/BALSAMIC/pull/914</a></p></li>
<li><p>ReadtheDocs for BALSAMIC annotation resources <a class="reference external" href="https://github.com/Clinical-Genomics/BALSAMIC/pull/916">https://github.com/Clinical-Genomics/BALSAMIC/pull/916</a></p></li>
<li><p>Delly CNV for tumor only workflow <a class="reference external" href="https://github.com/Clinical-Genomics/BALSAMIC/pull/923">https://github.com/Clinical-Genomics/BALSAMIC/pull/923</a></p></li>
<li><p>Delly CNV Read-depth profiles for tumor only workflows <a class="reference external" href="https://github.com/Clinical-Genomics/BALSAMIC/pull/924">https://github.com/Clinical-Genomics/BALSAMIC/pull/924</a></p></li>
<li><p>New metric to be extracted and validated: <code class="docutils literal notranslate"><span class="pre">NUMBER_OF_SITES</span></code> (<code class="docutils literal notranslate"><span class="pre">bcftools</span></code> counts) <a class="reference external" href="https://github.com/Clinical-Genomics/BALSAMIC/pull/925">https://github.com/Clinical-Genomics/BALSAMIC/pull/925</a></p></li>
</ul>
</section>
<section id="id48">
<h3>Changed:<a class="headerlink" href="#id48" title="Permalink to this heading">#</a></h3>
<ul class="simple">
<li><p>Merge QC metric extraction workflows <a class="reference external" href="https://github.com/Clinical-Genomics/BALSAMIC/pull/833">https://github.com/Clinical-Genomics/BALSAMIC/pull/833</a></p></li>
<li><p>Changed the base-image for balsamic container to 4.10.3-alpine <a class="reference external" href="https://github.com/Clinical-Genomics/BALSAMIC/pull/869">https://github.com/Clinical-Genomics/BALSAMIC/pull/869</a></p></li>
<li><p>Updated SVdb to 2.6.0 <a class="reference external" href="https://github.com/Clinical-Genomics/BALSAMIC/pull/901">https://github.com/Clinical-Genomics/BALSAMIC/pull/901</a></p></li>
<li><p>Upgrade black to 22.3.0</p></li>
<li><p>For UMI workflow, post filter <cite>gnomad_pop_freq</cite> value is changed from <cite>0.005</cite> to <cite>0.02</cite> <a class="reference external" href="https://github.com/Clinical-Genomics/BALSAMIC/pull/919">https://github.com/Clinical-Genomics/BALSAMIC/pull/919</a></p></li>
<li><p>updated delly to 0.9.1 <a class="reference external" href="https://github.com/Clinical-Genomics/BALSAMIC/pull/920">https://github.com/Clinical-Genomics/BALSAMIC/pull/920</a></p></li>
<li><p>container base_image (align_qc, annotate, coverage_qc, varcall_cnvkit, varcall_py36) to 4.10.3-alpine <a class="reference external" href="https://github.com/Clinical-Genomics/BALSAMIC/pull/921">https://github.com/Clinical-Genomics/BALSAMIC/pull/921</a></p></li>
<li><p>update container (align_qc, annotate, coverage_qc, varcall_cnvkit,varcall_py36) bioinfo tool versions  <a class="reference external" href="https://github.com/Clinical-Genomics/BALSAMIC/pull/921">https://github.com/Clinical-Genomics/BALSAMIC/pull/921</a></p></li>
<li><p>update tool versions (align_qc, annotate, coverage_qc, varcall_cnvkit) in methods and softwares docs <a class="reference external" href="https://github.com/Clinical-Genomics/BALSAMIC/pull/921">https://github.com/Clinical-Genomics/BALSAMIC/pull/921</a></p></li>
<li><p>Updated the list of files to be stored and delivered <a class="reference external" href="https://github.com/Clinical-Genomics/BALSAMIC/pull/915">https://github.com/Clinical-Genomics/BALSAMIC/pull/915</a></p></li>
<li><p>Moved <code class="docutils literal notranslate"><span class="pre">collect_custom_qc_metrics</span></code> rule from <code class="docutils literal notranslate"><span class="pre">multiqc.rule</span></code> <a class="reference external" href="https://github.com/Clinical-Genomics/BALSAMIC/pull/925">https://github.com/Clinical-Genomics/BALSAMIC/pull/925</a></p></li>
</ul>
</section>
<section id="id49">
<h3>Fixed:<a class="headerlink" href="#id49" title="Permalink to this heading">#</a></h3>
<ul class="simple">
<li><p>Automate balsamic version for readthedocs install page <a class="reference external" href="https://github.com/Clinical-Genomics/BALSAMIC/pull/888">https://github.com/Clinical-Genomics/BALSAMIC/pull/888</a></p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">collect_qc_metrics.py</span></code> failing for WGS cases with empty <code class="docutils literal notranslate"><span class="pre">capture_kit</span></code> argument <a class="reference external" href="https://github.com/Clinical-Genomics/BALSAMIC/pull/850">https://github.com/Clinical-Genomics/BALSAMIC/pull/850</a></p></li>
<li><p>QC metric validation for different panel bed version <a class="reference external" href="https://github.com/Clinical-Genomics/BALSAMIC/pull/855">https://github.com/Clinical-Genomics/BALSAMIC/pull/855</a></p></li>
<li><p>Fixed development version of <code class="docutils literal notranslate"><span class="pre">fpdf2</span></code> to <code class="docutils literal notranslate"><span class="pre">2.4.6</span></code> <a class="reference external" href="https://github.com/Clinical-Genomics/BALSAMIC/issues/878">https://github.com/Clinical-Genomics/BALSAMIC/issues/878</a></p></li>
<li><p>Added missing svdb index file <a class="reference external" href="https://github.com/Clinical-Genomics/BALSAMIC/issues/848">https://github.com/Clinical-Genomics/BALSAMIC/issues/848</a></p></li>
</ul>
</section>
<section id="id50">
<h3>Removed<a class="headerlink" href="#id50" title="Permalink to this heading">#</a></h3>
<ul class="simple">
<li><p><code class="docutils literal notranslate"><span class="pre">--qc-metrics/--no-qc-metrics</span></code> flag from the <code class="docutils literal notranslate"><span class="pre">balsamic</span> <span class="pre">report</span> <span class="pre">deliver</span></code> command <a class="reference external" href="https://github.com/Clinical-Genomics/BALSAMIC/pull/833">https://github.com/Clinical-Genomics/BALSAMIC/pull/833</a></p></li>
<li><p>Unused pon option for SNV calling with TNhaplotyper tumor-only <a class="reference external" href="https://github.com/Clinical-Genomics/BALSAMIC/pull/851">https://github.com/Clinical-Genomics/BALSAMIC/pull/851</a></p></li>
<li><p>SV and CNV callers from annotation and filtering <a class="reference external" href="https://github.com/Clinical-Genomics/BALSAMIC/pull/889">https://github.com/Clinical-Genomics/BALSAMIC/pull/889</a></p></li>
<li><p>vcfanno and COSMIC from SV annotation <a class="reference external" href="https://github.com/Clinical-Genomics/BALSAMIC/pull/891">https://github.com/Clinical-Genomics/BALSAMIC/pull/891</a></p></li>
<li><p>Removed <cite>MSK_impact</cite> and <cite>MSK_impact_noStrelka</cite> json files from config <a class="reference external" href="https://github.com/Clinical-Genomics/BALSAMIC/pull/903">https://github.com/Clinical-Genomics/BALSAMIC/pull/903</a></p></li>
<li><p>Cleanup of <cite>strelka</cite>, <cite>pindel</cite> , <cite>mutect2</cite> variables from BALSAMIC <a class="reference external" href="https://github.com/Clinical-Genomics/BALSAMIC/pull/903">https://github.com/Clinical-Genomics/BALSAMIC/pull/903</a></p></li>
<li><p>bcftools_stats from vep <a class="reference external" href="https://github.com/Clinical-Genomics/BALSAMIC/issues/898">https://github.com/Clinical-Genomics/BALSAMIC/issues/898</a></p></li>
<li><p>QC delivery report workflow (generating the <code class="docutils literal notranslate"><span class="pre">&lt;case&gt;_qc_report.html</span></code> file) <a class="reference external" href="https://github.com/Clinical-Genomics/BALSAMIC/issues/878">https://github.com/Clinical-Genomics/BALSAMIC/issues/878</a></p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">--sample-id-map</span></code> and <code class="docutils literal notranslate"><span class="pre">--case-id-map</span></code> flags from the <code class="docutils literal notranslate"><span class="pre">balsamic</span> <span class="pre">report</span> <span class="pre">deliver</span></code> command <a class="reference external" href="https://github.com/Clinical-Genomics/BALSAMIC/issues/878">https://github.com/Clinical-Genomics/BALSAMIC/issues/878</a></p></li>
<li><p>Removed <cite>gatk_haplotypecaller</cite> for reporting panel germline variants <a class="reference external" href="https://github.com/Clinical-Genomics/BALSAMIC/issues/918">https://github.com/Clinical-Genomics/BALSAMIC/issues/918</a></p></li>
</ul>
</section>
</section>
<section id="id51">
<h2>[8.2.10]<a class="headerlink" href="#id51" title="Permalink to this heading">#</a></h2>
<section id="id52">
<h3>Added:<a class="headerlink" href="#id52" title="Permalink to this heading">#</a></h3>
<ul class="simple">
<li><p><cite>libopenblas=0.3.20</cite> dependency to annotate container for fixing bcftools #909</p></li>
</ul>
</section>
<section id="fixes">
<h3>Fixes:<a class="headerlink" href="#fixes" title="Permalink to this heading">#</a></h3>
<ul class="simple">
<li><p>bcftools version locked at <cite>1.10</cite> #909</p></li>
</ul>
</section>
<section id="id53">
<h3>Changed:<a class="headerlink" href="#id53" title="Permalink to this heading">#</a></h3>
<ul class="simple">
<li><p>base image of balsamic container to <cite>4.10.3-alphine</cite> #909</p></li>
<li><p>Replaced annotate container tests with new code #909</p></li>
</ul>
</section>
<section id="id54">
<h3>Removed:<a class="headerlink" href="#id54" title="Permalink to this heading">#</a></h3>
<ul class="simple">
<li><p>Removed failed <cite>vcf2cytosure</cite> installation from annotate container #909</p></li>
</ul>
</section>
</section>
<section id="id55">
<h2>[8.2.9]<a class="headerlink" href="#id55" title="Permalink to this heading">#</a></h2>
<section id="id56">
<h3>Added:<a class="headerlink" href="#id56" title="Permalink to this heading">#</a></h3>
<ul class="simple">
<li><p>Added slurm qos tag <cite>express</cite> #885</p></li>
<li><p>Included more text about UMI-workflow variant calling settings to the readthedocs #888</p></li>
<li><p>Extend QCModel to include <cite>n_base_limit</cite> which outputs in config json <cite>QC</cite> dict</p></li>
</ul>
</section>
<section id="id57">
<h3>Fixes:<a class="headerlink" href="#id57" title="Permalink to this heading">#</a></h3>
<ul class="simple">
<li><p>Automate balsamic version for readthedocs install page #888</p></li>
</ul>
</section>
<section id="id58">
<h3>Changed:<a class="headerlink" href="#id58" title="Permalink to this heading">#</a></h3>
<ul class="simple">
<li><p>Upgrade black to 22.3.0</p></li>
<li><p>fastp default setting of <cite>n_base_limit</cite> is changed to <cite>50</cite> from <cite>5</cite></p></li>
</ul>
</section>
</section>
<section id="id59">
<h2>[8.2.8]<a class="headerlink" href="#id59" title="Permalink to this heading">#</a></h2>
<section id="id60">
<h3>Added:<a class="headerlink" href="#id60" title="Permalink to this heading">#</a></h3>
<ul class="simple">
<li><p>Added the readthedocs page for BALSAMIC variant-calling filters #867</p></li>
<li><p>Project requirements (setup.py) to build the docs #874</p></li>
<li><p>Generate cram from umi-consensus called bam files #865</p></li>
</ul>
</section>
<section id="id61">
<h3>Changed:<a class="headerlink" href="#id61" title="Permalink to this heading">#</a></h3>
<ul class="simple">
<li><p>Updated the bioinfo tools version numbers in BALSAMIC readthedocs #867</p></li>
<li><p>Sphinx version fixed to &lt;0.18 #874</p></li>
<li><p>Sphinx GitHub action triggers only on master branch PRs</p></li>
<li><p>VAF filter for reporting somatic variants (Vardict) is minimised to 0.7% from 1% #876</p></li>
</ul>
</section>
<section id="id62">
<h3>Fixes:<a class="headerlink" href="#id62" title="Permalink to this heading">#</a></h3>
<ul class="simple">
<li><p>cyvcf2 mock import for READTHEDOCS environment #874</p></li>
</ul>
</section>
</section>
<section id="id63">
<h2>[8.2.7]<a class="headerlink" href="#id63" title="Permalink to this heading">#</a></h2>
<section id="id64">
<h3>Fixes:<a class="headerlink" href="#id64" title="Permalink to this heading">#</a></h3>
<ul class="simple">
<li><p>Fixes fastqc timeout issues for wgs cases #861</p></li>
<li><p>Fix cluster configuration for vep and vcfanno #857</p></li>
</ul>
</section>
</section>
<section id="id65">
<h2>[8.2.6]<a class="headerlink" href="#id65" title="Permalink to this heading">#</a></h2>
<section id="id66">
<h3>Fixes:<a class="headerlink" href="#id66" title="Permalink to this heading">#</a></h3>
<ul class="simple">
<li><p>Set right qos in scheduler command #856</p></li>
</ul>
</section>
</section>
<section id="id67">
<h2>[8.2.5]<a class="headerlink" href="#id67" title="Permalink to this heading">#</a></h2>
<ul class="simple">
<li><p>balsamic.sif container installation during cache generation #841</p></li>
</ul>
<section id="id68">
<h3>Fixed:<a class="headerlink" href="#id68" title="Permalink to this heading">#</a></h3>
<ul class="simple">
<li><p>Execution of <cite>create_pdf</cite> python script inside the balsamic container #841</p></li>
</ul>
</section>
</section>
<section id="id69">
<h2>[8.2.4]<a class="headerlink" href="#id69" title="Permalink to this heading">#</a></h2>
<section id="id70">
<h3>Added:<a class="headerlink" href="#id70" title="Permalink to this heading">#</a></h3>
<ul class="simple">
<li><p><code class="docutils literal notranslate"><span class="pre">--hgvsg</span></code> annotation to VEP #830</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">ascatNgs</span></code> PDF delivery (plots &amp; statistics) #828</p></li>
</ul>
</section>
</section>
<section id="id71">
<h2>[8.2.3]<a class="headerlink" href="#id71" title="Permalink to this heading">#</a></h2>
<section id="id72">
<h3>Fixed:<a class="headerlink" href="#id72" title="Permalink to this heading">#</a></h3>
<ul class="simple">
<li><p>Add default for gender if <code class="docutils literal notranslate"><span class="pre">purecn</span></code> captures dual gender values #824</p></li>
</ul>
</section>
<section id="id73">
<h3>Changed:<a class="headerlink" href="#id73" title="Permalink to this heading">#</a></h3>
<ul class="simple">
<li><p>Updated <code class="docutils literal notranslate"><span class="pre">purecn</span></code> and its dependencies to latest versions</p></li>
</ul>
</section>
</section>
<section id="id74">
<h2>[8.2.2]<a class="headerlink" href="#id74" title="Permalink to this heading">#</a></h2>
<section id="id75">
<h3>Added:<a class="headerlink" href="#id75" title="Permalink to this heading">#</a></h3>
<ul class="simple">
<li><p><code class="docutils literal notranslate"><span class="pre">ascatNGS</span></code> tumor normal delivery #810</p></li>
</ul>
</section>
<section id="id76">
<h3>Changed:<a class="headerlink" href="#id76" title="Permalink to this heading">#</a></h3>
<ul class="simple">
<li><p>QC metrics delivery tag #820</p></li>
<li><p>Refactor tmb rule that contains redundant line #817</p></li>
</ul>
</section>
</section>
<section id="id77">
<h2>[8.2.1]<a class="headerlink" href="#id77" title="Permalink to this heading">#</a></h2>
<section id="id78">
<h3>Fixed:<a class="headerlink" href="#id78" title="Permalink to this heading">#</a></h3>
<ul class="simple">
<li><p><code class="docutils literal notranslate"><span class="pre">cnvkit</span></code> gender comparison operator bug #819</p></li>
</ul>
</section>
</section>
<section id="id79">
<h2>[8.2.0]<a class="headerlink" href="#id79" title="Permalink to this heading">#</a></h2>
<section id="id80">
<h3>Added:<a class="headerlink" href="#id80" title="Permalink to this heading">#</a></h3>
<ul class="simple">
<li><p>Added various basic filters to all variant callers irregardless of their delivery status #750</p></li>
<li><p>BALSAMIC container #728</p></li>
<li><p>BALSAMIC reference generation via cluster submission for both reference and container #686</p></li>
<li><p>Container specific tests #770</p></li>
<li><p>BALSAMIC quality control metrics extraction and validation #754</p></li>
<li><p>Delly is added as a submodule and removed from rest of the conda environments #787</p></li>
<li><p>Store research VCFs for all filtered and annotated VCF files</p></li>
<li><p>Added <cite>.,PASS</cite> to all structural variant filter rules to resolve the issues with missing calls in filtered file</p></li>
<li><p>Handling of QC metrics validation errors #783</p></li>
<li><p>Github Action workflow that builds the docs using Sphinx #809</p></li>
<li><p>Zenodo integration to create citable link #813</p></li>
<li><p>Panel BED specific QC conditions #800</p></li>
<li><p>Metric extraction to a YAML file for Vogue #802</p></li>
</ul>
</section>
<section id="id81">
<h3>Changed:<a class="headerlink" href="#id81" title="Permalink to this heading">#</a></h3>
<ul class="simple">
<li><p>refactored main workflow with more readible organization #614</p></li>
<li><p>refactored conda envs within container to be on base and container definition is uncoupled #759</p></li>
<li><p>renamed umi output file names to fix issue with picard HSmetrics #804</p></li>
<li><p>locked requirements for graphviz io 0.16 #811</p></li>
<li><p>QC metric validation is performed across all metrics of each of the samples #800</p></li>
</ul>
</section>
<section id="id82">
<h3>Removed:<a class="headerlink" href="#id82" title="Permalink to this heading">#</a></h3>
<ul class="simple">
<li><p>The option of running umiworkflow independently with balsamic command-line option “-a umi”</p></li>
<li><p>Removed source activate from reference and pon workflows #764</p></li>
</ul>
</section>
<section id="id83">
<h3>Fixed:<a class="headerlink" href="#id83" title="Permalink to this heading">#</a></h3>
<ul class="simple">
<li><p>Pip installation failure inside balsamic container #758</p></li>
<li><p>Fixed issue #768 with missing <code class="docutils literal notranslate"><span class="pre">vep_install</span></code> command in container</p></li>
<li><p>Fixed issue #765 with correct input bam files for SV rules</p></li>
<li><p>Continuation of CNVkit even if <code class="docutils literal notranslate"><span class="pre">PURECN</span></code> fails and fix <code class="docutils literal notranslate"><span class="pre">PureCN</span></code> conda paths #774 #775</p></li>
<li><p>Locked version for <code class="docutils literal notranslate"><span class="pre">cryptography</span></code> package</p></li>
<li><p>Bumped version for <code class="docutils literal notranslate"><span class="pre">bcftools</span></code> in cnvkit container</p></li>
<li><p>Fixed issues #776 and #777 with correct install paths for gatk and manta</p></li>
<li><p>Fixed issue #782 for missing AF in the vcf INFO field</p></li>
<li><p>Fixed issues #748 #749 with correct sample names</p></li>
<li><p>Fixed issue #767 for ascatngs hardcoded values</p></li>
<li><p>Fixed missing output option in bcftools filters for tnhaplotyper #793</p></li>
<li><p>Fixed issue #795 with increasing resources for vep and filter SV prior to vep</p></li>
<li><p>Building <code class="docutils literal notranslate"><span class="pre">wheel</span></code> for <code class="docutils literal notranslate"><span class="pre">cryptography</span></code> bug inside BALSAMIC container #801</p></li>
<li><p>Fixed badget for docker container master and develop status</p></li>
<li><p>ReadtheDocs building failure due to dependencies, fixed by locking versions #773</p></li>
<li><p>Dev requirements installation for Sphinx docs (Github Action) #812</p></li>
<li><p>Changed path for main Dockerfile version in <code class="docutils literal notranslate"><span class="pre">.bumpversion.cfg</span></code></p></li>
</ul>
</section>
</section>
<section id="id84">
<h2>[8.1.0]<a class="headerlink" href="#id84" title="Permalink to this heading">#</a></h2>
<section id="id85">
<h3>Added:<a class="headerlink" href="#id85" title="Permalink to this heading">#</a></h3>
<ul class="simple">
<li><p>Workflow to check PR tiltes to make easier to tell PR intents #724</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">bcftools</span> <span class="pre">stats</span></code>  to calculate Ti/Tv for all post annotate germline and somatic calls #93</p></li>
<li><p>Added reference download date to <code class="docutils literal notranslate"><span class="pre">reference.json</span></code> #726</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">ascatngs</span></code> hg38 references to constants #683</p></li>
<li><p>Added ClinVar as a source to download and to be annotated with VCFAnno #737</p></li>
</ul>
</section>
<section id="id86">
<h3>Changed:<a class="headerlink" href="#id86" title="Permalink to this heading">#</a></h3>
<ul class="simple">
<li><p>Updated docs for git FAQs #731</p></li>
<li><p>Rename panel of normal filename Clinical-Genomics/cgp-cancer-cnvcall#10</p></li>
</ul>
</section>
<section id="id87">
<h3>Fixed:<a class="headerlink" href="#id87" title="Permalink to this heading">#</a></h3>
<ul class="simple">
<li><p>Fixed bug with using varcall_py36 container with VarDict #739</p></li>
<li><p>Fixed a bug with VEP module in MultiQC by excluding #746</p></li>
<li><p>Fixed a bug with <code class="docutils literal notranslate"><span class="pre">bcftools</span> <span class="pre">stats</span></code> results failing in MultiQC #744</p></li>
</ul>
</section>
</section>
<section id="id88">
<h2>[8.0.2]<a class="headerlink" href="#id88" title="Permalink to this heading">#</a></h2>
<section id="id89">
<h3>Fixed:<a class="headerlink" href="#id89" title="Permalink to this heading">#</a></h3>
<ul class="simple">
<li><p>Fixed breaking shell command for VEP annotation rules #734</p></li>
</ul>
</section>
</section>
<section id="id90">
<h2>[8.0.1]<a class="headerlink" href="#id90" title="Permalink to this heading">#</a></h2>
<section id="id91">
<h3>Fixed:<a class="headerlink" href="#id91" title="Permalink to this heading">#</a></h3>
<ul class="simple">
<li><p>Fixed context for Dockerfile for release content #720</p></li>
</ul>
</section>
</section>
<section id="id92">
<h2>[8.0.0]<a class="headerlink" href="#id92" title="Permalink to this heading">#</a></h2>
<section id="id93">
<h3>Added:<a class="headerlink" href="#id93" title="Permalink to this heading">#</a></h3>
<ul class="simple">
<li><p><code class="docutils literal notranslate"><span class="pre">samtools</span></code> flagstats and stats to workflow and MultiQC</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">delly</span> <span class="pre">v0.8.7</span></code> somatic SV caller #644</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">delly</span></code> containter #644</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">bcftools</span> <span class="pre">v1.12</span></code> to <code class="docutils literal notranslate"><span class="pre">delly</span></code> container #644</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">tabix</span> <span class="pre">v0.2.6</span></code> to <code class="docutils literal notranslate"><span class="pre">delly</span></code> container #644</p></li>
<li><p>Passed SV calls from Manta to clinical delivery</p></li>
<li><p>An extra filter to VarDict tumor-normal to remove variants with STATUS=Germline, all other will still be around</p></li>
<li><p>Added <code class="docutils literal notranslate"><span class="pre">vcf2cytosure</span></code> to annotate container</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">git</span></code> to the container definition</p></li>
<li><p>prepare_delly_exclusion rule</p></li>
<li><p>Installation of <code class="docutils literal notranslate"><span class="pre">PureCN</span></code> rpackage in <code class="docutils literal notranslate"><span class="pre">cnvkit</span></code> container</p></li>
<li><p>Calculate tumor-purity and ploidy using <code class="docutils literal notranslate"><span class="pre">PureCN</span></code> for <code class="docutils literal notranslate"><span class="pre">cnvkit</span></code> call</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">ascatngs</span></code> as a submodule #672</p></li>
<li><p>GitHub action to build and test <code class="docutils literal notranslate"><span class="pre">ascatngs</span></code> container</p></li>
<li><p>Reference section to <code class="docutils literal notranslate"><span class="pre">docs/FAQ.rst</span></code></p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">ascatngs</span></code> download references from reference_file repository #672</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">delly</span></code> tumor only rule #644</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">ascatngs</span></code> download container #672</p></li>
<li><p>Documentation update on setting sentieon env variables in <code class="docutils literal notranslate"><span class="pre">bashrc</span></code></p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">ascatngs</span></code> tumor normal rule for wgs cases #672</p></li>
<li><p>Individual rules (i.e. ngs filters) for cnv and sv callers. Only Manta will be delivered and added to the list of output files. #708</p></li>
<li><p>Added “targeted” and “wgs” tags to variant callers to provide another layer of separation. #708</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">manta</span></code> convert inversion #709</p></li>
<li><p>Sentieon version to bioinformatic tool version parsing #685</p></li>
<li><p>added <code class="docutils literal notranslate"><span class="pre">CITATION.cff</span></code> to cite BALSAMIC</p></li>
</ul>
</section>
<section id="id94">
<h3>Changed:<a class="headerlink" href="#id94" title="Permalink to this heading">#</a></h3>
<ul class="simple">
<li><p>Upgrade to latest sentieon version 202010.02</p></li>
<li><p>New name <code class="docutils literal notranslate"><span class="pre">MarkDuplicates</span></code> to <code class="docutils literal notranslate"><span class="pre">picard_markduplicates</span></code> in <code class="docutils literal notranslate"><span class="pre">bwa_mem</span></code> rule and <code class="docutils literal notranslate"><span class="pre">cluster.json</span></code></p></li>
<li><p>New name rule <code class="docutils literal notranslate"><span class="pre">GATK_contest</span></code> to <code class="docutils literal notranslate"><span class="pre">gatk_contest</span></code></p></li>
<li><p>Avoid running pytest github actions workflow on <code class="docutils literal notranslate"><span class="pre">docs/**</span></code> and <code class="docutils literal notranslate"><span class="pre">CHANGELOG.rst</span></code> changes</p></li>
<li><p>Updated <code class="docutils literal notranslate"><span class="pre">snakemake</span></code> to <code class="docutils literal notranslate"><span class="pre">v6.5.3</span></code> #501</p></li>
<li><p>Update <code class="docutils literal notranslate"><span class="pre">GNOMAD</span></code> URL</p></li>
<li><p>Split Tumor-only <code class="docutils literal notranslate"><span class="pre">cnvkit</span> <span class="pre">batch</span></code> into individual commands</p></li>
<li><p>Improved TMB calculation issue #51</p></li>
<li><p>Generalized ascat, delly, and manta result in workflow. #708</p></li>
<li><p>Generalized workflow to eliminate duplicate entries and code. #708</p></li>
<li><p>Split Tumor-Normal <code class="docutils literal notranslate"><span class="pre">cnvkit</span> <span class="pre">batch</span></code> into individual commands</p></li>
<li><p>Moved params that are used in multiple rules to constants #711</p></li>
<li><p>Changed the way conda and non-conda bioinfo tools version are parsed</p></li>
<li><p>Python code formatter changed from Black to YAPF #619</p></li>
</ul>
</section>
<section id="id95">
<h3>Fixed:<a class="headerlink" href="#id95" title="Permalink to this heading">#</a></h3>
<ul class="simple">
<li><p>post-processing of the umi consensus in handling BI tags</p></li>
<li><p>vcf-filtered-clinical tag files will have all variants including PASS</p></li>
<li><p>Refactor snakemake <code class="docutils literal notranslate"><span class="pre">annotate</span></code> rules according to snakemake etiquette #636</p></li>
<li><p>Refactor snakemake <code class="docutils literal notranslate"><span class="pre">align</span></code> rules according to snakemake etiquette #636</p></li>
<li><p>Refactor snakemake <code class="docutils literal notranslate"><span class="pre">fastqc</span></code> <code class="docutils literal notranslate"><span class="pre">vep</span></code> contest and <code class="docutils literal notranslate"><span class="pre">mosdepth</span></code> rules according to <code class="docutils literal notranslate"><span class="pre">snakemake</span></code> etiquette #636</p></li>
<li><p>Order of columns in QC and coverage report issue #601</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">delly</span></code> not showing in workflow at runtime #644</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">ascatngs</span></code> documentation links in <code class="docutils literal notranslate"><span class="pre">FAQs</span></code> #672</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">varcall_py36</span></code> container build and push #703</p></li>
<li><p>Wrong spacing in reference json issue #704</p></li>
<li><p>Refactor snakemake <code class="docutils literal notranslate"><span class="pre">quality</span> <span class="pre">control</span></code> rules according to snakemake etiquette #636</p></li>
</ul>
</section>
<section id="id96">
<h3>Removed:<a class="headerlink" href="#id96" title="Permalink to this heading">#</a></h3>
<ul class="simple">
<li><p>Cleaned up unused container definitions and conda environment files</p></li>
<li><p>Remove cnvkit calling for WGS cases</p></li>
<li><p>Removed the install.sh script</p></li>
</ul>
</section>
</section>
<section id="id97">
<h2>[7.2.5]<a class="headerlink" href="#id97" title="Permalink to this heading">#</a></h2>
<section id="id98">
<h3>Changed:<a class="headerlink" href="#id98" title="Permalink to this heading">#</a></h3>
<ul class="simple">
<li><p>Updated COSMIC path to use version 94</p></li>
</ul>
</section>
</section>
<section id="id99">
<h2>[7.2.5]<a class="headerlink" href="#id99" title="Permalink to this heading">#</a></h2>
<section id="id100">
<h3>Changed:<a class="headerlink" href="#id100" title="Permalink to this heading">#</a></h3>
<ul class="simple">
<li><p>Updated path for gnomad and 1000genomes to a working path from Google Storage</p></li>
</ul>
</section>
</section>
<section id="id101">
<h2>[7.2.4]<a class="headerlink" href="#id101" title="Permalink to this heading">#</a></h2>
<section id="id102">
<h3>Changed:<a class="headerlink" href="#id102" title="Permalink to this heading">#</a></h3>
<ul class="simple">
<li><p>Updated sentieon util sort in umi to use Sentieon 20201002 version</p></li>
</ul>
</section>
</section>
<section id="id103">
<h2>[7.2.3]<a class="headerlink" href="#id103" title="Permalink to this heading">#</a></h2>
<section id="id104">
<h3>Fixed:<a class="headerlink" href="#id104" title="Permalink to this heading">#</a></h3>
<ul class="simple">
<li><p>Fixed memory issue with vcfanno in vep_somatic rule fixes #661</p></li>
</ul>
</section>
</section>
<section id="id105">
<h2>[7.2.2]<a class="headerlink" href="#id105" title="Permalink to this heading">#</a></h2>
<section id="id106">
<h3>Fixed:<a class="headerlink" href="#id106" title="Permalink to this heading">#</a></h3>
<ul class="simple">
<li><p>An error with Sentieon for better management of memory fixes #621</p></li>
</ul>
</section>
</section>
<section id="id107">
<h2>[7.2.1]<a class="headerlink" href="#id107" title="Permalink to this heading">#</a></h2>
<section id="id108">
<h3>Changed:<a class="headerlink" href="#id108" title="Permalink to this heading">#</a></h3>
<ul class="simple">
<li><p>Rename Github actions to reflect their content</p></li>
</ul>
</section>
</section>
<section id="id109">
<h2>[7.2.0]<a class="headerlink" href="#id109" title="Permalink to this heading">#</a></h2>
<section id="id110">
<h3>Added:<a class="headerlink" href="#id110" title="Permalink to this heading">#</a></h3>
<ul class="simple">
<li><p>Changelog reminder workflow to Github</p></li>
<li><p>Snakemake workflow for created PON reference</p></li>
<li><p>Balsamic cli config command(pon) for creating json for PON analysis</p></li>
<li><p>tumor lod option for passing tnscope-umi final variants</p></li>
<li><p>Git guide to make balsamic release in FAQ docs</p></li>
</ul>
</section>
<section id="id111">
<h3>Changed:<a class="headerlink" href="#id111" title="Permalink to this heading">#</a></h3>
<ul class="simple">
<li><p>Expanded multiqc result search dir to whole analysis dir</p></li>
<li><p>Simple test for docker container</p></li>
</ul>
</section>
<section id="id112">
<h3>Fixed:<a class="headerlink" href="#id112" title="Permalink to this heading">#</a></h3>
<ul class="simple">
<li><p>Correctly version bump for Dockerfile</p></li>
</ul>
</section>
<section id="id113">
<h3>Removed:<a class="headerlink" href="#id113" title="Permalink to this heading">#</a></h3>
<ul class="simple">
<li><p>Removed unused Dockerfile releases</p></li>
<li><p>Removed redundant genome version from <code class="docutils literal notranslate"><span class="pre">reference.json</span></code></p></li>
</ul>
</section>
</section>
<section id="id114">
<h2>[7.1.10]<a class="headerlink" href="#id114" title="Permalink to this heading">#</a></h2>
<section id="id115">
<h3>Fixed:<a class="headerlink" href="#id115" title="Permalink to this heading">#</a></h3>
<ul class="simple">
<li><p>Bug in <code class="docutils literal notranslate"><span class="pre">ngs_filter</span></code> rule set for tumor-only WGS</p></li>
<li><p>Missing delivery of tumor only WGS filter</p></li>
</ul>
</section>
</section>
<section id="id116">
<h2>[7.1.9]<a class="headerlink" href="#id116" title="Permalink to this heading">#</a></h2>
<section id="id117">
<h3>Changed:<a class="headerlink" href="#id117" title="Permalink to this heading">#</a></h3>
<ul class="simple">
<li><p>only pass variants are not part of delivery anymore</p></li>
<li><p>delivery tag file ids are properly matched with sample_name</p></li>
<li><p>tabix updated to 0.2.6</p></li>
<li><p>fastp updated to 0.20.1</p></li>
<li><p>samtools updated to 1.12</p></li>
<li><p>bedtools updated to 2.30.0</p></li>
</ul>
</section>
<section id="id118">
<h3>Removed:<a class="headerlink" href="#id118" title="Permalink to this heading">#</a></h3>
<ul class="simple">
<li><p>sentieon-dedup rule from delivery</p></li>
<li><p>Removed all pre filter pass from delivery</p></li>
</ul>
</section>
</section>
<section id="id119">
<h2>[7.1.8]<a class="headerlink" href="#id119" title="Permalink to this heading">#</a></h2>
<section id="id120">
<h3>Fixed:<a class="headerlink" href="#id120" title="Permalink to this heading">#</a></h3>
<ul class="simple">
<li><p>Target coverage (Picard HsMetrics) for UMI files is now correctly calculated.</p></li>
</ul>
</section>
<section id="id121">
<h3>Changed:<a class="headerlink" href="#id121" title="Permalink to this heading">#</a></h3>
<ul class="simple">
<li><p>TNscope calculated AF values are fetched and written to AFtable.txt.</p></li>
</ul>
</section>
</section>
<section id="id122">
<h2>[7.1.7]<a class="headerlink" href="#id122" title="Permalink to this heading">#</a></h2>
<section id="id123">
<h3>Added:<a class="headerlink" href="#id123" title="Permalink to this heading">#</a></h3>
<ul class="simple">
<li><p>ngs_filter_tnscope is also part of deliveries now</p></li>
</ul>
</section>
<section id="id124">
<h3>Changed:<a class="headerlink" href="#id124" title="Permalink to this heading">#</a></h3>
<ul class="simple">
<li><p>rankscore is now a research tag instead of clinical</p></li>
<li><p>Some typo and fixes in the coverage and constant metrics</p></li>
<li><p>Delivery process is more verbose</p></li>
</ul>
</section>
<section id="id125">
<h3>Fixed:<a class="headerlink" href="#id125" title="Permalink to this heading">#</a></h3>
<ul class="simple">
<li><p>CNVKit output is now properly imported in the deliveries and workflow</p></li>
</ul>
</section>
</section>
<section id="id126">
<h2>[7.1.6]<a class="headerlink" href="#id126" title="Permalink to this heading">#</a></h2>
<section id="id127">
<h3>Fixed:<a class="headerlink" href="#id127" title="Permalink to this heading">#</a></h3>
<ul class="simple">
<li><p>CSS style for qc coverage report is changed to landscape</p></li>
</ul>
</section>
</section>
<section id="id128">
<h2>[7.1.5]<a class="headerlink" href="#id128" title="Permalink to this heading">#</a></h2>
<section id="id129">
<h3>Changed:<a class="headerlink" href="#id129" title="Permalink to this heading">#</a></h3>
<ul class="simple">
<li><p>update download url for 1000genome WGS sites from ftp to http</p></li>
</ul>
</section>
</section>
<section id="id130">
<h2>[7.1.4]<a class="headerlink" href="#id130" title="Permalink to this heading">#</a></h2>
<section id="id131">
<h3>Changed:<a class="headerlink" href="#id131" title="Permalink to this heading">#</a></h3>
<ul class="simple">
<li><p>bump picard to version 2.25.0</p></li>
</ul>
</section>
</section>
<section id="id132">
<h2>[7.1.3]<a class="headerlink" href="#id132" title="Permalink to this heading">#</a></h2>
<section id="id133">
<h3>Fixed:<a class="headerlink" href="#id133" title="Permalink to this heading">#</a></h3>
<ul class="simple">
<li><p><code class="docutils literal notranslate"><span class="pre">assets</span></code> path is now added to bind path</p></li>
</ul>
</section>
</section>
<section id="id134">
<h2>[7.1.2]<a class="headerlink" href="#id134" title="Permalink to this heading">#</a></h2>
<section id="id135">
<h3>Fixed:<a class="headerlink" href="#id135" title="Permalink to this heading">#</a></h3>
<ul class="simple">
<li><p>umi_workflow config json is set as true for panel and wgs as false.</p></li>
<li><p>Rename umiconsensus bam file headers from {samplenames} to TUMOR/NORMAL.</p></li>
<li><p>Documentation autobuild on RTFD</p></li>
</ul>
</section>
</section>
<section id="id136">
<h2>[7.1.1]<a class="headerlink" href="#id136" title="Permalink to this heading">#</a></h2>
<section id="id137">
<h3>Fixed:<a class="headerlink" href="#id137" title="Permalink to this heading">#</a></h3>
<ul class="simple">
<li><p>Moved all requirements to setup.py, and added all package_data there. Clean up unused files.</p></li>
</ul>
</section>
</section>
<section id="id138">
<h2>[7.1.0]<a class="headerlink" href="#id138" title="Permalink to this heading">#</a></h2>
<section id="id139">
<h3>Removed<a class="headerlink" href="#id139" title="Permalink to this heading">#</a></h3>
<ul class="simple">
<li><p><code class="docutils literal notranslate"><span class="pre">tnsnv</span></code> removed from WGS analysis, both tumor-only and tumor-normal</p></li>
<li><p>GATK-BaseRecalibrator is removed from all workflows</p></li>
</ul>
</section>
<section id="id140">
<h3>Fixed<a class="headerlink" href="#id140" title="Permalink to this heading">#</a></h3>
<ul class="simple">
<li><p>Fixed issue 577 with missing <code class="docutils literal notranslate"><span class="pre">tumor.merged.bam</span></code> and <code class="docutils literal notranslate"><span class="pre">normal.merged.bam</span></code></p></li>
<li><p>Issue 448 with lingering tmp_dir. It is not deleted after analysis is properly finished.</p></li>
</ul>
</section>
<section id="id141">
<h3>Changed<a class="headerlink" href="#id141" title="Permalink to this heading">#</a></h3>
<ul class="simple">
<li><p>All variant calling rules use proper <code class="docutils literal notranslate"><span class="pre">tumor.merged.bam</span></code> or <code class="docutils literal notranslate"><span class="pre">normal.merged.bam</span></code> as inputs</p></li>
</ul>
</section>
</section>
<section id="id142">
<h2>[7.0.2]<a class="headerlink" href="#id142" title="Permalink to this heading">#</a></h2>
<section id="id143">
<h3>Added<a class="headerlink" href="#id143" title="Permalink to this heading">#</a></h3>
<ul class="simple">
<li><p>Updated docs with FAQ for UMI workflow</p></li>
</ul>
</section>
<section id="id144">
<h3>Fixed<a class="headerlink" href="#id144" title="Permalink to this heading">#</a></h3>
<ul class="simple">
<li><p>fix job scheduling bug for benchmarking</p></li>
<li><p>rankscore’s output is now a proper vcf.gz file</p></li>
<li><p>Manta rules now properly make a sample_name file</p></li>
</ul>
</section>
</section>
<section id="id145">
<h2>[7.0.1]<a class="headerlink" href="#id145" title="Permalink to this heading">#</a></h2>
<section id="id146">
<h3>Added<a class="headerlink" href="#id146" title="Permalink to this heading">#</a></h3>
<ul class="simple">
<li><p>github action workflow to autobuild release containers</p></li>
</ul>
</section>
</section>
<section id="id147">
<h2>[7.0.0]<a class="headerlink" href="#id147" title="Permalink to this heading">#</a></h2>
<section id="id148">
<h3>Added<a class="headerlink" href="#id148" title="Permalink to this heading">#</a></h3>
<ul class="simple">
<li><p><code class="docutils literal notranslate"><span class="pre">balsamic</span> <span class="pre">init</span></code> to download reference and related containers done in PRs #464 #538</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">balsamic</span> <span class="pre">config</span> <span class="pre">case</span></code> now only take a cache path instead of container and reference #538</p></li>
<li><p>UMI workflow added to main workflow in series of PRs #469 #477 #483 #498 #503 #514 #517</p></li>
<li><p>DRAGEN for WGS applications in PR #488</p></li>
<li><p>A framework for QC check PR #401</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">--quiet`</span></code> option for <code class="docutils literal notranslate"><span class="pre">run</span> <span class="pre">analysis</span></code> PR #491</p></li>
<li><p>Benchmark SLURM jobs after the analysis is finished PR #534</p></li>
<li><p>One container per conda environment (i.e. decouple containers) PR #511 #525 #522</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">--disable-variant-caller</span></code> command for <code class="docutils literal notranslate"><span class="pre">report</span> <span class="pre">deliver</span></code> PR #439</p></li>
<li><p>Added genmod and rankscore in series of two PRs #531 and #533</p></li>
<li><p>Variant filtering to Tumor-Normal in PR #534</p></li>
<li><p>Split SNV/InDels and SVs from TNScope variant caller PR #540</p></li>
<li><p>WGS Tumor only variant filters added in PR #548</p></li>
</ul>
</section>
<section id="id149">
<h3>Changed<a class="headerlink" href="#id149" title="Permalink to this heading">#</a></h3>
<ul class="simple">
<li><p>Update Manta to 1.6.0 PR #470</p></li>
<li><p>Update FastQC to 0.11.9 PR #532</p></li>
<li><p>Update BCFTools to 1.11 PR #537</p></li>
<li><p>Update Samtools to 1.11 PR #537</p></li>
<li><p>Increase resources and runtime for various workflows in PRs #482</p></li>
<li><p>Python package dependenicies versions fixed in PR #480</p></li>
<li><p>QoL changes to workflow in series of PR #471</p></li>
<li><p>Series of documentation updates in PRs #489 #553</p></li>
<li><p>QoL changes to scheduler script PR #491</p></li>
<li><p>QoL changes to how temporary directories are handlded PR #516</p></li>
<li><p>TNScope model apply rule merged with TNScope variant calling for tumor-normal in WGS #540</p></li>
<li><p>Decoupled <code class="docutils literal notranslate"><span class="pre">fastp</span></code> rule into two rules to make it possible to use it for UMI runs #570</p></li>
</ul>
</section>
<section id="id150">
<h3>Fixed<a class="headerlink" href="#id150" title="Permalink to this heading">#</a></h3>
<ul class="simple">
<li><p>A bug in Manta variant calling rules that didn’t name samples properly to TUMOR/NORMAL in the VCF file #572</p></li>
</ul>
</section>
</section>
<section id="id151">
<h2>[6.1.2]<a class="headerlink" href="#id151" title="Permalink to this heading">#</a></h2>
<section id="id152">
<h3>Changed<a class="headerlink" href="#id152" title="Permalink to this heading">#</a></h3>
<ul class="simple">
<li><p>Changed hk delivery tag for coverage-qc-report</p></li>
</ul>
</section>
</section>
<section id="id153">
<h2>[6.1.1]<a class="headerlink" href="#id153" title="Permalink to this heading">#</a></h2>
<section id="id154">
<h3>Fixed<a class="headerlink" href="#id154" title="Permalink to this heading">#</a></h3>
<ul class="simple">
<li><p>No UMI trimming for WGS applications #486</p></li>
<li><p>Fixed a bug where BALSAMIC was checking for sacct/jobid file in local mode PR #497</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">readlink</span></code> command in <code class="docutils literal notranslate"><span class="pre">vep_germline</span></code>, <code class="docutils literal notranslate"><span class="pre">vep_somatic</span></code>, <code class="docutils literal notranslate"><span class="pre">split_bed</span></code>, and <code class="docutils literal notranslate"><span class="pre">GATK_popVCF</span></code> #533</p></li>
<li><p>Fix various bugs for memory handling of Picardtools and its executable in PR #534</p></li>
<li><p>Fixed various issues with <code class="docutils literal notranslate"><span class="pre">gsutils</span></code> in PR #550</p></li>
</ul>
</section>
<section id="id155">
<h3>Removed<a class="headerlink" href="#id155" title="Permalink to this heading">#</a></h3>
<ul class="simple">
<li><p><code class="docutils literal notranslate"><span class="pre">gatk-register</span></code> command removed from installing GATK PR #496</p></li>
</ul>
</section>
</section>
<section id="id156">
<h2>[6.1.1]<a class="headerlink" href="#id156" title="Permalink to this heading">#</a></h2>
<ul class="simple">
<li><p>Fixed a bug with missing QC templates after <code class="docutils literal notranslate"><span class="pre">pip</span> <span class="pre">install</span></code></p></li>
</ul>
</section>
<section id="id157">
<h2>[6.1.0]<a class="headerlink" href="#id157" title="Permalink to this heading">#</a></h2>
<section id="id158">
<h3>Added<a class="headerlink" href="#id158" title="Permalink to this heading">#</a></h3>
<ul class="simple">
<li><p>CLI option to expand report generation for TGA and WES runs. Please see <code class="docutils literal notranslate"><span class="pre">balsamic</span> <span class="pre">report</span> <span class="pre">deliver</span> <span class="pre">--help</span></code></p></li>
<li><p>BALSAMIC now generates a custom HTML report for TGA and WES cases.</p></li>
</ul>
</section>
</section>
<section id="id159">
<h2>[6.0.4]<a class="headerlink" href="#id159" title="Permalink to this heading">#</a></h2>
<section id="id160">
<h3>Changed<a class="headerlink" href="#id160" title="Permalink to this heading">#</a></h3>
<ul class="simple">
<li><p>Reduces MQ cutoff from 50 to 40 to only remove obvious artifacts PR #535</p></li>
<li><p>Reduces AF cutoff from 0.02 to 0.01 PR #535</p></li>
</ul>
</section>
</section>
<section id="id161">
<h2>[6.0.3]<a class="headerlink" href="#id161" title="Permalink to this heading">#</a></h2>
<section id="id162">
<h3>Added<a class="headerlink" href="#id162" title="Permalink to this heading">#</a></h3>
<ul class="simple">
<li><p><code class="docutils literal notranslate"><span class="pre">config</span> <span class="pre">case</span></code> subcommand now has <code class="docutils literal notranslate"><span class="pre">--tumor-sample-name</span></code> and <code class="docutils literal notranslate"><span class="pre">--normal-sample-name</span></code></p></li>
</ul>
</section>
<section id="id163">
<h3>Fixed<a class="headerlink" href="#id163" title="Permalink to this heading">#</a></h3>
<ul class="simple">
<li><p>Manta resource allocation is now properly set PR #523</p></li>
<li><p>VarDict resource allocation in cluster.json increased (both core and time allocation) PR #523</p></li>
<li><p>minimum memory request for GATK mutect2 and haplotypecaller is removed and max memory increased PR #523</p></li>
</ul>
</section>
</section>
<section id="id164">
<h2>[6.0.2]<a class="headerlink" href="#id164" title="Permalink to this heading">#</a></h2>
<section id="id165">
<h3>Added<a class="headerlink" href="#id165" title="Permalink to this heading">#</a></h3>
<ul class="simple">
<li><p>Document for Snakemake rule grammar PR #489</p></li>
</ul>
</section>
<section id="id166">
<h3>Fixed<a class="headerlink" href="#id166" title="Permalink to this heading">#</a></h3>
<ul class="simple">
<li><p>removed <code class="docutils literal notranslate"><span class="pre">gatk3-register</span></code> command from Dockerfile(s) PR #508</p></li>
</ul>
</section>
</section>
<section id="id167">
<h2>[6.0.1]<a class="headerlink" href="#id167" title="Permalink to this heading">#</a></h2>
<section id="id168">
<h3>Added<a class="headerlink" href="#id168" title="Permalink to this heading">#</a></h3>
<ul class="simple">
<li><p>A secondary path for latest jobids submitted to cluster (slurm and qsub) PR #465</p></li>
</ul>
</section>
</section>
<section id="id169">
<h2>[6.0.0]<a class="headerlink" href="#id169" title="Permalink to this heading">#</a></h2>
<section id="id170">
<h3>Added<a class="headerlink" href="#id170" title="Permalink to this heading">#</a></h3>
<ul class="simple">
<li><p>UMI workflow using Sentieon tools. Analysis run available via <cite>balsamic run analysis –help</cite> command. PR #359</p></li>
<li><p>VCFutils to create VCF from flat text file. This is for internal purpose to generate validation VCF. PR #349</p></li>
<li><p>Download option for hg38 (not validated) PR #407</p></li>
<li><p>Option to disable variant callers for WES runs. PR #417</p></li>
</ul>
</section>
<section id="id171">
<h3>Fixed<a class="headerlink" href="#id171" title="Permalink to this heading">#</a></h3>
<ul class="simple">
<li><p>Missing cyvcf2 dependency, and changed conda environment for base environment PR #413</p></li>
<li><p>Missing numpy dependency PR #426</p></li>
</ul>
</section>
<section id="id172">
<h3>Changed<a class="headerlink" href="#id172" title="Permalink to this heading">#</a></h3>
<ul class="simple">
<li><p>COSMIC db for hg19 updated to v90 PR #407</p></li>
<li><p>Fastp trimming is now a two-pass trimming and adapter trimming is always enabled. This might affect coverage slightly PR #422</p></li>
<li><p>All containers start with a clean environment #425</p></li>
<li><p>All Sentieon environment variables are now added to config when workflow executes #425</p></li>
<li><p>Branching model will be changed to gitflow</p></li>
</ul>
</section>
</section>
<section id="id173">
<h2>[5.1.0]<a class="headerlink" href="#id173" title="Permalink to this heading">#</a></h2>
<section id="id174">
<h3>Fixed<a class="headerlink" href="#id174" title="Permalink to this heading">#</a></h3>
<ul class="simple">
<li><p>Vardict-java version fixed. This is due to bad dependency and releases available on conda. Anaconda is not yet update with vardict 1.8, but vardict-java 1.8 is there. This causes various random breaks with Vardict’s TSV output. #403</p></li>
</ul>
</section>
<section id="id175">
<h3>Changed<a class="headerlink" href="#id175" title="Permalink to this heading">#</a></h3>
<ul class="simple">
<li><p>Refactored Docker files a bit, preparation for decoupling #403</p></li>
</ul>
</section>
<section id="id176">
<h3>Removed<a class="headerlink" href="#id176" title="Permalink to this heading">#</a></h3>
<ul class="simple">
<li><p>In preparation for GATK4, IndelRealigner is removed #404</p></li>
</ul>
</section>
</section>
<section id="id177">
<h2>[5.0.1]<a class="headerlink" href="#id177" title="Permalink to this heading">#</a></h2>
<section id="id178">
<h3>Added<a class="headerlink" href="#id178" title="Permalink to this heading">#</a></h3>
<ul class="simple">
<li><p>Temp directory for various rules and workflow wide temp directory #396</p></li>
</ul>
</section>
<section id="id179">
<h3>Changed<a class="headerlink" href="#id179" title="Permalink to this heading">#</a></h3>
<ul class="simple">
<li><p>Refactored tags for housekeeper delivery to make them unique #395</p></li>
<li><p>Increased core requirements for mutect2 #396</p></li>
<li><p>GATK3.8 related utils run via jar file instead of gatk3 #396</p></li>
</ul>
</section>
</section>
<section id="id180">
<h2>[5.0.0]<a class="headerlink" href="#id180" title="Permalink to this heading">#</a></h2>
<section id="id181">
<h3>Added<a class="headerlink" href="#id181" title="Permalink to this heading">#</a></h3>
<ul class="simple">
<li><p>Config.json and DAG draph included in Housekeeper report #372</p></li>
<li><p>New output names added to cnvkit_single and cnvkit_paired #372</p></li>
<li><p>New output names added to vep.rule #372</p></li>
<li><p>Delivery option to CLI and what to delivery with delivery params in rules that are needed to be delivered #376</p></li>
<li><p>Reference data model with validation #371</p></li>
<li><p>Added container path to install script #388</p></li>
</ul>
</section>
<section id="id182">
<h3>Changed<a class="headerlink" href="#id182" title="Permalink to this heading">#</a></h3>
<ul class="simple">
<li><p>Delivery file format simplified #376</p></li>
<li><p>VEP rules have “all” and “pass” as output #376</p></li>
<li><p>Downloaded reference structure changed #371</p></li>
<li><p>genome/refseq.flat renamed to genome/refGene.flat #371</p></li>
<li><p>reverted CNVKit to version 0.9.4 #390</p></li>
</ul>
</section>
<section id="id183">
<h3>Fixed<a class="headerlink" href="#id183" title="Permalink to this heading">#</a></h3>
<ul class="simple">
<li><p>Missing pygments to requirements.txt to fix travis CI #364</p></li>
<li><p>Wildcard resolve for deliveries of vep_germline #374</p></li>
<li><p>Missing index file from deliverables #383</p></li>
<li><p>Ambiguous deliveries in vep_somatic and ngs_filters #387</p></li>
<li><p>Updated documentation to match with installation #391</p></li>
</ul>
</section>
<section id="id184">
<h3>Removed<a class="headerlink" href="#id184" title="Permalink to this heading">#</a></h3>
<ul class="simple">
<li><p>Temp files removed from list of outputs in vep.rule #372</p></li>
<li><p>samtools.rule and merged it with bwa_mem #375</p></li>
</ul>
</section>
</section>
<section id="id185">
<h2>[4.5.0]<a class="headerlink" href="#id185" title="Permalink to this heading">#</a></h2>
<section id="id186">
<h3>Added<a class="headerlink" href="#id186" title="Permalink to this heading">#</a></h3>
<ul class="simple">
<li><p>Models to build config case JSON. The models and descriptions of their contents can now be found
in BALSAMIC/utils/models.py</p></li>
<li><p>Added analysis_type to <cite>report deliver</cite> command</p></li>
<li><p>Added report and delivery capability to Alignment workflow</p></li>
<li><p>run_validate.sh now has -d to handle path to analysis_dir (for internal use only) #361</p></li>
</ul>
</section>
<section id="id187">
<h3>Changed<a class="headerlink" href="#id187" title="Permalink to this heading">#</a></h3>
<ul class="simple">
<li><p>Fastq files are no longer being copied as part of creation of the case config file.
A symlink is now created at the destination path instead</p></li>
<li><p>Config structure is no longer contained in a collestion of JSON files.
The config models are now built using Pydantic and are contained in BALSAMIC/utils/models.py</p></li>
</ul>
</section>
<section id="id188">
<h3>Removed<a class="headerlink" href="#id188" title="Permalink to this heading">#</a></h3>
<ul class="simple">
<li><p>Removed command line option “–fastq-prefix” from config case command</p></li>
<li><p>Removed command line option “–config-path” from config case command.
The config is now always saved with default name “case_id.json”</p></li>
<li><p>Removed command line option “–overwrite-config” from config-case command
The command is now always executed with “–overwrite-config True” behavior</p></li>
</ul>
</section>
<section id="refactored">
<h3>Refactored<a class="headerlink" href="#refactored" title="Permalink to this heading">#</a></h3>
<ul class="simple">
<li><p>Refactored BALSAMIC/commands/config/case.py:
Utility functions are moved to BALSAMIC/utils/cli.py
Models for config fields can be found at BALSAMIC/utils/models.py
Context aborts and logging now contained in pilot function
Tests created to support new architecture</p></li>
<li><p>Reduce analysis directory’s storage</p></li>
</ul>
</section>
<section id="id189">
<h3>Fixed<a class="headerlink" href="#id189" title="Permalink to this heading">#</a></h3>
<ul class="simple">
<li><p>Report generation warnings supressed by adding workdirectory</p></li>
<li><p>Missing tag name for germline annotated calls #356</p></li>
<li><p>Bind path is not added as None if analysis type is wgs #357</p></li>
<li><p>Changes vardict to vardict-java #361</p></li>
</ul>
</section>
</section>
<section id="id190">
<h2>[4.4.0]<a class="headerlink" href="#id190" title="Permalink to this heading">#</a></h2>
<section id="id191">
<h3>Added<a class="headerlink" href="#id191" title="Permalink to this heading">#</a></h3>
<ul class="simple">
<li><p>pydantic to validate various models namely variant caller filters</p></li>
</ul>
</section>
<section id="id192">
<h3>Changed<a class="headerlink" href="#id192" title="Permalink to this heading">#</a></h3>
<ul class="simple">
<li><p>Variant caller filters moved into pydantic</p></li>
<li><p>Install script and setup.py</p></li>
<li><p>refactored install script with more log output and added a conda env suffix option</p></li>
<li><p>refactored docker container and decoupled various parts of the workflow</p></li>
</ul>
</section>
</section>
<section id="id193">
<h2>[4.3.0]<a class="headerlink" href="#id193" title="Permalink to this heading">#</a></h2>
<section id="id194">
<h3>Added<a class="headerlink" href="#id194" title="Permalink to this heading">#</a></h3>
<ul class="simple">
<li><p>Added cram files for targeted sequencing runs fixes #286</p></li>
<li><p>Added <cite>mosdepth</cite> to calculate coverage for whole exome and targeted sequencing</p></li>
<li><p>Filter models added for tumor-only mode</p></li>
<li><p>Enabling adapter trim enables pe adapter trim option for fastp</p></li>
<li><p>Annotate germline variant calls</p></li>
<li><p>Baitset name to picard hsmetrics</p></li>
</ul>
</section>
<section id="deprecated">
<h3>Deprecated<a class="headerlink" href="#deprecated" title="Permalink to this heading">#</a></h3>
<ul class="simple">
<li><p>Sambamba coverage and rules will be deprecated</p></li>
</ul>
</section>
<section id="id195">
<h3>Fixed<a class="headerlink" href="#id195" title="Permalink to this heading">#</a></h3>
<ul class="simple">
<li><p>Fixed latest tag in install script</p></li>
<li><p>Fixed lack of naming final annotated VCF TUMOR/NORMAL</p></li>
</ul>
</section>
<section id="id196">
<h3>Changed<a class="headerlink" href="#id196" title="Permalink to this heading">#</a></h3>
<ul class="simple">
<li><p>Increased run time for various slurm jobs fixes #314</p></li>
<li><p>Enabled SV calls for VarDict tumor-only</p></li>
<li><p>Updated <cite>ensembl-vep</cite> to v100.2</p></li>
</ul>
</section>
</section>
<section id="id197">
<h2>[4.2.4]<a class="headerlink" href="#id197" title="Permalink to this heading">#</a></h2>
<section id="id198">
<h3>Fixed<a class="headerlink" href="#id198" title="Permalink to this heading">#</a></h3>
<ul class="simple">
<li><p>Fixed sort issue with bedfiles after 100 slop</p></li>
</ul>
</section>
</section>
<section id="id199">
<h2>[4.2.3]<a class="headerlink" href="#id199" title="Permalink to this heading">#</a></h2>
<section id="id200">
<h3>Added<a class="headerlink" href="#id200" title="Permalink to this heading">#</a></h3>
<ul class="simple">
<li><p>Added Docker container definition for release and bumpversion</p></li>
</ul>
</section>
<section id="id201">
<h3>Changed<a class="headerlink" href="#id201" title="Permalink to this heading">#</a></h3>
<ul class="simple">
<li><p>Quality of life change to rtfd docs</p></li>
</ul>
</section>
<section id="id202">
<h3>Fixed<a class="headerlink" href="#id202" title="Permalink to this heading">#</a></h3>
<ul class="simple">
<li><p>Fix Docker container with faulty git checkout</p></li>
</ul>
</section>
</section>
<section id="id203">
<h2>[4.2.2]<a class="headerlink" href="#id203" title="Permalink to this heading">#</a></h2>
<section id="id204">
<h3>Added<a class="headerlink" href="#id204" title="Permalink to this heading">#</a></h3>
<ul class="simple">
<li><p>Add “SENTIEON_TMPDIR” to wgs workflow</p></li>
</ul>
</section>
</section>
<section id="id205">
<h2>[4.2.1]<a class="headerlink" href="#id205" title="Permalink to this heading">#</a></h2>
<section id="id206">
<h3>Changed<a class="headerlink" href="#id206" title="Permalink to this heading">#</a></h3>
<ul class="simple">
<li><p>Add docker container pull for correct version of install script</p></li>
</ul>
</section>
</section>
<section id="id207">
<h2>[4.2.0]<a class="headerlink" href="#id207" title="Permalink to this heading">#</a></h2>
<section id="id208">
<h3>Added<a class="headerlink" href="#id208" title="Permalink to this heading">#</a></h3>
<ul class="simple">
<li><p>CNV output as VCF</p></li>
<li><p>Vep output for PASSed variants</p></li>
<li><p>Report command with status and delivery subcommands</p></li>
</ul>
</section>
<section id="id209">
<h3>Changed<a class="headerlink" href="#id209" title="Permalink to this heading">#</a></h3>
<ul class="simple">
<li><p>Bed files are slopped 100bp for variant calling fix #262</p></li>
<li><p>Disable vcfmerge</p></li>
<li><p>Picard markduplicate output moved from log to output</p></li>
<li><p>Vep upgraded to 99.1</p></li>
<li><p>Removed SVs from vardict</p></li>
<li><p>Refactored delivery plugins to produce a file with list of output files from workflow</p></li>
<li><p>Updated snakemake to 5.13</p></li>
</ul>
</section>
<section id="id210">
<h3>Fixed<a class="headerlink" href="#id210" title="Permalink to this heading">#</a></h3>
<ul class="simple">
<li><p>Fixed a bug where threads were not sent properly to rules</p></li>
</ul>
</section>
<section id="id211">
<h3>Removed<a class="headerlink" href="#id211" title="Permalink to this heading">#</a></h3>
<ul class="simple">
<li><p>Removed coverage annotation from mutect2</p></li>
<li><p>Removed source deactivate from rules to suppress conda warning</p></li>
<li><p>Removed <code class="docutils literal notranslate"><span class="pre">plugins</span> <span class="pre">delivery</span></code> subcommand</p></li>
<li><p>Removed annotation for germline caller results</p></li>
</ul>
</section>
</section>
<section id="id212">
<h2>[4.1.0]<a class="headerlink" href="#id212" title="Permalink to this heading">#</a></h2>
<section id="id213">
<h3>Added<a class="headerlink" href="#id213" title="Permalink to this heading">#</a></h3>
<ul class="simple">
<li><p>VEP now also produces a tab delimited file</p></li>
<li><p>CNVkit rules output genemetrics and gene break file</p></li>
<li><p>Added reference genome to be able to calculate AT/CG dropouts by Picard</p></li>
<li><p>coverage plot plugin part of issue #75</p></li>
<li><p>callable regions for CNV calling of tumor-only</p></li>
</ul>
</section>
<section id="id214">
<h3>Changed<a class="headerlink" href="#id214" title="Permalink to this heading">#</a></h3>
<ul class="simple">
<li><p>Increased time for indel realigner and base recalib rules</p></li>
<li><p>decoupled vep stat from vep main rule</p></li>
<li><p>changed qsub command to match UGE</p></li>
<li><p>scout plugin updated</p></li>
</ul>
</section>
<section id="id215">
<h3>Fixed<a class="headerlink" href="#id215" title="Permalink to this heading">#</a></h3>
<ul class="simple">
<li><p>WGS qc rules - updated with correct options
(picard - CollectMultipleMetrics, sentieon - CoverageMetrics)</p></li>
<li><p>Log warning if WES workflow cannot find SENTIEON* env variables</p></li>
<li><p>Fixes issue with cnvkit and WGS samples #268</p></li>
<li><p>Fix #267 coverage issue with long deletions in vardict</p></li>
</ul>
</section>
</section>
<section id="id216">
<h2>[4.0.1] - 2019-11-08<a class="headerlink" href="#id216" title="Permalink to this heading">#</a></h2>
<section id="id217">
<h3>Added<a class="headerlink" href="#id217" title="Permalink to this heading">#</a></h3>
<ul class="simple">
<li><p>dependencies for workflow report</p></li>
<li><p>sentieon variant callers germline and somatic for wes cases</p></li>
</ul>
</section>
<section id="id218">
<h3>Changed<a class="headerlink" href="#id218" title="Permalink to this heading">#</a></h3>
<ul class="simple">
<li><p>housekeeper file path changed from basename to absolute</p></li>
<li><p>scout template for sample location changed from delivery_report to scout</p></li>
<li><p>rule names added to benchmark files</p></li>
</ul>
</section>
</section>
<section id="id219">
<h2>[4.0.0] - 2019-11-04<a class="headerlink" href="#id219" title="Permalink to this heading">#</a></h2>
<p>SGE qsub support release</p>
<section id="id220">
<h3>Added<a class="headerlink" href="#id220" title="Permalink to this heading">#</a></h3>
<ul class="simple">
<li><p><code class="docutils literal notranslate"><span class="pre">install.sh</span></code> now also downloads latest container</p></li>
<li><p>Docker image for balsamic as part of ci</p></li>
<li><p>Support for qsub alongside with slurm on <code class="docutils literal notranslate"><span class="pre">run</span> <span class="pre">analysis</span> <span class="pre">--profile</span></code></p></li>
</ul>
</section>
<section id="id221">
<h3>Changed<a class="headerlink" href="#id221" title="Permalink to this heading">#</a></h3>
<ul class="simple">
<li><p>Documentation updated</p></li>
<li><p>Test fastq data and test panel bed file with real but dummy data</p></li>
</ul>
</section>
</section>
<section id="id222">
<h2>[3.3.1] - 2019-10-28<a class="headerlink" href="#id222" title="Permalink to this heading">#</a></h2>
<section id="id223">
<h3>Fixed<a class="headerlink" href="#id223" title="Permalink to this heading">#</a></h3>
<ul class="simple">
<li><p>Various links for reference genome is updated with working URL</p></li>
<li><p>Config reference command now print correct output file</p></li>
</ul>
</section>
</section>
<section id="id224">
<h2>[3.3.0] - 2019-10-24<a class="headerlink" href="#id224" title="Permalink to this heading">#</a></h2>
<p>somatic vcfmerge release</p>
<section id="id225">
<h3>Added<a class="headerlink" href="#id225" title="Permalink to this heading">#</a></h3>
<ul class="simple">
<li><p>QC metrics for WGS workflow</p></li>
<li><p>refGene.txt download to reference.json and reference workflow</p></li>
<li><p>A new conda environment within container</p></li>
<li><p>A new base container built via Docker (centos7:miniconda3_4_6_14)</p></li>
<li><p>VCFmerge package as VCF merge rule (<a class="reference external" href="https://github.com/hassanfa/VCFmerge">https://github.com/hassanfa/VCFmerge</a>)</p></li>
<li><p>A container for develop branch</p></li>
<li><p>Benchmark rules to variant callers</p></li>
</ul>
</section>
<section id="id226">
<h3>Changed<a class="headerlink" href="#id226" title="Permalink to this heading">#</a></h3>
<ul class="simple">
<li><p>SLURM resource allocation for various variancalling rules optimized</p></li>
<li><p>mergetype rule updated and only accepts one single tumor instead of multiple</p></li>
</ul>
</section>
</section>
<section id="id227">
<h2>[3.2.3] - 2019-10-24<a class="headerlink" href="#id227" title="Permalink to this heading">#</a></h2>
<section id="id228">
<h3>Fixed<a class="headerlink" href="#id228" title="Permalink to this heading">#</a></h3>
<ul class="simple">
<li><p>Removed unused output files from cnvkit which caused to fail on targetted analysis</p></li>
</ul>
</section>
</section>
<section id="id229">
<h2>[3.2.2] - 2019-10-23<a class="headerlink" href="#id229" title="Permalink to this heading">#</a></h2>
<section id="id230">
<h3>Fixed<a class="headerlink" href="#id230" title="Permalink to this heading">#</a></h3>
<ul class="simple">
<li><p>Removed target file from cnvkit batch</p></li>
</ul>
</section>
</section>
<section id="id231">
<h2>[3.2.1] - 2019-10-23<a class="headerlink" href="#id231" title="Permalink to this heading">#</a></h2>
<section id="id232">
<h3>Fixed<a class="headerlink" href="#id232" title="Permalink to this heading">#</a></h3>
<ul class="simple">
<li><p>CNVkit single missing reference file added</p></li>
</ul>
</section>
</section>
<section id="id233">
<h2>[3.2.0] - 2019-10-11<a class="headerlink" href="#id233" title="Permalink to this heading">#</a></h2>
<section id="adds">
<h3>Adds:<a class="headerlink" href="#adds" title="Permalink to this heading">#</a></h3>
<ul class="simple">
<li><p>CNVkit to WGS workflow</p></li>
<li><p>get_thread for runs</p></li>
</ul>
</section>
<section id="id234">
<h3>Changed:<a class="headerlink" href="#id234" title="Permalink to this heading">#</a></h3>
<ul class="simple">
<li><p>Optimized resources for SLURM jobs</p></li>
</ul>
</section>
<section id="id235">
<h3>Removed:<a class="headerlink" href="#id235" title="Permalink to this heading">#</a></h3>
<ul class="simple">
<li><p>Removed hsmetrics for non-mark duplicate bam files</p></li>
</ul>
</section>
</section>
<section id="id236">
<h2>[3.1.4] - 2019-10-08<a class="headerlink" href="#id236" title="Permalink to this heading">#</a></h2>
<section id="id237">
<h3>Fixed<a class="headerlink" href="#id237" title="Permalink to this heading">#</a></h3>
<ul class="simple">
<li><p>Fixes a bug where missing capture kit bed file error for WGS cases</p></li>
</ul>
</section>
</section>
<section id="id238">
<h2>[3.1.3] - 2019-10-07<a class="headerlink" href="#id238" title="Permalink to this heading">#</a></h2>
<section id="id239">
<h3>Fixed<a class="headerlink" href="#id239" title="Permalink to this heading">#</a></h3>
<ul class="simple">
<li><p>benchmark path bug issue #221</p></li>
</ul>
</section>
</section>
<section id="id240">
<h2>[3.1.2] - 2019-10-07<a class="headerlink" href="#id240" title="Permalink to this heading">#</a></h2>
<section id="id241">
<h3>Fixed<a class="headerlink" href="#id241" title="Permalink to this heading">#</a></h3>
<ul class="simple">
<li><p>libreadline.so.6 symlinking and proper centos version for container</p></li>
</ul>
</section>
</section>
<section id="id242">
<h2>[3.1.1] - 2019-10-03<a class="headerlink" href="#id242" title="Permalink to this heading">#</a></h2>
<section id="id243">
<h3>Fixed<a class="headerlink" href="#id243" title="Permalink to this heading">#</a></h3>
<ul class="simple">
<li><p>Proper tag retrieval for release
### Changed</p></li>
<li><p>BALSAMIC container change to latest and version added to help line</p></li>
</ul>
</section>
</section>
<section id="id244">
<h2>[3.1.0] - 2019-10-03<a class="headerlink" href="#id244" title="Permalink to this heading">#</a></h2>
<p>TL;DR:</p>
<ul class="simple">
<li><p>QoL changes to WGS workflow</p></li>
<li><p>Simplified installation by moving all tools to a container</p></li>
</ul>
<section id="id245">
<h3>Added<a class="headerlink" href="#id245" title="Permalink to this heading">#</a></h3>
<ul class="simple">
<li><p>Benchmarking using psutil</p></li>
<li><p>ML variant calling for WGS</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">--singularity</span></code> option to <code class="docutils literal notranslate"><span class="pre">config</span> <span class="pre">case</span></code> and <code class="docutils literal notranslate"><span class="pre">config</span> <span class="pre">reference</span></code></p></li>
</ul>
</section>
<section id="id246">
<h3>Fixed<a class="headerlink" href="#id246" title="Permalink to this heading">#</a></h3>
<ul class="simple">
<li><p>Fixed a bug with boolean values in analysis.json</p></li>
</ul>
</section>
<section id="id247">
<h3>Changed<a class="headerlink" href="#id247" title="Permalink to this heading">#</a></h3>
<ul class="simple">
<li><p><code class="docutils literal notranslate"><span class="pre">install.sh</span></code> simplified and will be depricated</p></li>
<li><p>Singularity container updated</p></li>
<li><p>Common somatic and germline variant callers are put in single file</p></li>
<li><p>Variant calling workflow and analysis config files merged together</p></li>
</ul>
</section>
<section id="id248">
<h3>Removed<a class="headerlink" href="#id248" title="Permalink to this heading">#</a></h3>
<ul class="simple">
<li><p><code class="docutils literal notranslate"><span class="pre">balsamic</span> <span class="pre">install</span></code> is removed</p></li>
<li><p>Conda environments for py36 and py27 are removed</p></li>
</ul>
</section>
</section>
<section id="id249">
<h2>[3.0.1] - 2019-09-11<a class="headerlink" href="#id249" title="Permalink to this heading">#</a></h2>
<section id="id250">
<h3>Fixed<a class="headerlink" href="#id250" title="Permalink to this heading">#</a></h3>
<ul class="simple">
<li><p>Permissions on <code class="docutils literal notranslate"><span class="pre">analysis/qc</span></code> dir are 777 now</p></li>
</ul>
</section>
</section>
<section id="id251">
<h2>[3.0.0] - 2019-09-05<a class="headerlink" href="#id251" title="Permalink to this heading">#</a></h2>
<p>This is major release.
TL;DR:</p>
<ul class="simple">
<li><p>Major changes to CLI. See documentation for updates.</p></li>
<li><p>New additions to reference generation and reference config file generation and complete overhaul</p></li>
<li><p>Major changes to reposityory structure, conda environments.</p></li>
</ul>
<section id="id252">
<h3>Added<a class="headerlink" href="#id252" title="Permalink to this heading">#</a></h3>
<ul class="simple">
<li><p>Creating and downloading reference files: <code class="docutils literal notranslate"><span class="pre">balsamic</span> <span class="pre">config</span> <span class="pre">reference</span></code> and <code class="docutils literal notranslate"><span class="pre">balsamic</span> <span class="pre">run</span> <span class="pre">reference</span></code></p></li>
<li><p>Container definitions for install and running BALSAMIC</p></li>
<li><p>Bunch of tests, setup coveralls and travis.</p></li>
<li><p>Added Mutliqc, fastp to rule utilities</p></li>
<li><p>Create Housekeeper and Scout files after analysis completes</p></li>
<li><p>Added Sentieon tumor-normal and tumor only workflows</p></li>
<li><p>Added trimming option while creating workflow</p></li>
<li><p>Added multiple tumor sample QC analysis</p></li>
<li><p>Added pindle for indel variant calling</p></li>
<li><p>Added Analysis finish file in the analysis directory</p></li>
</ul>
</section>
<section id="id253">
<h3>Fixed<a class="headerlink" href="#id253" title="Permalink to this heading">#</a></h3>
<ul class="simple">
<li><p>Multiple fixes to snakemake rules</p></li>
</ul>
</section>
<section id="id254">
<h3>Changed<a class="headerlink" href="#id254" title="Permalink to this heading">#</a></h3>
<ul class="simple">
<li><p>Running analysis through: <code class="docutils literal notranslate"><span class="pre">balsamic</span> <span class="pre">run</span> <span class="pre">analysis</span></code></p></li>
<li><p>Cluster account and email info added to <code class="docutils literal notranslate"><span class="pre">balsamic</span> <span class="pre">run</span> <span class="pre">analysis</span></code></p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">umi</span></code> workflow through <code class="docutils literal notranslate"><span class="pre">--umi</span></code> tag. [workflow still in evaluation]</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">sample-id</span></code> replaced by <code class="docutils literal notranslate"><span class="pre">case-id</span></code></p></li>
<li><p>Plan to remove FastQC as well</p></li>
</ul>
</section>
<section id="id255">
<h3>Removed<a class="headerlink" href="#id255" title="Permalink to this heading">#</a></h3>
<ul class="simple">
<li><p><code class="docutils literal notranslate"><span class="pre">balsamic</span> <span class="pre">config</span> <span class="pre">report</span></code> and <code class="docutils literal notranslate"><span class="pre">balsamic</span> <span class="pre">report</span></code></p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">sample.config</span></code> and <code class="docutils literal notranslate"><span class="pre">reference.json</span></code> from config directory</p></li>
<li><p>Removed cutadapt from workflows</p></li>
</ul>
</section>
</section>
<section id="id256">
<h2>[2.9.8] - 2019-01-01<a class="headerlink" href="#id256" title="Permalink to this heading">#</a></h2>
<section id="id257">
<h3>Fixed<a class="headerlink" href="#id257" title="Permalink to this heading">#</a></h3>
<ul class="simple">
<li><p>picard hsmetrics now has 50000 cov max</p></li>
<li><p>cnvkit single wildcard resolve bug fixed</p></li>
</ul>
</section>
</section>
<section id="id258">
<h2>[2.9.7] - 2019-02-28<a class="headerlink" href="#id258" title="Permalink to this heading">#</a></h2>
<section id="id259">
<h3>Fixed<a class="headerlink" href="#id259" title="Permalink to this heading">#</a></h3>
<ul class="simple">
<li><p>Various fixes to umi_single mode</p></li>
<li><p>analysis_finish file does not block reruns anymore</p></li>
<li><p>Added missing single_umi to analysis workflow cli</p></li>
</ul>
</section>
<section id="id260">
<h3>Changed<a class="headerlink" href="#id260" title="Permalink to this heading">#</a></h3>
<ul class="simple">
<li><p>vardict in single mode has lower AF threshold filter (0.005 -&gt; 0.001)</p></li>
</ul>
</section>
</section>
<section id="id261">
<h2>[2.9.6] - 2019-02-25<a class="headerlink" href="#id261" title="Permalink to this heading">#</a></h2>
<section id="id262">
<h3>Fixed<a class="headerlink" href="#id262" title="Permalink to this heading">#</a></h3>
<ul class="simple">
<li><p>Reference to issue #141, fix for 3 other workflows</p></li>
<li><p>CNVkit rule update for refflat file</p></li>
</ul>
</section>
</section>
<section id="id263">
<h2>[2.9.5] - 2019-02-25<a class="headerlink" href="#id263" title="Permalink to this heading">#</a></h2>
<section id="id264">
<h3>Added<a class="headerlink" href="#id264" title="Permalink to this heading">#</a></h3>
<ul class="simple">
<li><p>An analysis finish file is generated with date and time inside (%Y-%M-%d T%T %:z)</p></li>
</ul>
</section>
</section>
<section id="id265">
<h2>[2.9.4] - 2019-02-13<a class="headerlink" href="#id265" title="Permalink to this heading">#</a></h2>
<section id="id266">
<h3>Fixed<a class="headerlink" href="#id266" title="Permalink to this heading">#</a></h3>
<ul class="simple">
<li><p>picard version update to 2.18.11 github.com/hassanfa/picard</p></li>
</ul>
</section>
</section>
<section id="id267">
<h2>[2.9.3] - 2019-02-12<a class="headerlink" href="#id267" title="Permalink to this heading">#</a></h2>
<section id="id268">
<h3>Fixed<a class="headerlink" href="#id268" title="Permalink to this heading">#</a></h3>
<ul class="simple">
<li><p>Mutect single mode table generation fix</p></li>
<li><p>Vardict single mode MVL annotation fix</p></li>
</ul>
</section>
</section>
<section id="id269">
<h2>[2.9.2] - 2019-02-04<a class="headerlink" href="#id269" title="Permalink to this heading">#</a></h2>
<section id="id270">
<h3>Added<a class="headerlink" href="#id270" title="Permalink to this heading">#</a></h3>
<ul class="simple">
<li><p>CNVkit single sample mode now in workflow</p></li>
<li><p>MVL list from cheng et al. 2015 moved to assets</p></li>
</ul>
</section>
</section>
<section id="id271">
<h2>[2.9.1] - 2019-01-22<a class="headerlink" href="#id271" title="Permalink to this heading">#</a></h2>
<section id="id272">
<h3>Added<a class="headerlink" href="#id272" title="Permalink to this heading">#</a></h3>
<ul class="simple">
<li><p>Simple table for somatic variant callers for single sample mode added</p></li>
</ul>
</section>
<section id="id273">
<h3>Fixed<a class="headerlink" href="#id273" title="Permalink to this heading">#</a></h3>
<ul class="simple">
<li><p>Fixes an issue with conda that unset variables threw an error issue #141</p></li>
</ul>
</section>
</section>
<section id="id274">
<h2>[2.9.0] - 2019-01-04<a class="headerlink" href="#id274" title="Permalink to this heading">#</a></h2>
<section id="id275">
<h3>Changed<a class="headerlink" href="#id275" title="Permalink to this heading">#</a></h3>
<ul class="simple">
<li><p>Readme structure and example</p></li>
<li><p>Mutect2’s single sample output is similar to paired now</p></li>
<li><p>cli path structure update</p></li>
</ul>
</section>
<section id="id276">
<h3>Added<a class="headerlink" href="#id276" title="Permalink to this heading">#</a></h3>
<ul class="simple">
<li><p>test data and sample inputs</p></li>
<li><p>A dag PDF will be generated when config is made</p></li>
<li><p>umi specific variant calling</p></li>
</ul>
</section>
</section>
<section id="id277">
<h2>[2.8.1] - 2018-11-28<a class="headerlink" href="#id277" title="Permalink to this heading">#</a></h2>
<section id="id278">
<h3>Fixed<a class="headerlink" href="#id278" title="Permalink to this heading">#</a></h3>
<ul class="simple">
<li><p>VEP’s perl module errors</p></li>
<li><p>CoverageRep.R now properly takes protein_coding transcatipts only</p></li>
</ul>
</section>
</section>
<section id="id279">
<h2>[2.8.0] - 2018-11-23<a class="headerlink" href="#id279" title="Permalink to this heading">#</a></h2>
<p>UMI single sample align and QC</p>
<section id="id280">
<h3>Added<a class="headerlink" href="#id280" title="Permalink to this heading">#</a></h3>
<ul class="simple">
<li><p>Added rules and workflows for UMI analysis: QC and alignment</p></li>
</ul>
</section>
</section>
<section id="id281">
<h2>[2.7.4] - 2018-11-23<a class="headerlink" href="#id281" title="Permalink to this heading">#</a></h2>
<p>Germline single sample</p>
<section id="id282">
<h3>Added<a class="headerlink" href="#id282" title="Permalink to this heading">#</a></h3>
<ul class="simple">
<li><p>Germline single sample addition
### Changed</p></li>
<li><p>Minor fixes to some rules to make them compatible with tumor mode</p></li>
</ul>
</section>
</section>
<section id="id283">
<h2>[2.7.3] - 2018-11-20<a class="headerlink" href="#id283" title="Permalink to this heading">#</a></h2>
<section id="id284">
<h3>Fixed<a class="headerlink" href="#id284" title="Permalink to this heading">#</a></h3>
<ul class="simple">
<li><p>Various bugs with DAG to keep popvcf and splitbed depending on merge bam file</p></li>
<li><p>install script script fixed and help added</p></li>
</ul>
</section>
</section>
<section id="id285">
<h2>[2.7.2] - 2018-11-15<a class="headerlink" href="#id285" title="Permalink to this heading">#</a></h2>
<section id="id286">
<h3>Changed<a class="headerlink" href="#id286" title="Permalink to this heading">#</a></h3>
<ul class="simple">
<li><p>Vardict, Strelka, and Manta separated from GATK best practice pipeline</p></li>
</ul>
</section>
</section>
<section id="id287">
<h2>[2.7.1] - 2018-11-13<a class="headerlink" href="#id287" title="Permalink to this heading">#</a></h2>
<section id="id288">
<h3>Fixed<a class="headerlink" href="#id288" title="Permalink to this heading">#</a></h3>
<ul class="simple">
<li><p>minro bugs with strelka_germline and freebayes merge
### Changed</p></li>
<li><p>removed ERC from haplotypecaller</p></li>
</ul>
</section>
</section>
<section id="id289">
<h2>[2.7.0] - 2018-11-08<a class="headerlink" href="#id289" title="Permalink to this heading">#</a></h2>
<p>Germline patch</p>
<section id="id290">
<h3>Added<a class="headerlink" href="#id290" title="Permalink to this heading">#</a></h3>
<ul class="simple">
<li><p>Germline caller tested and added to the paired analysis workflow: Freebayes, HaplotypeCaller, Strelka, Manta</p></li>
</ul>
</section>
<section id="id291">
<h3>Changed<a class="headerlink" href="#id291" title="Permalink to this heading">#</a></h3>
<ul class="simple">
<li><p>Analysis config files updated</p></li>
<li><p>Output directory structure changed</p></li>
<li><p>vep rule is now a single rule</p></li>
<li><p>Bunch of rule names updated and shortened, specifically in Picard and GATK</p></li>
<li><p>Variant caller rules are all updated and changed</p></li>
<li><p>output vcf file names are now more sensible: {SNV,SV}.{somatic,germline}.sampleId.variantCaller.vcf.gz</p></li>
<li><p>Job limit increased to 300</p></li>
</ul>
</section>
<section id="id292">
<h3>Removed<a class="headerlink" href="#id292" title="Permalink to this heading">#</a></h3>
<ul class="simple">
<li><p>removed bcftools.rule for var id annotation</p></li>
</ul>
</section>
<section id="id293">
<h3>Changed<a class="headerlink" href="#id293" title="Permalink to this heading">#</a></h3>
</section>
<section id="id294">
<h3>Fixed<a class="headerlink" href="#id294" title="Permalink to this heading">#</a></h3>
</section>
</section>
<section id="id295">
<h2>[2.6.3] - 2018-11-01<a class="headerlink" href="#id295" title="Permalink to this heading">#</a></h2>
<section id="id296">
<h3>Changed<a class="headerlink" href="#id296" title="Permalink to this heading">#</a></h3>
<ul class="simple">
<li><p>Ugly and godforsaken <code class="docutils literal notranslate"><span class="pre">runSbatch.py</span></code> is now dumping sacct files with job IDs. Yikes!</p></li>
</ul>
</section>
</section>
<section id="id297">
<h2>[2.6.2] - 2018-10-31<a class="headerlink" href="#id297" title="Permalink to this heading">#</a></h2>
<section id="id298">
<h3>Fixed<a class="headerlink" href="#id298" title="Permalink to this heading">#</a></h3>
<ul class="simple">
<li><p>added <code class="docutils literal notranslate"><span class="pre">--fastq-prefix</span></code> option for <code class="docutils literal notranslate"><span class="pre">config</span> <span class="pre">sample</span></code> to set fastq prefix name. Linking is not changed.</p></li>
</ul>
</section>
</section>
<section id="id299">
<h2>[2.6.1] - 2018-10-29<a class="headerlink" href="#id299" title="Permalink to this heading">#</a></h2>
<section id="id300">
<h3>Fixed<a class="headerlink" href="#id300" title="Permalink to this heading">#</a></h3>
<ul class="simple">
<li><p>patched a bug for copying results for strelka and manta which was introduced in <code class="docutils literal notranslate"><span class="pre">2.5.0</span></code></p></li>
</ul>
</section>
</section>
<section id="id301">
<h2>[2.5.0] - 2018-10-22<a class="headerlink" href="#id301" title="Permalink to this heading">#</a></h2>
<section id="id302">
<h3>Changed<a class="headerlink" href="#id302" title="Permalink to this heading">#</a></h3>
<ul class="simple">
<li><p><code class="docutils literal notranslate"><span class="pre">variant_panel</span></code> changed to <code class="docutils literal notranslate"><span class="pre">capture_kit</span></code></p></li>
<li><p>sample config file takes balsamic version</p></li>
<li><p>bioinfo tool config moved bioinfotool to cli_utils from <code class="docutils literal notranslate"><span class="pre">config</span> <span class="pre">report</span></code></p></li>
</ul>
</section>
<section id="id303">
<h3>Added<a class="headerlink" href="#id303" title="Permalink to this heading">#</a></h3>
<ul class="simple">
<li><p>bioinfo tool versions is now added to analysis config file</p></li>
</ul>
</section>
</section>
<section id="id304">
<h2>[2.4.0] - 2018-10-22<a class="headerlink" href="#id304" title="Permalink to this heading">#</a></h2>
<section id="id305">
<h3>Changed<a class="headerlink" href="#id305" title="Permalink to this heading">#</a></h3>
<ul class="simple">
<li><p><code class="docutils literal notranslate"><span class="pre">balsamic</span> <span class="pre">run</span></code> has 3 stop points: paired variant calling, single mode variant calling, and QC/Alignment mode.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">balsamic</span> <span class="pre">run</span> <span class="pre">[OPTIONS]</span> <span class="pre">-S</span> <span class="pre">...</span></code> is depricated, but it supersedes <code class="docutils literal notranslate"><span class="pre">analysis_type</span></code> mode if provided.</p></li>
</ul>
</section>
</section>
<section id="id306">
<h2>[2.3.3] - 2018-10-22<a class="headerlink" href="#id306" title="Permalink to this heading">#</a></h2>
<section id="id307">
<h3>Added<a class="headerlink" href="#id307" title="Permalink to this heading">#</a></h3>
<ul class="simple">
<li><p>CSV output for variants in each variant caller based on variant filters</p></li>
<li><p>DAG image of workflow
### Changed</p></li>
<li><p>Input for variant filter has a default value</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">delivery_report</span></code> is no created during config generation</p></li>
<li><p>Variant reporter R script cmd updated in <code class="docutils literal notranslate"><span class="pre">balsamic</span> <span class="pre">report</span></code></p></li>
</ul>
</section>
</section>
<section id="id308">
<h2>[2.3.2] - 2018-10-19<a class="headerlink" href="#id308" title="Permalink to this heading">#</a></h2>
<section id="id309">
<h3>Changed<a class="headerlink" href="#id309" title="Permalink to this heading">#</a></h3>
<ul class="simple">
<li><p>Fastq files are now always linked to <code class="docutils literal notranslate"><span class="pre">fastq</span></code> directory within the analysis directory</p></li>
</ul>
</section>
<section id="id310">
<h3>Added<a class="headerlink" href="#id310" title="Permalink to this heading">#</a></h3>
<ul class="simple">
<li><p><code class="docutils literal notranslate"><span class="pre">balsamic</span> <span class="pre">config</span> <span class="pre">sample</span></code> now accepts individual files and paths. See README for usage.</p></li>
</ul>
</section>
</section>
<section id="id311">
<h2>[2.3.1] - 2018-09-25<a class="headerlink" href="#id311" title="Permalink to this heading">#</a></h2>
<section id="id312">
<h3>Added<a class="headerlink" href="#id312" title="Permalink to this heading">#</a></h3>
<ul class="simple">
<li><p>CollectHSmetric now run twice for before and after markduplicate</p></li>
</ul>
</section>
</section>
<section id="id313">
<h2>[2.3.0] - 2018-09-25<a class="headerlink" href="#id313" title="Permalink to this heading">#</a></h2>
<section id="id314">
<h3>Changed<a class="headerlink" href="#id314" title="Permalink to this heading">#</a></h3>
<ul class="simple">
<li><p>Sample config file now includes a list of chromosomes in the panel bed file</p></li>
</ul>
</section>
<section id="id315">
<h3>Fixed<a class="headerlink" href="#id315" title="Permalink to this heading">#</a></h3>
<ul class="simple">
<li><p>Non-matching chrom won’t break the splitbed rule anymore</p></li>
<li><p>collectqc rules now properly parse tab delimited metric files</p></li>
</ul>
</section>
</section>
<section id="id316">
<h2>[2.2.0] - 2018-09-11<a class="headerlink" href="#id316" title="Permalink to this heading">#</a></h2>
<section id="id317">
<h3>Added<a class="headerlink" href="#id317" title="Permalink to this heading">#</a></h3>
<ul class="simple">
<li><p>Coverage plot to report</p></li>
<li><p>target coverage file to report json</p></li>
<li><p>post-cutadapt fastqc to collectqc</p></li>
<li><p>A header to report pdf</p></li>
<li><p>list of bioinfo tools used in the analysis added to report
### Changed</p></li>
<li><p>VariantRep.R now accepts multiple inputs for each parameter (see help)</p></li>
<li><p>AF values for MSKIMPACT config
### Fixed</p></li>
<li><p>Output figure for coverageplot is now fully square :-)</p></li>
</ul>
</section>
</section>
<section id="id318">
<h2>[2.1.0] - 2018-09-11<a class="headerlink" href="#id318" title="Permalink to this heading">#</a></h2>
<section id="id319">
<h3>Added<a class="headerlink" href="#id319" title="Permalink to this heading">#</a></h3>
<ul class="simple">
<li><p>normalized coverage plot script</p></li>
<li><p>fastq file IO check for config creation</p></li>
<li><p>added qos option to <code class="docutils literal notranslate"><span class="pre">balsamic</span> <span class="pre">run</span></code>
### Fixed</p></li>
<li><p>Sambamba depth coverage parameters</p></li>
<li><p>bug with picard markduplicate flag</p></li>
</ul>
</section>
</section>
<section id="id320">
<h2>[2.0.2] - 2018-09-11<a class="headerlink" href="#id320" title="Permalink to this heading">#</a></h2>
<section id="id321">
<h3>Added<a class="headerlink" href="#id321" title="Permalink to this heading">#</a></h3>
<ul class="simple">
<li><p>Added qos option for setting qos to run jobs with a default value of low</p></li>
</ul>
</section>
</section>
<section id="id322">
<h2>[2.0.1] - 2018-09-10<a class="headerlink" href="#id322" title="Permalink to this heading">#</a></h2>
<section id="id323">
<h3>Fixed<a class="headerlink" href="#id323" title="Permalink to this heading">#</a></h3>
<ul class="simple">
<li><p>Fixed package dependencies with vep and installation</p></li>
</ul>
</section>
</section>
<section id="id324">
<h2>[2.0.0] - 2018-09-05<a class="headerlink" href="#id324" title="Permalink to this heading">#</a></h2>
<p>Variant reporter patch and cli update</p>
<section id="id325">
<h3>Added<a class="headerlink" href="#id325" title="Permalink to this heading">#</a></h3>
<ul class="simple">
<li><p>Added <code class="docutils literal notranslate"><span class="pre">balsamic</span> <span class="pre">config</span> <span class="pre">sample</span></code> and <code class="docutils literal notranslate"><span class="pre">balsamic</span> <span class="pre">config</span> <span class="pre">report</span></code> to generate run analysis and reporting config</p></li>
<li><p>Added <code class="docutils literal notranslate"><span class="pre">VariantRep.R</span></code> script to information from merged variant table: variant summry, TMB, and much more</p></li>
<li><p>Added a workflow for single sample mode alignment and QC only</p></li>
<li><p>Added QC skimming script to qccollect to generate nicely formatted information from picard
### Changed</p></li>
<li><p>Change to CLI for running and creating config</p></li>
<li><p>Major overhaul to coverage report script. It’s now simpler and more readable!
### Fixed</p></li>
<li><p>Fixed sambamba depth to include mapping quality</p></li>
<li><p>Markduplicate now is now by default on marking mode, and will NOT remove duplicates</p></li>
<li><p>Minor formatting and script beautification happened</p></li>
</ul>
</section>
</section>
<section id="id326">
<h2>[1.13.1] - 2018-08-17<a class="headerlink" href="#id326" title="Permalink to this heading">#</a></h2>
<section id="id327">
<h3>Fixed<a class="headerlink" href="#id327" title="Permalink to this heading">#</a></h3>
<ul class="simple">
<li><p>fixed a typo in MSKMVL config</p></li>
<li><p>fixed a bug in strelka_simple for correct column orders</p></li>
</ul>
</section>
</section>
<section id="id328">
<h2>[1.13.0] - 2018-08-10<a class="headerlink" href="#id328" title="Permalink to this heading">#</a></h2>
<section id="id329">
<h3>Added<a class="headerlink" href="#id329" title="Permalink to this heading">#</a></h3>
<ul class="simple">
<li><p>rule for all three variant callers for paired analysis now generate a simple VCF file</p></li>
<li><p>rule for all three variant callers for paired analysis to convert VCF into table format</p></li>
<li><p>MVL config file and MVL annotation to VCF calls for SNV/INDEL callers</p></li>
<li><p>CALLER annotation added to SNV/INDEL callers</p></li>
<li><p>exome specific option for strelka paired</p></li>
<li><p>create_config subcommand is now more granular, it accepts all enteries from sample.json as commandline arguments</p></li>
<li><p>Added tabQuery to the assets as a tool to query the tabulated output of summarized VCF</p></li>
<li><p>Added MQ annotation field to Mutect2 output see #67
### Changed</p></li>
<li><p>Leaner VCF output from mutect2 with coverage and MQ annotation according to #64</p></li>
<li><p>variant ids are now updated from simple VCF file
### Fixed</p></li>
<li><p>Fixed a bug with sambamba depth coverage reporting wrong exon and panel coverage see #68</p></li>
<li><p>The json output is now properly formatted using yapf</p></li>
<li><p>Strelka rule doesn’t filter out PASS variants anymore fixes issue #63</p></li>
</ul>
</section>
</section>
<section id="id330">
<h2>[1.12.0] - 2018-07-06<a class="headerlink" href="#id330" title="Permalink to this heading">#</a></h2>
<p>Coverage report patch</p>
<section id="id331">
<h3>Added<a class="headerlink" href="#id331" title="Permalink to this heading">#</a></h3>
<ul class="simple">
<li><p>Added a new script to retrieve coverage report for a list of gene(s) and transcripts(s)</p></li>
<li><p>Added sambamba exon depth rule for coverage report</p></li>
<li><p>Added a new entry in reference json for exon bed file, this file generated using: <a class="reference external" href="https://github.com/hassanfa/GFFtoolkit">https://github.com/hassanfa/GFFtoolkit</a>
### Changed</p></li>
<li><p>sambamba_depth rule changed to sambama_panel_depth</p></li>
<li><p>sambamba depth now has fix-mate-overlaps parameter enabled</p></li>
<li><p>sambamba string filter changed to <code class="docutils literal notranslate"><span class="pre">unmapped</span> <span class="pre">or</span> <span class="pre">mate\_is\_unmapped)</span> <span class="pre">and</span> <span class="pre">not</span> <span class="pre">duplicate</span> <span class="pre">and</span> <span class="pre">not</span> <span class="pre">failed\_quality\_control</span></code>.</p></li>
<li><p>sambamba depth for both panel and exon work on picard flag (rmdup or mrkdup).
### Fixed</p></li>
<li><p>Fixed sambamba panel depth rule for redundant coverage parameter</p></li>
</ul>
</section>
</section>
<section id="id332">
<h2>[1.11.0] - 2018-07-05<a class="headerlink" href="#id332" title="Permalink to this heading">#</a></h2>
<p>create config patch for single and paired mode</p>
<section id="id333">
<h3>Changed<a class="headerlink" href="#id333" title="Permalink to this heading">#</a></h3>
<ul class="simple">
<li><p>create_config is now accepting a paired|single mode instead of analysis json template (see help for changes). It is
not backward compatible
### Added</p></li>
<li><p>analysis_{paired single}.json for creating config. Analysis.json is now obsolete.
### Fixed</p></li>
<li><p>A bug with writing output for analysis config, and creating the path if it doesn’t exist.</p></li>
<li><p>A bug with manta rule to correctly set output files in config.</p></li>
<li><p>A bug that strelka was still included in sample analysis.</p></li>
</ul>
</section>
</section>
<section id="id334">
<h2>[1.10.0] - 2018-06-07<a class="headerlink" href="#id334" title="Permalink to this heading">#</a></h2>
<section id="id335">
<h3>Added<a class="headerlink" href="#id335" title="Permalink to this heading">#</a></h3>
<ul class="simple">
<li><p>Markduplicate flag to analysis config</p></li>
</ul>
</section>
</section>
<section id="id336">
<h2>[1.9.0] - 2018-06-04<a class="headerlink" href="#id336" title="Permalink to this heading">#</a></h2>
<section id="id337">
<h3>Added<a class="headerlink" href="#id337" title="Permalink to this heading">#</a></h3>
<ul class="simple">
<li><p>Single mode for vardict, manta, and mutect.</p></li>
<li><p>merge type for tumor only
### Changed</p></li>
<li><p>Single mode variant calling now has all variant calling rules
### Fixed</p></li>
<li><p>run_analaysis now accepts workflows for testing pyrposes</p></li>
</ul>
</section>
</section>
<section id="id338">
<h2>[1.8.0] - 2018-06-01<a class="headerlink" href="#id338" title="Permalink to this heading">#</a></h2>
<section id="id339">
<h3>Changed<a class="headerlink" href="#id339" title="Permalink to this heading">#</a></h3>
<ul class="simple">
<li><p>picard create bed interval rule moved into collect hsmetric</p></li>
<li><p>split bed is dependent on bam merge rule</p></li>
<li><p>vardict env now has specific build rather than URL download (conda doesn’t support URLs anymore)
### Fixed</p></li>
<li><p>new logs and scripts dirs are not re-created if they are empty</p></li>
</ul>
</section>
</section>
<section id="id340">
<h2>[1.7.0] - 2018-05-31<a class="headerlink" href="#id340" title="Permalink to this heading">#</a></h2>
<section id="id341">
<h3>Added<a class="headerlink" href="#id341" title="Permalink to this heading">#</a></h3>
<ul class="simple">
<li><p>A source altered picard to generated more quality metrics output is added to installation and rules</p></li>
</ul>
</section>
</section>
<section id="id342">
<h2>[1.6.0] - 2018-05-30<a class="headerlink" href="#id342" title="Permalink to this heading">#</a></h2>
<section id="id343">
<h3>Added<a class="headerlink" href="#id343" title="Permalink to this heading">#</a></h3>
<ul class="simple">
<li><p>report subcommand for generating a pdf report from a json input file</p></li>
<li><p>Added fastqc after removing adapter
### Changed</p></li>
<li><p>Markduplicate now has both REMOVE and MARK (rmdup vs mrkdup)</p></li>
<li><p>CollectHSMetrics now has more steps on PCT_TARGET_BASES</p></li>
</ul>
</section>
</section>
<section id="id344">
<h2>[1.5.0] - 2018-05-28<a class="headerlink" href="#id344" title="Permalink to this heading">#</a></h2>
<section id="id345">
<h3>Changed<a class="headerlink" href="#id345" title="Permalink to this heading">#</a></h3>
<ul class="simple">
<li><p>New log and script directories are now created for each re-run
### Fixed</p></li>
<li><p>Picardtools’ memory issue addressed for large samples</p></li>
</ul>
</section>
</section>
<section id="id346">
<h2>[1.4.0] - 2018-05-18<a class="headerlink" href="#id346" title="Permalink to this heading">#</a></h2>
<section id="id347">
<h3>Added<a class="headerlink" href="#id347" title="Permalink to this heading">#</a></h3>
<ul class="simple">
<li><p>single sample analysis mode</p></li>
<li><p>alignment and insert size metrics are added to the workflow
### Changed</p></li>
<li><p>collectqc and contest have their own rule for paired (tumor vs normal) and single (tumor only) sample.</p></li>
</ul>
</section>
</section>
<section id="id348">
<h2>[1.3.0] - 2018-05-13<a class="headerlink" href="#id348" title="Permalink to this heading">#</a></h2>
<section id="id349">
<h3>Added<a class="headerlink" href="#id349" title="Permalink to this heading">#</a></h3>
<ul class="simple">
<li><p>bed file for panel analysis is now mandatory to create analaysis config</p></li>
</ul>
</section>
</section>
<section id="id350">
<h2>[1.2.3] - 2018-05-13<a class="headerlink" href="#id350" title="Permalink to this heading">#</a></h2>
<section id="id351">
<h3>Changed<a class="headerlink" href="#id351" title="Permalink to this heading">#</a></h3>
<ul class="simple">
<li><p>vep execution path</p></li>
<li><p>working directory for snakemake</p></li>
</ul>
</section>
</section>
<section id="id352">
<h2>[1.2.2] - 2018-05-04<a class="headerlink" href="#id352" title="Permalink to this heading">#</a></h2>
<section id="id353">
<h3>Added<a class="headerlink" href="#id353" title="Permalink to this heading">#</a></h3>
<ul class="simple">
<li><p>sbatch submitter and cluster config now has an mail field
### Changed</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">create_config</span></code> now only requires sample and output json. The rest are optional</p></li>
</ul>
</section>
</section>
<section id="id354">
<h2>[1.2.0] - 2018-05-02<a class="headerlink" href="#id354" title="Permalink to this heading">#</a></h2>
<section id="id355">
<h3>Added<a class="headerlink" href="#id355" title="Permalink to this heading">#</a></h3>
<ul class="simple">
<li><p>snakefile and cluster config in run analysis are now optional with a default value</p></li>
</ul>
</section>
</section>
<section id="id356">
<h2>[1.1.2] - 2018-04-27<a class="headerlink" href="#id356" title="Permalink to this heading">#</a></h2>
<section id="id357">
<h3>Fixed<a class="headerlink" href="#id357" title="Permalink to this heading">#</a></h3>
<ul class="simple">
<li><p>vardict installation was failing without conda-forge channel</p></li>
<li><p>gatk installation was failing without correct jar file</p></li>
</ul>
</section>
</section>
<section id="id358">
<h2>[1.1.1] - 2018-04-27<a class="headerlink" href="#id358" title="Permalink to this heading">#</a></h2>
<section id="id359">
<h3>Fixed<a class="headerlink" href="#id359" title="Permalink to this heading">#</a></h3>
<ul class="simple">
<li><p>gatk-register tmp directory</p></li>
</ul>
</section>
</section>
<section id="id360">
<h2>[1.1.0] - 2018-04-26<a class="headerlink" href="#id360" title="Permalink to this heading">#</a></h2>
<section id="id361">
<h3>Added<a class="headerlink" href="#id361" title="Permalink to this heading">#</a></h3>
<ul class="simple">
<li><p>create config sub command added as a new feature to create input config file</p></li>
<li><p>templates to generate a config file for analysis added</p></li>
<li><p>code style template for YAPF input created. see: <a class="reference external" href="https://github.com/google/yapf">https://github.com/google/yapf</a></p></li>
<li><p>vt conda env added</p></li>
</ul>
</section>
<section id="id362">
<h3>Changed<a class="headerlink" href="#id362" title="Permalink to this heading">#</a></h3>
<ul class="simple">
<li><p>install script changed to create an output config</p></li>
<li><p>README updated with usage</p></li>
</ul>
</section>
<section id="id363">
<h3>Fixed<a class="headerlink" href="#id363" title="Permalink to this heading">#</a></h3>
<ul class="simple">
<li><p>fastq location for analysis config is now fixed</p></li>
<li><p>lambda rules removed from cutadapt and fastq</p></li>
</ul>
</section>
</section>
<section id="rc2-2018-04-18">
<h2>[1.0.3-rc2] - 2018-04-18<a class="headerlink" href="#rc2-2018-04-18" title="Permalink to this heading">#</a></h2>
<section id="id364">
<h3>Added<a class="headerlink" href="#id364" title="Permalink to this heading">#</a></h3>
<ul class="simple">
<li><p>Added sbatch submitter to handle it outside snakemake
### Changed</p></li>
<li><p>sample config file structure changed</p></li>
<li><p>coding styles updated</p></li>
</ul>
</section>
</section>
<section id="rc2-2018-04-17">
<h2>[1.0.2-rc2] - 2018-04-17<a class="headerlink" href="#rc2-2018-04-17" title="Permalink to this heading">#</a></h2>
<section id="id365">
<h3>Added<a class="headerlink" href="#id365" title="Permalink to this heading">#</a></h3>
<ul class="simple">
<li><p>Added vt environment
### Fixed</p></li>
<li><p>conda envs are now have D prefix instead of P (develop vs production)</p></li>
<li><p>install_conda subcommand now accepts a proper conda prefix</p></li>
</ul>
</section>
</section>
<section id="rc2-2018-04-16">
<h2>[1.0.1-rc2] - 2018-04-16<a class="headerlink" href="#rc2-2018-04-16" title="Permalink to this heading">#</a></h2>
<section id="id366">
<h3>Fixed<a class="headerlink" href="#id366" title="Permalink to this heading">#</a></h3>
<ul class="simple">
<li><p>snakemake rules are now externally linked</p></li>
</ul>
</section>
</section>
<section id="id367">
<h2>[1.0.0-rc2] - 2018-04-16<a class="headerlink" href="#id367" title="Permalink to this heading">#</a></h2>
<section id="id368">
<h3>Added<a class="headerlink" href="#id368" title="Permalink to this heading">#</a></h3>
<ul class="simple">
<li><p>run_analysis subcommand</p></li>
<li><p>Mutational Signature R script with CLI</p></li>
<li><p>unittest to install_conda</p></li>
<li><p>a method to semi-dynamically retrieve suitable conda env for each rule</p></li>
</ul>
</section>
<section id="id369">
<h3>Fixed<a class="headerlink" href="#id369" title="Permalink to this heading">#</a></h3>
<ul class="simple">
<li><p>install.sh updated with gatk and proper log output</p></li>
<li><p>conda environments updated</p></li>
<li><p>vardict now has its own environment and it should not raise anymore errors</p></li>
</ul>
</section>
</section>
<section id="rc1-2018-04-05">
<h2>[1.0.0-rc1] - 2018-04-05<a class="headerlink" href="#rc1-2018-04-05" title="Permalink to this heading">#</a></h2>
<section id="id370">
<h3>Added<a class="headerlink" href="#id370" title="Permalink to this heading">#</a></h3>
<ul class="simple">
<li><p>install.sh to install balsamic</p></li>
<li><p>balsamic barebone cli</p></li>
<li><p>subcommand to install required environments</p></li>
<li><p>README.md updated with basic installation instructions</p></li>
</ul>
</section>
<section id="id371">
<h3>Fixed<a class="headerlink" href="#id371" title="Permalink to this heading">#</a></h3>
<ul class="simple">
<li><p>conda environment yaml files</p></li>
</ul>
</section>
</section>
</section>

        </article>
      </div>
      <footer>
        
        <div class="related-pages">
          <a class="next-page" href="bioinfo_softwares.html">
              <div class="page-info">
                <div class="context">
                  <span>Next</span>
                </div>
                <div class="title">Tools and software</div>
              </div>
              <svg class="furo-related-icon"><use href="#svg-arrow-right"></use></svg>
            </a>
          <a class="prev-page" href="balsamic_methods.html">
              <svg class="furo-related-icon"><use href="#svg-arrow-right"></use></svg>
              <div class="page-info">
                <div class="context">
                  <span>Previous</span>
                </div>
                
                <div class="title">Method description</div>
                
              </div>
            </a>
        </div>
        <div class="bottom-of-page">
          <div class="left-details">
            <div class="copyright">
                Copyright &#169; 2023, Clinical Genomics
            </div>
            Made with <a href="https://www.sphinx-doc.org/">Sphinx</a> and <a class="muted-link" href="https://pradyunsg.me">@pradyunsg</a>'s
            
            <a href="https://github.com/pradyunsg/furo">Furo</a>
            
          </div>
          <div class="right-details">
            
          </div>
        </div>
        
      </footer>
    </div>
    <aside class="toc-drawer">
      
      
      <div class="toc-sticky toc-scroll">
        <div class="toc-title-container">
          <span class="toc-title">
            On this page
          </span>
        </div>
        <div class="toc-tree-container">
          <div class="toc-tree">
            <ul>
<li><a class="reference internal" href="#">Changelog</a><ul>
<li><a class="reference internal" href="#id1">[13.0.0]</a><ul>
<li><a class="reference internal" href="#added">Added:</a></li>
<li><a class="reference internal" href="#changed">Changed:</a></li>
<li><a class="reference internal" href="#fixed">Fixed:</a></li>
<li><a class="reference internal" href="#removed">Removed:</a></li>
</ul>
</li>
<li><a class="reference internal" href="#id2">[12.0.2]</a><ul>
<li><a class="reference internal" href="#id3">Fixed:</a></li>
<li><a class="reference internal" href="#id4">Changed:</a></li>
</ul>
</li>
<li><a class="reference internal" href="#id5">[12.0.1]</a><ul>
<li><a class="reference internal" href="#id6">Added:</a></li>
<li><a class="reference internal" href="#id7">Fixed:</a></li>
</ul>
</li>
<li><a class="reference internal" href="#id8">[12.0.0]</a><ul>
<li><a class="reference internal" href="#id9">Added:</a></li>
<li><a class="reference internal" href="#id10">Changed:</a></li>
<li><a class="reference internal" href="#id11">Fixed:</a></li>
</ul>
</li>
<li><a class="reference internal" href="#id12">[11.2.0]</a><ul>
<li><a class="reference internal" href="#id13">Fixed:</a></li>
</ul>
</li>
<li><a class="reference internal" href="#id14">[11.1.0]</a><ul>
<li><a class="reference internal" href="#id15">Added:</a></li>
<li><a class="reference internal" href="#id16">Changed:</a></li>
<li><a class="reference internal" href="#id17">Fixed:</a></li>
</ul>
</li>
<li><a class="reference internal" href="#id18">[11.0.2]</a><ul>
<li><a class="reference internal" href="#id19">Changed:</a></li>
<li><a class="reference internal" href="#id20">Fixed:</a></li>
</ul>
</li>
<li><a class="reference internal" href="#id21">[11.0.1]</a><ul>
<li><a class="reference internal" href="#id22">Fixed:</a></li>
</ul>
</li>
<li><a class="reference internal" href="#id23">[11.0.0]</a><ul>
<li><a class="reference internal" href="#id24">Added:</a></li>
<li><a class="reference internal" href="#id25">Changed:</a></li>
<li><a class="reference internal" href="#id26">Fixed:</a></li>
<li><a class="reference internal" href="#id27">Removed:</a></li>
</ul>
</li>
<li><a class="reference internal" href="#id28">[10.0.5]</a><ul>
<li><a class="reference internal" href="#id29">Changed:</a></li>
</ul>
</li>
<li><a class="reference internal" href="#id30">[10.0.4]</a><ul>
<li><a class="reference internal" href="#id31">Fixed:</a></li>
</ul>
</li>
<li><a class="reference internal" href="#id32">[10.0.3]</a><ul>
<li><a class="reference internal" href="#id33">Fixed:</a></li>
</ul>
</li>
<li><a class="reference internal" href="#id34">[10.0.2]</a><ul>
<li><a class="reference internal" href="#id35">Fixed:</a></li>
</ul>
</li>
<li><a class="reference internal" href="#id36">[10.0.1]</a><ul>
<li><a class="reference internal" href="#id37">Fixed:</a></li>
<li><a class="reference internal" href="#id38">Changed:</a></li>
</ul>
</li>
<li><a class="reference internal" href="#id39">[10.0.0]</a><ul>
<li><a class="reference internal" href="#id40">Added:</a></li>
<li><a class="reference internal" href="#id41">Changed:</a></li>
<li><a class="reference internal" href="#id42">Fixed:</a></li>
<li><a class="reference internal" href="#id43">Removed</a></li>
</ul>
</li>
<li><a class="reference internal" href="#id44">[9.0.1]</a><ul>
<li><a class="reference internal" href="#id45">Fixed:</a></li>
</ul>
</li>
<li><a class="reference internal" href="#id46">[9.0.0]</a><ul>
<li><a class="reference internal" href="#id47">Added:</a></li>
<li><a class="reference internal" href="#id48">Changed:</a></li>
<li><a class="reference internal" href="#id49">Fixed:</a></li>
<li><a class="reference internal" href="#id50">Removed</a></li>
</ul>
</li>
<li><a class="reference internal" href="#id51">[8.2.10]</a><ul>
<li><a class="reference internal" href="#id52">Added:</a></li>
<li><a class="reference internal" href="#fixes">Fixes:</a></li>
<li><a class="reference internal" href="#id53">Changed:</a></li>
<li><a class="reference internal" href="#id54">Removed:</a></li>
</ul>
</li>
<li><a class="reference internal" href="#id55">[8.2.9]</a><ul>
<li><a class="reference internal" href="#id56">Added:</a></li>
<li><a class="reference internal" href="#id57">Fixes:</a></li>
<li><a class="reference internal" href="#id58">Changed:</a></li>
</ul>
</li>
<li><a class="reference internal" href="#id59">[8.2.8]</a><ul>
<li><a class="reference internal" href="#id60">Added:</a></li>
<li><a class="reference internal" href="#id61">Changed:</a></li>
<li><a class="reference internal" href="#id62">Fixes:</a></li>
</ul>
</li>
<li><a class="reference internal" href="#id63">[8.2.7]</a><ul>
<li><a class="reference internal" href="#id64">Fixes:</a></li>
</ul>
</li>
<li><a class="reference internal" href="#id65">[8.2.6]</a><ul>
<li><a class="reference internal" href="#id66">Fixes:</a></li>
</ul>
</li>
<li><a class="reference internal" href="#id67">[8.2.5]</a><ul>
<li><a class="reference internal" href="#id68">Fixed:</a></li>
</ul>
</li>
<li><a class="reference internal" href="#id69">[8.2.4]</a><ul>
<li><a class="reference internal" href="#id70">Added:</a></li>
</ul>
</li>
<li><a class="reference internal" href="#id71">[8.2.3]</a><ul>
<li><a class="reference internal" href="#id72">Fixed:</a></li>
<li><a class="reference internal" href="#id73">Changed:</a></li>
</ul>
</li>
<li><a class="reference internal" href="#id74">[8.2.2]</a><ul>
<li><a class="reference internal" href="#id75">Added:</a></li>
<li><a class="reference internal" href="#id76">Changed:</a></li>
</ul>
</li>
<li><a class="reference internal" href="#id77">[8.2.1]</a><ul>
<li><a class="reference internal" href="#id78">Fixed:</a></li>
</ul>
</li>
<li><a class="reference internal" href="#id79">[8.2.0]</a><ul>
<li><a class="reference internal" href="#id80">Added:</a></li>
<li><a class="reference internal" href="#id81">Changed:</a></li>
<li><a class="reference internal" href="#id82">Removed:</a></li>
<li><a class="reference internal" href="#id83">Fixed:</a></li>
</ul>
</li>
<li><a class="reference internal" href="#id84">[8.1.0]</a><ul>
<li><a class="reference internal" href="#id85">Added:</a></li>
<li><a class="reference internal" href="#id86">Changed:</a></li>
<li><a class="reference internal" href="#id87">Fixed:</a></li>
</ul>
</li>
<li><a class="reference internal" href="#id88">[8.0.2]</a><ul>
<li><a class="reference internal" href="#id89">Fixed:</a></li>
</ul>
</li>
<li><a class="reference internal" href="#id90">[8.0.1]</a><ul>
<li><a class="reference internal" href="#id91">Fixed:</a></li>
</ul>
</li>
<li><a class="reference internal" href="#id92">[8.0.0]</a><ul>
<li><a class="reference internal" href="#id93">Added:</a></li>
<li><a class="reference internal" href="#id94">Changed:</a></li>
<li><a class="reference internal" href="#id95">Fixed:</a></li>
<li><a class="reference internal" href="#id96">Removed:</a></li>
</ul>
</li>
<li><a class="reference internal" href="#id97">[7.2.5]</a><ul>
<li><a class="reference internal" href="#id98">Changed:</a></li>
</ul>
</li>
<li><a class="reference internal" href="#id99">[7.2.5]</a><ul>
<li><a class="reference internal" href="#id100">Changed:</a></li>
</ul>
</li>
<li><a class="reference internal" href="#id101">[7.2.4]</a><ul>
<li><a class="reference internal" href="#id102">Changed:</a></li>
</ul>
</li>
<li><a class="reference internal" href="#id103">[7.2.3]</a><ul>
<li><a class="reference internal" href="#id104">Fixed:</a></li>
</ul>
</li>
<li><a class="reference internal" href="#id105">[7.2.2]</a><ul>
<li><a class="reference internal" href="#id106">Fixed:</a></li>
</ul>
</li>
<li><a class="reference internal" href="#id107">[7.2.1]</a><ul>
<li><a class="reference internal" href="#id108">Changed:</a></li>
</ul>
</li>
<li><a class="reference internal" href="#id109">[7.2.0]</a><ul>
<li><a class="reference internal" href="#id110">Added:</a></li>
<li><a class="reference internal" href="#id111">Changed:</a></li>
<li><a class="reference internal" href="#id112">Fixed:</a></li>
<li><a class="reference internal" href="#id113">Removed:</a></li>
</ul>
</li>
<li><a class="reference internal" href="#id114">[7.1.10]</a><ul>
<li><a class="reference internal" href="#id115">Fixed:</a></li>
</ul>
</li>
<li><a class="reference internal" href="#id116">[7.1.9]</a><ul>
<li><a class="reference internal" href="#id117">Changed:</a></li>
<li><a class="reference internal" href="#id118">Removed:</a></li>
</ul>
</li>
<li><a class="reference internal" href="#id119">[7.1.8]</a><ul>
<li><a class="reference internal" href="#id120">Fixed:</a></li>
<li><a class="reference internal" href="#id121">Changed:</a></li>
</ul>
</li>
<li><a class="reference internal" href="#id122">[7.1.7]</a><ul>
<li><a class="reference internal" href="#id123">Added:</a></li>
<li><a class="reference internal" href="#id124">Changed:</a></li>
<li><a class="reference internal" href="#id125">Fixed:</a></li>
</ul>
</li>
<li><a class="reference internal" href="#id126">[7.1.6]</a><ul>
<li><a class="reference internal" href="#id127">Fixed:</a></li>
</ul>
</li>
<li><a class="reference internal" href="#id128">[7.1.5]</a><ul>
<li><a class="reference internal" href="#id129">Changed:</a></li>
</ul>
</li>
<li><a class="reference internal" href="#id130">[7.1.4]</a><ul>
<li><a class="reference internal" href="#id131">Changed:</a></li>
</ul>
</li>
<li><a class="reference internal" href="#id132">[7.1.3]</a><ul>
<li><a class="reference internal" href="#id133">Fixed:</a></li>
</ul>
</li>
<li><a class="reference internal" href="#id134">[7.1.2]</a><ul>
<li><a class="reference internal" href="#id135">Fixed:</a></li>
</ul>
</li>
<li><a class="reference internal" href="#id136">[7.1.1]</a><ul>
<li><a class="reference internal" href="#id137">Fixed:</a></li>
</ul>
</li>
<li><a class="reference internal" href="#id138">[7.1.0]</a><ul>
<li><a class="reference internal" href="#id139">Removed</a></li>
<li><a class="reference internal" href="#id140">Fixed</a></li>
<li><a class="reference internal" href="#id141">Changed</a></li>
</ul>
</li>
<li><a class="reference internal" href="#id142">[7.0.2]</a><ul>
<li><a class="reference internal" href="#id143">Added</a></li>
<li><a class="reference internal" href="#id144">Fixed</a></li>
</ul>
</li>
<li><a class="reference internal" href="#id145">[7.0.1]</a><ul>
<li><a class="reference internal" href="#id146">Added</a></li>
</ul>
</li>
<li><a class="reference internal" href="#id147">[7.0.0]</a><ul>
<li><a class="reference internal" href="#id148">Added</a></li>
<li><a class="reference internal" href="#id149">Changed</a></li>
<li><a class="reference internal" href="#id150">Fixed</a></li>
</ul>
</li>
<li><a class="reference internal" href="#id151">[6.1.2]</a><ul>
<li><a class="reference internal" href="#id152">Changed</a></li>
</ul>
</li>
<li><a class="reference internal" href="#id153">[6.1.1]</a><ul>
<li><a class="reference internal" href="#id154">Fixed</a></li>
<li><a class="reference internal" href="#id155">Removed</a></li>
</ul>
</li>
<li><a class="reference internal" href="#id156">[6.1.1]</a></li>
<li><a class="reference internal" href="#id157">[6.1.0]</a><ul>
<li><a class="reference internal" href="#id158">Added</a></li>
</ul>
</li>
<li><a class="reference internal" href="#id159">[6.0.4]</a><ul>
<li><a class="reference internal" href="#id160">Changed</a></li>
</ul>
</li>
<li><a class="reference internal" href="#id161">[6.0.3]</a><ul>
<li><a class="reference internal" href="#id162">Added</a></li>
<li><a class="reference internal" href="#id163">Fixed</a></li>
</ul>
</li>
<li><a class="reference internal" href="#id164">[6.0.2]</a><ul>
<li><a class="reference internal" href="#id165">Added</a></li>
<li><a class="reference internal" href="#id166">Fixed</a></li>
</ul>
</li>
<li><a class="reference internal" href="#id167">[6.0.1]</a><ul>
<li><a class="reference internal" href="#id168">Added</a></li>
</ul>
</li>
<li><a class="reference internal" href="#id169">[6.0.0]</a><ul>
<li><a class="reference internal" href="#id170">Added</a></li>
<li><a class="reference internal" href="#id171">Fixed</a></li>
<li><a class="reference internal" href="#id172">Changed</a></li>
</ul>
</li>
<li><a class="reference internal" href="#id173">[5.1.0]</a><ul>
<li><a class="reference internal" href="#id174">Fixed</a></li>
<li><a class="reference internal" href="#id175">Changed</a></li>
<li><a class="reference internal" href="#id176">Removed</a></li>
</ul>
</li>
<li><a class="reference internal" href="#id177">[5.0.1]</a><ul>
<li><a class="reference internal" href="#id178">Added</a></li>
<li><a class="reference internal" href="#id179">Changed</a></li>
</ul>
</li>
<li><a class="reference internal" href="#id180">[5.0.0]</a><ul>
<li><a class="reference internal" href="#id181">Added</a></li>
<li><a class="reference internal" href="#id182">Changed</a></li>
<li><a class="reference internal" href="#id183">Fixed</a></li>
<li><a class="reference internal" href="#id184">Removed</a></li>
</ul>
</li>
<li><a class="reference internal" href="#id185">[4.5.0]</a><ul>
<li><a class="reference internal" href="#id186">Added</a></li>
<li><a class="reference internal" href="#id187">Changed</a></li>
<li><a class="reference internal" href="#id188">Removed</a></li>
<li><a class="reference internal" href="#refactored">Refactored</a></li>
<li><a class="reference internal" href="#id189">Fixed</a></li>
</ul>
</li>
<li><a class="reference internal" href="#id190">[4.4.0]</a><ul>
<li><a class="reference internal" href="#id191">Added</a></li>
<li><a class="reference internal" href="#id192">Changed</a></li>
</ul>
</li>
<li><a class="reference internal" href="#id193">[4.3.0]</a><ul>
<li><a class="reference internal" href="#id194">Added</a></li>
<li><a class="reference internal" href="#deprecated">Deprecated</a></li>
<li><a class="reference internal" href="#id195">Fixed</a></li>
<li><a class="reference internal" href="#id196">Changed</a></li>
</ul>
</li>
<li><a class="reference internal" href="#id197">[4.2.4]</a><ul>
<li><a class="reference internal" href="#id198">Fixed</a></li>
</ul>
</li>
<li><a class="reference internal" href="#id199">[4.2.3]</a><ul>
<li><a class="reference internal" href="#id200">Added</a></li>
<li><a class="reference internal" href="#id201">Changed</a></li>
<li><a class="reference internal" href="#id202">Fixed</a></li>
</ul>
</li>
<li><a class="reference internal" href="#id203">[4.2.2]</a><ul>
<li><a class="reference internal" href="#id204">Added</a></li>
</ul>
</li>
<li><a class="reference internal" href="#id205">[4.2.1]</a><ul>
<li><a class="reference internal" href="#id206">Changed</a></li>
</ul>
</li>
<li><a class="reference internal" href="#id207">[4.2.0]</a><ul>
<li><a class="reference internal" href="#id208">Added</a></li>
<li><a class="reference internal" href="#id209">Changed</a></li>
<li><a class="reference internal" href="#id210">Fixed</a></li>
<li><a class="reference internal" href="#id211">Removed</a></li>
</ul>
</li>
<li><a class="reference internal" href="#id212">[4.1.0]</a><ul>
<li><a class="reference internal" href="#id213">Added</a></li>
<li><a class="reference internal" href="#id214">Changed</a></li>
<li><a class="reference internal" href="#id215">Fixed</a></li>
</ul>
</li>
<li><a class="reference internal" href="#id216">[4.0.1] - 2019-11-08</a><ul>
<li><a class="reference internal" href="#id217">Added</a></li>
<li><a class="reference internal" href="#id218">Changed</a></li>
</ul>
</li>
<li><a class="reference internal" href="#id219">[4.0.0] - 2019-11-04</a><ul>
<li><a class="reference internal" href="#id220">Added</a></li>
<li><a class="reference internal" href="#id221">Changed</a></li>
</ul>
</li>
<li><a class="reference internal" href="#id222">[3.3.1] - 2019-10-28</a><ul>
<li><a class="reference internal" href="#id223">Fixed</a></li>
</ul>
</li>
<li><a class="reference internal" href="#id224">[3.3.0] - 2019-10-24</a><ul>
<li><a class="reference internal" href="#id225">Added</a></li>
<li><a class="reference internal" href="#id226">Changed</a></li>
</ul>
</li>
<li><a class="reference internal" href="#id227">[3.2.3] - 2019-10-24</a><ul>
<li><a class="reference internal" href="#id228">Fixed</a></li>
</ul>
</li>
<li><a class="reference internal" href="#id229">[3.2.2] - 2019-10-23</a><ul>
<li><a class="reference internal" href="#id230">Fixed</a></li>
</ul>
</li>
<li><a class="reference internal" href="#id231">[3.2.1] - 2019-10-23</a><ul>
<li><a class="reference internal" href="#id232">Fixed</a></li>
</ul>
</li>
<li><a class="reference internal" href="#id233">[3.2.0] - 2019-10-11</a><ul>
<li><a class="reference internal" href="#adds">Adds:</a></li>
<li><a class="reference internal" href="#id234">Changed:</a></li>
<li><a class="reference internal" href="#id235">Removed:</a></li>
</ul>
</li>
<li><a class="reference internal" href="#id236">[3.1.4] - 2019-10-08</a><ul>
<li><a class="reference internal" href="#id237">Fixed</a></li>
</ul>
</li>
<li><a class="reference internal" href="#id238">[3.1.3] - 2019-10-07</a><ul>
<li><a class="reference internal" href="#id239">Fixed</a></li>
</ul>
</li>
<li><a class="reference internal" href="#id240">[3.1.2] - 2019-10-07</a><ul>
<li><a class="reference internal" href="#id241">Fixed</a></li>
</ul>
</li>
<li><a class="reference internal" href="#id242">[3.1.1] - 2019-10-03</a><ul>
<li><a class="reference internal" href="#id243">Fixed</a></li>
</ul>
</li>
<li><a class="reference internal" href="#id244">[3.1.0] - 2019-10-03</a><ul>
<li><a class="reference internal" href="#id245">Added</a></li>
<li><a class="reference internal" href="#id246">Fixed</a></li>
<li><a class="reference internal" href="#id247">Changed</a></li>
<li><a class="reference internal" href="#id248">Removed</a></li>
</ul>
</li>
<li><a class="reference internal" href="#id249">[3.0.1] - 2019-09-11</a><ul>
<li><a class="reference internal" href="#id250">Fixed</a></li>
</ul>
</li>
<li><a class="reference internal" href="#id251">[3.0.0] - 2019-09-05</a><ul>
<li><a class="reference internal" href="#id252">Added</a></li>
<li><a class="reference internal" href="#id253">Fixed</a></li>
<li><a class="reference internal" href="#id254">Changed</a></li>
<li><a class="reference internal" href="#id255">Removed</a></li>
</ul>
</li>
<li><a class="reference internal" href="#id256">[2.9.8] - 2019-01-01</a><ul>
<li><a class="reference internal" href="#id257">Fixed</a></li>
</ul>
</li>
<li><a class="reference internal" href="#id258">[2.9.7] - 2019-02-28</a><ul>
<li><a class="reference internal" href="#id259">Fixed</a></li>
<li><a class="reference internal" href="#id260">Changed</a></li>
</ul>
</li>
<li><a class="reference internal" href="#id261">[2.9.6] - 2019-02-25</a><ul>
<li><a class="reference internal" href="#id262">Fixed</a></li>
</ul>
</li>
<li><a class="reference internal" href="#id263">[2.9.5] - 2019-02-25</a><ul>
<li><a class="reference internal" href="#id264">Added</a></li>
</ul>
</li>
<li><a class="reference internal" href="#id265">[2.9.4] - 2019-02-13</a><ul>
<li><a class="reference internal" href="#id266">Fixed</a></li>
</ul>
</li>
<li><a class="reference internal" href="#id267">[2.9.3] - 2019-02-12</a><ul>
<li><a class="reference internal" href="#id268">Fixed</a></li>
</ul>
</li>
<li><a class="reference internal" href="#id269">[2.9.2] - 2019-02-04</a><ul>
<li><a class="reference internal" href="#id270">Added</a></li>
</ul>
</li>
<li><a class="reference internal" href="#id271">[2.9.1] - 2019-01-22</a><ul>
<li><a class="reference internal" href="#id272">Added</a></li>
<li><a class="reference internal" href="#id273">Fixed</a></li>
</ul>
</li>
<li><a class="reference internal" href="#id274">[2.9.0] - 2019-01-04</a><ul>
<li><a class="reference internal" href="#id275">Changed</a></li>
<li><a class="reference internal" href="#id276">Added</a></li>
</ul>
</li>
<li><a class="reference internal" href="#id277">[2.8.1] - 2018-11-28</a><ul>
<li><a class="reference internal" href="#id278">Fixed</a></li>
</ul>
</li>
<li><a class="reference internal" href="#id279">[2.8.0] - 2018-11-23</a><ul>
<li><a class="reference internal" href="#id280">Added</a></li>
</ul>
</li>
<li><a class="reference internal" href="#id281">[2.7.4] - 2018-11-23</a><ul>
<li><a class="reference internal" href="#id282">Added</a></li>
</ul>
</li>
<li><a class="reference internal" href="#id283">[2.7.3] - 2018-11-20</a><ul>
<li><a class="reference internal" href="#id284">Fixed</a></li>
</ul>
</li>
<li><a class="reference internal" href="#id285">[2.7.2] - 2018-11-15</a><ul>
<li><a class="reference internal" href="#id286">Changed</a></li>
</ul>
</li>
<li><a class="reference internal" href="#id287">[2.7.1] - 2018-11-13</a><ul>
<li><a class="reference internal" href="#id288">Fixed</a></li>
</ul>
</li>
<li><a class="reference internal" href="#id289">[2.7.0] - 2018-11-08</a><ul>
<li><a class="reference internal" href="#id290">Added</a></li>
<li><a class="reference internal" href="#id291">Changed</a></li>
<li><a class="reference internal" href="#id292">Removed</a></li>
<li><a class="reference internal" href="#id293">Changed</a></li>
<li><a class="reference internal" href="#id294">Fixed</a></li>
</ul>
</li>
<li><a class="reference internal" href="#id295">[2.6.3] - 2018-11-01</a><ul>
<li><a class="reference internal" href="#id296">Changed</a></li>
</ul>
</li>
<li><a class="reference internal" href="#id297">[2.6.2] - 2018-10-31</a><ul>
<li><a class="reference internal" href="#id298">Fixed</a></li>
</ul>
</li>
<li><a class="reference internal" href="#id299">[2.6.1] - 2018-10-29</a><ul>
<li><a class="reference internal" href="#id300">Fixed</a></li>
</ul>
</li>
<li><a class="reference internal" href="#id301">[2.5.0] - 2018-10-22</a><ul>
<li><a class="reference internal" href="#id302">Changed</a></li>
<li><a class="reference internal" href="#id303">Added</a></li>
</ul>
</li>
<li><a class="reference internal" href="#id304">[2.4.0] - 2018-10-22</a><ul>
<li><a class="reference internal" href="#id305">Changed</a></li>
</ul>
</li>
<li><a class="reference internal" href="#id306">[2.3.3] - 2018-10-22</a><ul>
<li><a class="reference internal" href="#id307">Added</a></li>
</ul>
</li>
<li><a class="reference internal" href="#id308">[2.3.2] - 2018-10-19</a><ul>
<li><a class="reference internal" href="#id309">Changed</a></li>
<li><a class="reference internal" href="#id310">Added</a></li>
</ul>
</li>
<li><a class="reference internal" href="#id311">[2.3.1] - 2018-09-25</a><ul>
<li><a class="reference internal" href="#id312">Added</a></li>
</ul>
</li>
<li><a class="reference internal" href="#id313">[2.3.0] - 2018-09-25</a><ul>
<li><a class="reference internal" href="#id314">Changed</a></li>
<li><a class="reference internal" href="#id315">Fixed</a></li>
</ul>
</li>
<li><a class="reference internal" href="#id316">[2.2.0] - 2018-09-11</a><ul>
<li><a class="reference internal" href="#id317">Added</a></li>
</ul>
</li>
<li><a class="reference internal" href="#id318">[2.1.0] - 2018-09-11</a><ul>
<li><a class="reference internal" href="#id319">Added</a></li>
</ul>
</li>
<li><a class="reference internal" href="#id320">[2.0.2] - 2018-09-11</a><ul>
<li><a class="reference internal" href="#id321">Added</a></li>
</ul>
</li>
<li><a class="reference internal" href="#id322">[2.0.1] - 2018-09-10</a><ul>
<li><a class="reference internal" href="#id323">Fixed</a></li>
</ul>
</li>
<li><a class="reference internal" href="#id324">[2.0.0] - 2018-09-05</a><ul>
<li><a class="reference internal" href="#id325">Added</a></li>
</ul>
</li>
<li><a class="reference internal" href="#id326">[1.13.1] - 2018-08-17</a><ul>
<li><a class="reference internal" href="#id327">Fixed</a></li>
</ul>
</li>
<li><a class="reference internal" href="#id328">[1.13.0] - 2018-08-10</a><ul>
<li><a class="reference internal" href="#id329">Added</a></li>
</ul>
</li>
<li><a class="reference internal" href="#id330">[1.12.0] - 2018-07-06</a><ul>
<li><a class="reference internal" href="#id331">Added</a></li>
</ul>
</li>
<li><a class="reference internal" href="#id332">[1.11.0] - 2018-07-05</a><ul>
<li><a class="reference internal" href="#id333">Changed</a></li>
</ul>
</li>
<li><a class="reference internal" href="#id334">[1.10.0] - 2018-06-07</a><ul>
<li><a class="reference internal" href="#id335">Added</a></li>
</ul>
</li>
<li><a class="reference internal" href="#id336">[1.9.0] - 2018-06-04</a><ul>
<li><a class="reference internal" href="#id337">Added</a></li>
</ul>
</li>
<li><a class="reference internal" href="#id338">[1.8.0] - 2018-06-01</a><ul>
<li><a class="reference internal" href="#id339">Changed</a></li>
</ul>
</li>
<li><a class="reference internal" href="#id340">[1.7.0] - 2018-05-31</a><ul>
<li><a class="reference internal" href="#id341">Added</a></li>
</ul>
</li>
<li><a class="reference internal" href="#id342">[1.6.0] - 2018-05-30</a><ul>
<li><a class="reference internal" href="#id343">Added</a></li>
</ul>
</li>
<li><a class="reference internal" href="#id344">[1.5.0] - 2018-05-28</a><ul>
<li><a class="reference internal" href="#id345">Changed</a></li>
</ul>
</li>
<li><a class="reference internal" href="#id346">[1.4.0] - 2018-05-18</a><ul>
<li><a class="reference internal" href="#id347">Added</a></li>
</ul>
</li>
<li><a class="reference internal" href="#id348">[1.3.0] - 2018-05-13</a><ul>
<li><a class="reference internal" href="#id349">Added</a></li>
</ul>
</li>
<li><a class="reference internal" href="#id350">[1.2.3] - 2018-05-13</a><ul>
<li><a class="reference internal" href="#id351">Changed</a></li>
</ul>
</li>
<li><a class="reference internal" href="#id352">[1.2.2] - 2018-05-04</a><ul>
<li><a class="reference internal" href="#id353">Added</a></li>
</ul>
</li>
<li><a class="reference internal" href="#id354">[1.2.0] - 2018-05-02</a><ul>
<li><a class="reference internal" href="#id355">Added</a></li>
</ul>
</li>
<li><a class="reference internal" href="#id356">[1.1.2] - 2018-04-27</a><ul>
<li><a class="reference internal" href="#id357">Fixed</a></li>
</ul>
</li>
<li><a class="reference internal" href="#id358">[1.1.1] - 2018-04-27</a><ul>
<li><a class="reference internal" href="#id359">Fixed</a></li>
</ul>
</li>
<li><a class="reference internal" href="#id360">[1.1.0] - 2018-04-26</a><ul>
<li><a class="reference internal" href="#id361">Added</a></li>
<li><a class="reference internal" href="#id362">Changed</a></li>
<li><a class="reference internal" href="#id363">Fixed</a></li>
</ul>
</li>
<li><a class="reference internal" href="#rc2-2018-04-18">[1.0.3-rc2] - 2018-04-18</a><ul>
<li><a class="reference internal" href="#id364">Added</a></li>
</ul>
</li>
<li><a class="reference internal" href="#rc2-2018-04-17">[1.0.2-rc2] - 2018-04-17</a><ul>
<li><a class="reference internal" href="#id365">Added</a></li>
</ul>
</li>
<li><a class="reference internal" href="#rc2-2018-04-16">[1.0.1-rc2] - 2018-04-16</a><ul>
<li><a class="reference internal" href="#id366">Fixed</a></li>
</ul>
</li>
<li><a class="reference internal" href="#id367">[1.0.0-rc2] - 2018-04-16</a><ul>
<li><a class="reference internal" href="#id368">Added</a></li>
<li><a class="reference internal" href="#id369">Fixed</a></li>
</ul>
</li>
<li><a class="reference internal" href="#rc1-2018-04-05">[1.0.0-rc1] - 2018-04-05</a><ul>
<li><a class="reference internal" href="#id370">Added</a></li>
<li><a class="reference internal" href="#id371">Fixed</a></li>
</ul>
</li>
</ul>
</li>
</ul>

          </div>
        </div>
      </div>
      
      
    </aside>
  </div>
</div><script data-url_root="./" id="documentation_options" src="_static/documentation_options.js"></script>
    <script src="_static/doctools.js"></script>
    <script src="_static/sphinx_highlight.js"></script>
    <script src="_static/scripts/furo.js"></script>
    </body>
</html>