<!doctype html>
<html class="no-js" lang="en">
  <head><meta charset="utf-8"/>
    <meta name="viewport" content="width=device-width,initial-scale=1"/>
    <meta name="color-scheme" content="light dark"><meta name="generator" content="Docutils 0.18.1: http://docutils.sourceforge.net/" />
<link rel="index" title="Index" href="genindex.html" /><link rel="search" title="Search" href="search.html" /><link rel="next" title="Documentation Guidline" href="README.html" /><link rel="prev" title="Tools and software" href="bioinfo_softwares.html" />

    <!-- Generated with Sphinx 6.2.1 and Furo 2023.07.26 -->
        <title>References and other resources - BALSAMIC documentation</title>
      <link rel="stylesheet" type="text/css" href="_static/pygments.css" />
    <link rel="stylesheet" type="text/css" href="_static/styles/furo.css?digest=369552022d0b975c8e74270ce6eabe0fb7978f24" />
    <link rel="stylesheet" type="text/css" href="_static/styles/furo-extensions.css?digest=30d1aed668e5c3a91c3e3bf6a60b675221979f0e" />
    
    


<style>
  body {
    --color-code-background: #f8f8f8;
  --color-code-foreground: black;
  
  }
  @media not print {
    body[data-theme="dark"] {
      --color-code-background: #202020;
  --color-code-foreground: #d0d0d0;
  
    }
    @media (prefers-color-scheme: dark) {
      body:not([data-theme="light"]) {
        --color-code-background: #202020;
  --color-code-foreground: #d0d0d0;
  
      }
    }
  }
</style></head>
  <body>
    
    <script>
      document.body.dataset.theme = localStorage.getItem("theme") || "auto";
    </script>
    

<svg xmlns="http://www.w3.org/2000/svg" style="display: none;">
  <symbol id="svg-toc" viewBox="0 0 24 24">
    <title>Contents</title>
    <svg stroke="currentColor" fill="currentColor" stroke-width="0" viewBox="0 0 1024 1024">
      <path d="M408 442h480c4.4 0 8-3.6 8-8v-56c0-4.4-3.6-8-8-8H408c-4.4 0-8 3.6-8 8v56c0 4.4 3.6 8 8 8zm-8 204c0 4.4 3.6 8 8 8h480c4.4 0 8-3.6 8-8v-56c0-4.4-3.6-8-8-8H408c-4.4 0-8 3.6-8 8v56zm504-486H120c-4.4 0-8 3.6-8 8v56c0 4.4 3.6 8 8 8h784c4.4 0 8-3.6 8-8v-56c0-4.4-3.6-8-8-8zm0 632H120c-4.4 0-8 3.6-8 8v56c0 4.4 3.6 8 8 8h784c4.4 0 8-3.6 8-8v-56c0-4.4-3.6-8-8-8zM115.4 518.9L271.7 642c5.8 4.6 14.4.5 14.4-6.9V388.9c0-7.4-8.5-11.5-14.4-6.9L115.4 505.1a8.74 8.74 0 0 0 0 13.8z"/>
    </svg>
  </symbol>
  <symbol id="svg-menu" viewBox="0 0 24 24">
    <title>Menu</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="2" stroke-linecap="round" stroke-linejoin="round" class="feather-menu">
      <line x1="3" y1="12" x2="21" y2="12"></line>
      <line x1="3" y1="6" x2="21" y2="6"></line>
      <line x1="3" y1="18" x2="21" y2="18"></line>
    </svg>
  </symbol>
  <symbol id="svg-arrow-right" viewBox="0 0 24 24">
    <title>Expand</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="2" stroke-linecap="round" stroke-linejoin="round" class="feather-chevron-right">
      <polyline points="9 18 15 12 9 6"></polyline>
    </svg>
  </symbol>
  <symbol id="svg-sun" viewBox="0 0 24 24">
    <title>Light mode</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="1.5" stroke-linecap="round" stroke-linejoin="round" class="feather-sun">
      <circle cx="12" cy="12" r="5"></circle>
      <line x1="12" y1="1" x2="12" y2="3"></line>
      <line x1="12" y1="21" x2="12" y2="23"></line>
      <line x1="4.22" y1="4.22" x2="5.64" y2="5.64"></line>
      <line x1="18.36" y1="18.36" x2="19.78" y2="19.78"></line>
      <line x1="1" y1="12" x2="3" y2="12"></line>
      <line x1="21" y1="12" x2="23" y2="12"></line>
      <line x1="4.22" y1="19.78" x2="5.64" y2="18.36"></line>
      <line x1="18.36" y1="5.64" x2="19.78" y2="4.22"></line>
    </svg>
  </symbol>
  <symbol id="svg-moon" viewBox="0 0 24 24">
    <title>Dark mode</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="1.5" stroke-linecap="round" stroke-linejoin="round" class="icon-tabler-moon">
      <path stroke="none" d="M0 0h24v24H0z" fill="none" />
      <path d="M12 3c.132 0 .263 0 .393 0a7.5 7.5 0 0 0 7.92 12.446a9 9 0 1 1 -8.313 -12.454z" />
    </svg>
  </symbol>
  <symbol id="svg-sun-half" viewBox="0 0 24 24">
    <title>Auto light/dark mode</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="1.5" stroke-linecap="round" stroke-linejoin="round" class="icon-tabler-shadow">
      <path stroke="none" d="M0 0h24v24H0z" fill="none"/>
      <circle cx="12" cy="12" r="9" />
      <path d="M13 12h5" />
      <path d="M13 15h4" />
      <path d="M13 18h1" />
      <path d="M13 9h4" />
      <path d="M13 6h1" />
    </svg>
  </symbol>
</svg>

<input type="checkbox" class="sidebar-toggle" name="__navigation" id="__navigation">
<input type="checkbox" class="sidebar-toggle" name="__toc" id="__toc">
<label class="overlay sidebar-overlay" for="__navigation">
  <div class="visually-hidden">Hide navigation sidebar</div>
</label>
<label class="overlay toc-overlay" for="__toc">
  <div class="visually-hidden">Hide table of contents sidebar</div>
</label>



<div class="page">
  <header class="mobile-header">
    <div class="header-left">
      <label class="nav-overlay-icon" for="__navigation">
        <div class="visually-hidden">Toggle site navigation sidebar</div>
        <i class="icon"><svg><use href="#svg-menu"></use></svg></i>
      </label>
    </div>
    <div class="header-center">
      <a href="index.html"><div class="brand">BALSAMIC  documentation</div></a>
    </div>
    <div class="header-right">
      <div class="theme-toggle-container theme-toggle-header">
        <button class="theme-toggle">
          <div class="visually-hidden">Toggle Light / Dark / Auto color theme</div>
          <svg class="theme-icon-when-auto"><use href="#svg-sun-half"></use></svg>
          <svg class="theme-icon-when-dark"><use href="#svg-moon"></use></svg>
          <svg class="theme-icon-when-light"><use href="#svg-sun"></use></svg>
        </button>
      </div>
      <label class="toc-overlay-icon toc-header-icon" for="__toc">
        <div class="visually-hidden">Toggle table of contents sidebar</div>
        <i class="icon"><svg><use href="#svg-toc"></use></svg></i>
      </label>
    </div>
  </header>
  <aside class="sidebar-drawer">
    <div class="sidebar-container">
      
      <div class="sidebar-sticky"><a class="sidebar-brand" href="index.html">
  
  <div class="sidebar-logo-container">
    <img class="sidebar-logo" src="_static/balsamic_logo.png" alt="Logo"/>
  </div>
  
  <span class="sidebar-brand-text">BALSAMIC  documentation</span>
  
</a><form class="sidebar-search-container" method="get" action="search.html" role="search">
  <input class="sidebar-search" placeholder="Search" name="q" aria-label="Search">
  <input type="hidden" name="check_keywords" value="yes">
  <input type="hidden" name="area" value="default">
</form>
<div id="searchbox"></div><div class="sidebar-scroll"><div class="sidebar-tree">
  <p class="caption" role="heading"><span class="caption-text">Installation and Usage</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="install.html">Installation</a></li>
<li class="toctree-l1"><a class="reference internal" href="user_guide.html">Short tutorial</a></li>
<li class="toctree-l1"><a class="reference internal" href="cli_package.html">Command line arguments</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">Analysis and Interpretation</span></p>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="balsamic_filters.html">Calling and filtering variants</a></li>
<li class="toctree-l1"><a class="reference internal" href="balsamic_sv_cnv.html">Structural and Copy Number variants</a></li>
<li class="toctree-l1"><a class="reference internal" href="balsamic_annotation.html">Annotation resources</a></li>
<li class="toctree-l1"><a class="reference internal" href="balsamic_pon.html">Panel of Normals (PON)</a></li>
<li class="toctree-l1"><a class="reference internal" href="balsamic_pon.html#cnvkit-pon">CNVkit PON</a></li>
<li class="toctree-l1"><a class="reference internal" href="balsamic_pon.html#gens-pon">GENS PON</a></li>
<li class="toctree-l1"><a class="reference internal" href="balsamic_methods.html">Method description</a></li>
<li class="toctree-l1"><a class="reference internal" href="history.html">Changelog</a></li>
<li class="toctree-l1"><a class="reference internal" href="bioinfo_softwares.html">Tools and software</a></li>
<li class="toctree-l1 current current-page"><a class="current reference internal" href="#">References and other resources</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">Development guide</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="README.html">Documentation Guidline</a></li>
<li class="toctree-l1"><a class="reference internal" href="programming_guide.html">Coding etiquettes</a></li>
<li class="toctree-l1"><a class="reference internal" href="semver.html">Semantic versioning</a></li>
<li class="toctree-l1"><a class="reference internal" href="FAQs.html">Frequently Asked Questions (FAQs)</a></li>
</ul>

</div>
</div>

      </div>
      
    </div>
  </aside>
  <div class="main">
    <div class="content">
      <div class="article-container">
        <a href="#" class="back-to-top muted-link">
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24">
            <path d="M13 20h-2V8l-5.5 5.5-1.42-1.42L12 4.16l7.92 7.92-1.42 1.42L13 8v12z"></path>
          </svg>
          <span>Back to top</span>
        </a>
        <div class="content-icon-container">
          
<div class="theme-toggle-container theme-toggle-content">
            <button class="theme-toggle">
              <div class="visually-hidden">Toggle Light / Dark / Auto color theme</div>
              <svg class="theme-icon-when-auto"><use href="#svg-sun-half"></use></svg>
              <svg class="theme-icon-when-dark"><use href="#svg-moon"></use></svg>
              <svg class="theme-icon-when-light"><use href="#svg-sun"></use></svg>
            </button>
          </div>
          <label class="toc-overlay-icon toc-content-icon" for="__toc">
            <div class="visually-hidden">Toggle table of contents sidebar</div>
            <i class="icon"><svg><use href="#svg-toc"></use></svg></i>
          </label>
        </div>
        <article role="main">
          <section id="references-and-other-resources">
<h1>References and other resources<a class="headerlink" href="#references-and-other-resources" title="Permalink to this heading">#</a></h1>
<p><em>Main resources including knowledge base and databases necessary for pipeline development</em></p>
<ol class="arabic simple">
<li><p><strong>MSK-Impact pipeline</strong>: <a class="reference external" href="https://www.mskcc.org/msk-impact">https://www.mskcc.org/msk-impact</a></p></li>
<li><p><strong>TCGA</strong>: <a class="reference external" href="https://cancergenome.nih.gov/">https://cancergenome.nih.gov/</a></p></li>
<li><p><strong>COSMIC</strong>: <a class="reference external" href="http://cancer.sanger.ac.uk/cosmic">http://cancer.sanger.ac.uk/cosmic</a></p></li>
<li><p><strong>dbSNP</strong>:  Database of single nucleotide polymorphisms (SNPs) and multiple small-scale variations that include insertions/deletions, microsatellites, and non-polymorphic variants. <a class="reference external" href="https://www.ncbi.nlm.nih.gov/snp/">https://www.ncbi.nlm.nih.gov/snp/</a> Download link: <a class="reference external" href="ftp://ftp.ncbi.nlm.nih.gov/snp/organisms/human_9606_b150_GRCh38p7/VCF/All_20170710.vcf.gz">ftp://ftp.ncbi.nlm.nih.gov/snp/organisms/human_9606_b150_GRCh38p7/VCF/All_20170710.vcf.gz</a></p></li>
<li><p><strong>ClinVar</strong>: ClinVar aggregates information about genomic variation and its relationship to human health. <a class="reference external" href="https://www.ncbi.nlm.nih.gov/clinvar/">https://www.ncbi.nlm.nih.gov/clinvar/</a> Download link: <a class="reference external" href="ftp://ftp.ncbi.nlm.nih.gov/pub/clinvar/vcf_GRCh38/clinvar_20171029.vcf.gz">ftp://ftp.ncbi.nlm.nih.gov/pub/clinvar/vcf_GRCh38/clinvar_20171029.vcf.gz</a></p></li>
<li><p><strong>SweGen</strong>: This dataset contains whole-genome variant frequencies for 1000 Swedish individuals generated within the SweGen project. Download link:  <a class="reference external" href="https://swefreq.nbis.se/">https://swefreq.nbis.se/</a></p></li>
<li><p><strong>ExAC</strong>: The Exome Aggregation Consortium (ExAC) is a coalition of investigators seeking to aggregate and harmonize exome sequencing data from a wide variety of large-scale sequencing projects, and to make summary data available for the wider scientific community. <a class="reference external" href="http://exac.broadinstitute.org/">http://exac.broadinstitute.org/</a> Download link: <a class="reference external" href="ftp://ftp.broadinstitute.org/pub/ExAC_release/release1/ExAC.r1.sites.vep.vcf.gz">ftp://ftp.broadinstitute.org/pub/ExAC_release/release1/ExAC.r1.sites.vep.vcf.gz</a></p></li>
<li><p><strong>GTEx</strong>: The Genotype-Tissue Expression (GTEx) project aims to provide to the scientific community a resource with which to study human gene expression and regulation and its relationship to genetic variation. <a class="reference external" href="https://www.gtexportal.org/home/">https://www.gtexportal.org/home/</a> Download URL by applying through: <a class="reference external" href="https://www.ncbi.nlm.nih.gov/projects/gap/cgi-bin/study.cgi?study_id=phs000424.v6.p1">https://www.ncbi.nlm.nih.gov/projects/gap/cgi-bin/study.cgi?study_id=phs000424.v6.p1</a></p></li>
<li><p><strong>OMIM</strong>: OMIM®, Online Mendelian Inheritance in Man®, An Online Catalog of Human Genes and Genetic Disorders. <a class="reference external" href="https://www.omim.org/">https://www.omim.org/</a> Download link: <a class="reference external" href="https://omim.org/downloads/">https://omim.org/downloads/</a> (registration required)</p></li>
<li><p><strong>Drug resistance</strong>: An effort by Cosmic to annotate mutations identified in the literature as resistance mutations, including those conferring acquired resistance (after treatment) and intrinsic resistance (before treatment). Available through Cosmic: <a class="reference external" href="http://cancer.sanger.ac.uk/cosmic/drug_resistance">http://cancer.sanger.ac.uk/cosmic/drug_resistance</a></p></li>
<li><p><strong>Mutational signatures</strong>: Signatures of Mutational Processes in Human Cancer. Available through Cosmic: <a class="reference external" href="http://cancer.sanger.ac.uk/cosmic/signatures">http://cancer.sanger.ac.uk/cosmic/signatures</a></p></li>
<li><p><strong>DGVa</strong>: The Database of Genomic Variants archive (DGVa) is a repository that provides archiving, accessioning and distribution of publicly available genomic structural variants, in all species. <a class="reference external" href="https://www.ebi.ac.uk/dgva">https://www.ebi.ac.uk/dgva</a></p></li>
<li><p><strong>Cancer genomics workflow</strong>: MGI’s CWL Cancer Pipelines. <a class="reference external" href="https://github.com/genome/cancer-genomics-workflow/wiki">https://github.com/genome/cancer-genomics-workflow/wiki</a></p></li>
<li><p><strong>GIAB</strong>: The priority of GIAB is authoritative characterization of human genomes for use in analytical validation and technology development, optimization, and demonstration. <a class="reference external" href="https://github.com/genome-in-a-bottle">https://github.com/genome-in-a-bottle</a></p></li>
<li><p><strong>dbNSFP</strong>: dbNSFP is a database developed for functional prediction and annotation of all potential non-synonymous single-nucleotide variants (nsSNVs) in the human genome. <a class="reference external" href="https://sites.google.com/site/jpopgen/dbNSFP">https://sites.google.com/site/jpopgen/dbNSFP</a></p></li>
<li><p><strong>1000Genomes</strong>: The goal of the 1000 Genomes Project was to find most genetic variants with frequencies of at least 1% in the populations studied. <a class="reference external" href="http://www.internationalgenome.org/">http://www.internationalgenome.org/</a> Download link: <a class="reference external" href="ftp://ftp.1000genomes.ebi.ac.uk/vol1/ftp/">ftp://ftp.1000genomes.ebi.ac.uk/vol1/ftp/</a></p></li>
<li><p><strong>HapMap3</strong>: The International HapMap Project was an organization that aimed to develop a haplotype map (HapMap) of the human genome, to describe the common patterns of human genetic variation. HapMap 3 is the third phase of the International HapMap project. <a class="reference external" href="http://www.sanger.ac.uk/resources/downloads/human/hapmap3.html">http://www.sanger.ac.uk/resources/downloads/human/hapmap3.html</a> Download link: <a class="reference external" href="ftp://ftp.ncbi.nlm.nih.gov/hapmap/">ftp://ftp.ncbi.nlm.nih.gov/hapmap/</a></p></li>
<li><p><strong>GRCh38.p11</strong>: GRCh38.p11 is the eleventh patch release for the GRCh38 (human) reference assembly. <a class="reference external" href="https://www.ncbi.nlm.nih.gov/grc/human">https://www.ncbi.nlm.nih.gov/grc/human</a> Download link: <a class="reference external" href="ftp://ftp.ncbi.nlm.nih.gov/genomes/all/GCA/000/001/405/">ftp://ftp.ncbi.nlm.nih.gov/genomes/all/GCA/000/001/405/</a></p></li>
<li><p><strong>dbVar</strong>: dbVar is NCBI’s database of genomic structural variation – insertions, deletions, duplications, inversions, mobile element insertions, translocations, and complex chromosomal rearrangements <a class="reference external" href="https://www.ncbi.nlm.nih.gov/dbvar">https://www.ncbi.nlm.nih.gov/dbvar</a> Download link: <a class="reference external" href="https://www.ncbi.nlm.nih.gov/dbvar/content/ftp_manifest/">https://www.ncbi.nlm.nih.gov/dbvar/content/ftp_manifest/</a></p></li>
<li><p><strong>Drug sensitivity in cancer</strong>: Identifying molecular features of cancers that predict response to anti-cancer drugs. <a class="reference external" href="http://www.cancerrxgene.org/">http://www.cancerrxgene.org/</a> Download link: <a class="reference external" href="ftp://ftp.sanger.ac.uk/pub4/cancerrxgene/releases">ftp://ftp.sanger.ac.uk/pub4/cancerrxgene/releases</a></p></li>
<li><p><strong>VarSome</strong>: VarSome is a knowledge base and aggregator for human genomic variants. <a class="reference external" href="https://varsome.com/about/">https://varsome.com/about/</a></p></li>
<li><p><strong>CADD</strong>: CADD is a tool for scoring the deleteriousness of single nucleotide variants as well as insertion/deletions variants in the human genome. CADD can quantitatively prioritize functional, deleterious, and disease causal variants across a wide range of functional categories, effect sizes and genetic architectures and can be used prioritize causal variation in both research and clinical settings.</p></li>
</ol>
<section id="sample-datasets">
<h2>Sample datasets<a class="headerlink" href="#sample-datasets" title="Permalink to this heading">#</a></h2>
<ol class="arabic simple">
<li><p><strong>TCRB</strong>: The Texas Cancer Research Biobank (TCRB) was created to bridge the gap between doctors and scientific researchers to improve the prevention, diagnosis and treatment of cancer. This work occurred with funding from the Cancer Prevention &amp; Research Institute of Texas (CPRIT) from 2010-2014. <a class="reference external" href="http://txcrb.org/data.html">http://txcrb.org/data.html</a> Article: <a class="reference external" href="https://www.nature.com/articles/sdata201610">https://www.nature.com/articles/sdata201610</a></p></li>
</ol>
</section>
<section id="relevant-publications">
<h2>Relevant publications<a class="headerlink" href="#relevant-publications" title="Permalink to this heading">#</a></h2>
<p><em>Including methodological benchmarking</em></p>
<ol class="arabic simple">
<li><p><strong>MSK-IMPACT</strong>:</p>
<ul class="simple">
<li><p><strong>Original pipeline</strong>: Cheng, D. T., Mitchell, T. N., Zehir, A., Shah, R. H., Benayed, R., Syed, A., … Berger, M. F. (2015). Memorial sloan kettering-integrated mutation profiling of actionable cancer targets (MSK-IMPACT): A hybridization capture-based next-generation sequencing clinical assay for solid tumor molecular oncology. Journal of Molecular Diagnostics, 17(3), 251–264. <a class="reference external" href="https://doi.org/10.1016/j.jmoldx.2014.12.006">https://doi.org/10.1016/j.jmoldx.2014.12.006</a></p></li>
<li><p><strong>Case study</strong>: Cheng, D. T., Prasad, M., Chekaluk, Y., Benayed, R., Sadowska, J., Zehir, A., … Zhang, L. (2017). Comprehensive detection of germline variants by MSK-IMPACT, a clinical diagnostic platform for solid tumor molecular oncology and concurrent cancer predisposition testing. BMC Medical Genomics, 10(1), 33. <a class="reference external" href="https://doi.org/10.1186/s12920-017-0271-4">https://doi.org/10.1186/s12920-017-0271-4</a></p></li>
<li><p><strong>Case study</strong>: Zehir, A., Benayed, R., Shah, R. H., Syed, A., Middha, S., Kim, H. R., … Berger, M. F. (2017). Mutational landscape of metastatic cancer revealed from prospective clinical sequencing of 10,000 patients. Nature Medicine, 23(6), 703–713. <a class="reference external" href="https://doi.org/10.1038/nm.4333">https://doi.org/10.1038/nm.4333</a></p></li>
</ul>
</li>
<li><p><strong>Application of MSK-IMPACT</strong>: Zehir, A., Benayed, R., Shah, R. H., Syed, A., Middha, S., Kim, H. R., … Berger, M. F. (2017). Mutational landscape of metastatic cancer revealed from prospective clinical sequencing of 10,000 patients. Nature Medicine, 23(6), 703–713. <a class="reference external" href="https://doi.org/10.1038/nm.4333">https://doi.org/10.1038/nm.4333</a></p></li>
<li><p><strong>Review on bioinformatic pipelins</strong>: Leipzig, J. (2017). A review of bioinformatic pipeline frameworks. Briefings in Bioinformatics, 18(3), 530–536. <a class="reference external" href="https://doi.org/10.1093/bib/bbw020">https://doi.org/10.1093/bib/bbw020</a></p></li>
<li><p><strong>Mutational signature reviews</strong>:</p>
<ul class="simple">
<li><p>Helleday, T., Eshtad, S., &amp; Nik-Zainal, S. (2014). Mechanisms underlying mutational signatures in human cancers. Nature Reviews Genetics, 15(9), 585–598. <a class="reference external" href="https://doi.org/10.1038/nrg3729">https://doi.org/10.1038/nrg3729</a></p></li>
<li><p>Alexandrov, L. B., &amp; Stratton, M. R. (2014). Mutational signatures: The patterns of somatic mutations hidden in cancer genomes. Current Opinion in Genetics and Development, 24(1), 52–60. <a class="reference external" href="https://doi.org/10.1016/j.gde.2013.11.014">https://doi.org/10.1016/j.gde.2013.11.014</a></p></li>
</ul>
</li>
<li><p><strong>Review on structural variation detection tools</strong>:</p>
<ul class="simple">
<li><p>Lin, K., Bonnema, G., Sanchez-Perez, G., &amp; De Ridder, D. (2014). Making the difference: Integrating structural variation detection tools. Briefings in Bioinformatics, 16(5), 852–864. <a class="reference external" href="https://doi.org/10.1093/bib/bbu047">https://doi.org/10.1093/bib/bbu047</a></p></li>
<li><p>Tattini, L., D’Aurizio, R., &amp; Magi, A. (2015). Detection of Genomic Structural Variants from Next-Generation Sequencing Data. Frontiers in Bioengineering and Biotechnology, 3(June), 1–8. <a class="reference external" href="https://doi.org/10.3389/fbioe.2015.00092">https://doi.org/10.3389/fbioe.2015.00092</a></p></li>
</ul>
</li>
<li><p><strong>Two case studies and a pipeline (unpublished)</strong>: Noll, A. C., Miller, N. A., Smith, L. D., Yoo, B., Fiedler, S., Cooley, L. D., … Kingsmore, S. F. (2016). Clinical detection of deletion structural variants in whole-genome sequences. Npj Genomic Medicine, 1(1), 16026. <a class="reference external" href="https://doi.org/10.1038/npjgenmed.2016.26">https://doi.org/10.1038/npjgenmed.2016.26</a></p></li>
<li><p><strong>Review on driver gene methods</strong>: Tokheim, C. J., Papadopoulos, N., Kinzler, K. W., Vogelstein, B., &amp; Karchin, R. (2016). Evaluating the evaluation of cancer driver genes. Proceedings of the National Academy of Sciences, 113(50), 14330–14335. <a class="reference external" href="https://doi.org/10.1073/pnas.1616440113">https://doi.org/10.1073/pnas.1616440113</a></p></li>
</ol>
<p><em>Resource, or general notable papers including resource and KB papers related to cancer genomics</em></p>
<ol class="arabic simple">
<li><p><strong>GIAB</strong>: Zook, J. M., Catoe, D., McDaniel, J., Vang, L., Spies, N., Sidow, A., … Salit, M. (2016). Extensive sequencing of seven human genomes to characterize benchmark reference materials. Scientific Data, 3, 160025. <a class="reference external" href="https://doi.org/10.1038/sdata.2016.25">https://doi.org/10.1038/sdata.2016.25</a></p></li>
</ol>
</section>
<section id="methods-and-tools">
<h2>Methods and tools<a class="headerlink" href="#methods-and-tools" title="Permalink to this heading">#</a></h2>
<p><em>Excluding multiple method comparison or benchmarking tools</em></p>
<ul class="simple">
<li><p><strong>BreakDancer</strong>: Chen, K., Wallis, J. W., Mclellan, M. D., Larson, D. E., Kalicki, J. M., Pohl, C. S., … Elaine, R. (2013). BreakDancer - An algorithm for high resolution mapping of genomic structure variation. Nature Methods, 6(9), 677–681. <a class="reference external" href="https://doi.org/10.1038/nmeth.1363">https://doi.org/10.1038/nmeth.1363</a></p></li>
<li><p><strong>Pindel</strong>: Ye, K., Schulz, M. H., Long, Q., Apweiler, R., &amp; Ning, Z. (2009). Pindel: A pattern growth approach to detect break points of large deletions and medium sized insertions from paired-end short reads. Bioinformatics, 25(21), 2865–2871. <a class="reference external" href="https://doi.org/10.1093/bioinformatics/btp394">https://doi.org/10.1093/bioinformatics/btp394</a></p></li>
<li><p><strong>SVDetect</strong>: Zeitouni, B., Boeva, V., Janoueix-Lerosey, I., Loeillet, S., Legoix-né, P., Nicolas, A., … Barillot, E. (2010). SVDetect: A tool to identify genomic structural variations from paired-end and mate-pair sequencing data. Bioinformatics, 26(15), 1895–1896. <a class="reference external" href="https://doi.org/10.1093/bioinformatics/btq293">https://doi.org/10.1093/bioinformatics/btq293</a></p></li>
<li><p><strong>Purityest</strong>: Su, X., Zhang, L., Zhang, J., Meric-bernstam, F., &amp; Weinstein, J. N. (2012). Purityest: Estimating purity of human tumor samples using next-generation sequencing data. Bioinformatics, 28(17), 2265–2266. <a class="reference external" href="https://doi.org/10.1093/bioinformatics/bts365">https://doi.org/10.1093/bioinformatics/bts365</a></p></li>
<li><p><strong>PurBayes</strong>: Larson, N. B., &amp; Fridley, B. L. (2013). PurBayes: Estimating tumor cellularity and subclonality in next-generation sequencing data. Bioinformatics, 29(15), 1888–1889. <a class="reference external" href="https://doi.org/10.1093/bioinformatics/btt293">https://doi.org/10.1093/bioinformatics/btt293</a></p></li>
<li><p><strong>ANNOVAR</strong>: Wang, K., Li, M., &amp; Hakonarson, H. (2010). ANNOVAR: Functional annotation of genetic variants from high-throughput sequencing data. Nucleic Acids Research, 38(16), 1–7. <a class="reference external" href="https://doi.org/10.1093/nar/gkq603">https://doi.org/10.1093/nar/gkq603</a></p></li>
<li><p><strong>ASCAT</strong>: Van Loo, P., Nordgard, S. H., Lingjaerde, O. C., Russnes, H. G., Rye, I. H., Sun, W., … Kristensen, V. N. (2010). Allele-specific copy number analysis of tumors. Proceedings of the National Academy of Sciences, 107(39), 16910–16915. <a class="reference external" href="https://doi.org/10.1073/pnas.1009843107">https://doi.org/10.1073/pnas.1009843107</a></p></li>
<li><p><strong>Treeomics</strong>: Reiter, J. G., Makohon-Moore, A. P., Gerold, J. M., Bozic, I., Chatterjee, K., Iacobuzio-Donahue, C. A., … Nowak, M. A. (2017). Reconstructing metastatic seeding patterns of human cancers. Nature Communications, 8, 14114. <a class="reference external" href="https://doi.org/10.1038/ncomms14114">https://doi.org/10.1038/ncomms14114</a></p></li>
<li><p><strong>deconstructSigs</strong>: Rosenthal, R., McGranahan, N., Herrero, J., Taylor, B. S., &amp; Swanton, C. (2016). deconstructSigs: delineating mutational processes in single tumors distinguishes DNA repair deficiencies and patterns of carcinoma evolution. Genome Biology, 17(1), 31. <a class="reference external" href="https://doi.org/10.1186/s13059-016-0893-4">https://doi.org/10.1186/s13059-016-0893-4</a></p></li>
<li><p><strong>MutationalPatterns</strong>: Blokzijl, F., Janssen, R., van Boxtel, R., &amp; Cuppen, E. (2017). MutationalPatterns: comprehensive genome-wide analysis of mutational processes. bioRxiv, 1–20. <a class="reference external" href="https://doi.org/https://doi.org/10.1101/071761">https://doi.org/https://doi.org/10.1101/071761</a></p></li>
<li><p><strong>MaSuRCA</strong>: Zimin, A. V., Marçais, G., Puiu, D., Roberts, M., Salzberg, S. L., &amp; Yorke, J. A. (2013). The MaSuRCA genome assembler. Bioinformatics, 29(21), 2669–2677. <a class="reference external" href="https://doi.org/10.1093/bioinformatics/btt476">https://doi.org/10.1093/bioinformatics/btt476</a></p></li>
<li><p><strong>VarDict</strong>: Lai, Z., Markovets, A., Ahdesmaki, M., Chapman, B., Hofmann, O., Mcewen, R., … Dry, J. R. (2016). VarDict: A novel and versatile variant caller for next-generation sequencing in cancer research. Nucleic Acids Research, 44(11), 1–11. <a class="reference external" href="https://doi.org/10.1093/nar/gkw227">https://doi.org/10.1093/nar/gkw227</a></p></li>
<li><p><strong>vt</strong>: Tan, A., Abecasis, G. R., &amp; Kang, H. M. (2015). Unified representation of genetic variants. Bioinformatics, 31(13), 2202–2204. <a class="reference external" href="https://doi.org/10.1093/bioinformatics/btv112">https://doi.org/10.1093/bioinformatics/btv112</a></p></li>
<li><p><strong>peddy</strong>: Pedersen, B. S., &amp; Quinlan, A. R. (2017). Who’s Who? Detecting and Resolving Sample Anomalies in Human DNA Sequencing Studies with Peddy. American Journal of Human Genetics, 100(3), 406–413. <a class="reference external" href="https://doi.org/10.1016/j.ajhg.2017.01.017">https://doi.org/10.1016/j.ajhg.2017.01.017</a></p></li>
<li><p><strong>GQT</strong>: Layer, R. M., Kindlon, N., Karczewski, K. J., &amp; Quinlan, A. R. (2015). Efficient genotype compression and analysis of large genetic-variation data sets. Nature Methods, 13(1). <a class="reference external" href="https://doi.org/10.1038/nmeth.3654">https://doi.org/10.1038/nmeth.3654</a></p></li>
</ul>
<p><em>Tool sets and softwares required at various steps of pipeline development</em></p>
<ol class="arabic simple">
<li><p><strong>FastQC</strong>: Quality control tool. <a class="reference external" href="https://www.bioinformatics.babraham.ac.uk/projects/fastqc/">https://www.bioinformatics.babraham.ac.uk/projects/fastqc/</a></p></li>
<li><p><strong>Cutadapt</strong>: Adapter removal tool. <a class="reference external" href="https://cutadapt.readthedocs.io/en/stable/">https://cutadapt.readthedocs.io/en/stable/</a></p></li>
<li><p><strong>Trim Galore!</strong>: FastQC and Cutadapt wrapper. <a class="reference external" href="https://www.bioinformatics.babraham.ac.uk/projects/trim_galore/">https://www.bioinformatics.babraham.ac.uk/projects/trim_galore/</a></p></li>
<li><p><strong>Picardtools</strong>: BAM/SAM/VCF/CRAM manipulator. <a class="reference external" href="http://broadinstitute.github.io/picard/">http://broadinstitute.github.io/picard/</a></p>
<ul class="simple">
<li><p><strong>MarkDuplicate</strong>: Mark duplicate reads and potentially remove them</p></li>
<li><p><strong>LiftoverVcf</strong>: liftover VCF between builds</p></li>
<li><p><strong>CollectHsMetric</strong>: Collects hybrid-selection (HS) metrics for a SAM or BAM file</p></li>
<li><p><strong>CollectAlignmentSummaryMetrics</strong>: Produces a summary of alignment metrics from a SAM or BAM file</p></li>
<li><p><strong>CollectGcBiasMetrics</strong>: Collect metrics regarding GC bias</p></li>
<li><p><strong>CollectWgsMetrics</strong>: Collect metrics about coverage and performance of whole genome sequencing (WGS) experiments</p></li>
</ul>
</li>
<li><p><strong>GATK</strong>: A variant discovery tool: <a class="reference external" href="https://gatk.broadinstitute.org/hc/en-us">https://gatk.broadinstitute.org/hc/en-us</a></p>
<ul class="simple">
<li><p><strong>BaseRecalibrator</strong>: Detect systematic error in base quality score</p></li>
<li><p><strong>Somatic Indel Realigner</strong>: Local Realignment around Indels</p></li>
<li><p><strong>ContEst</strong>: Estimate cross sample contamination</p></li>
<li><p><strong>DepthOfCoverage</strong>: Assess sequence coverage by sample, read group, or libraries</p></li>
<li><p><strong>DuplicateReadFilter</strong>: remove duplicated from flag set by MarkDuplicates</p></li>
</ul>
</li>
<li><p><strong>Samtools</strong>: Reading/writing/editing/indexing/viewing SAM/BAM/CRAM format <a class="reference external" href="http://www.htslib.org/">http://www.htslib.org/</a></p></li>
<li><p><strong>Sambamba</strong>: Tools for working with SAM/BAM/CRAM data <a class="reference external" href="http://lomereiter.github.io/sambamba/">http://lomereiter.github.io/sambamba/</a></p></li>
<li><p><strong>bcftools</strong>: Reading/writing BCF2/VCF/gVCF files and calling/filtering/summarising SNP and short indel sequence variants <a class="reference external" href="http://www.htslib.org/doc/bcftools.html">http://www.htslib.org/doc/bcftools.html</a></p></li>
<li><p><strong>vcftools</strong>: VCFtools is a program package designed for working with VCF files, such as those generated by the 1000 Genomes Project. <a class="reference external" href="https://vcftools.github.io/index.html">https://vcftools.github.io/index.html</a></p></li>
<li><p><strong>Delly2</strong>: An integrated structural variant prediction method that can discover, genotype and visualize deletions, tandem duplications, inversions and translocations <a class="reference external" href="https://github.com/dellytools/delly">https://github.com/dellytools/delly</a></p></li>
<li><p><strong>PLINK</strong>: PLINK: Whole genome data analysis toolset <a class="reference external" href="https://www.cog-genomics.org/plink2">https://www.cog-genomics.org/plink2</a></p></li>
<li><p><strong>freebayes</strong>: a haplotype-based variant detector. <a class="reference external" href="https://github.com/ekg/freebayes">https://github.com/ekg/freebayes</a></p></li>
<li><p><strong>AscatNGS</strong>: Allele-Specific Copy Number Analysis of Tumors, tumor purity and ploidy <a class="reference external" href="https://github.com/cancerit/ascatNgs">https://github.com/cancerit/ascatNgs</a></p></li>
<li><p><strong>MutationalPatterns</strong>: R package for extracting and visualizing mutational patterns in base substitution catalogues <a class="reference external" href="https://github.com/UMCUGenetics/MutationalPatterns">https://github.com/UMCUGenetics/MutationalPatterns</a></p></li>
<li><p><strong>desconstructSigs</strong>: identification of mutational signatures within a single tumor sample <a class="reference external" href="https://github.com/raerose01/deconstructSigs">https://github.com/raerose01/deconstructSigs</a></p></li>
<li><p><strong>treeOmics</strong>: Decrypting somatic mutation patterns to reveal the evolution of cancer <a class="reference external" href="https://github.com/johannesreiter/treeomics">https://github.com/johannesreiter/treeomics</a></p></li>
<li><p><strong>controlFreeC</strong>: Copy number and allelic content caller <a class="reference external" href="http://boevalab.com/FREEC/">http://boevalab.com/FREEC/</a></p></li>
<li><p><strong>MuTect2</strong>: Call somatic SNPs and indels via local re-assembly of haplotypes <a class="reference external" href="https://gatk.broadinstitute.org/hc/en-us/articles/360037593851-Mutect2">https://gatk.broadinstitute.org/hc/en-us/articles/360037593851-Mutect2</a></p></li>
<li><p><strong>Annovar</strong>: annotation of detected genetic variation <a class="reference external" href="http://annovar.openbioinformatics.org/en/latest/">http://annovar.openbioinformatics.org/en/latest/</a></p></li>
<li><p><strong>Strelka</strong>: Small variant caller <a class="reference external" href="https://github.com/Illumina/strelka">https://github.com/Illumina/strelka</a></p></li>
<li><p><strong>Manta</strong>: Structural variant caller <a class="reference external" href="https://github.com/Illumina/manta">https://github.com/Illumina/manta</a></p></li>
<li><p><strong>PurBayes</strong>: estimate tumor purity and clonality</p></li>
<li><p><strong>VarDict</strong>: variant caller for both single and paired sample variant calling from BAM files <a class="reference external" href="https://github.com/AstraZeneca-NGS/VarDict">https://github.com/AstraZeneca-NGS/VarDict</a></p></li>
<li><p><strong>SNPeff/SNPSift</strong>: Genomic variant annotations and functional effect prediction toolbox. <a class="reference external" href="http://snpeff.sourceforge.net/">http://snpeff.sourceforge.net/</a> and <a class="reference external" href="http://snpeff.sourceforge.net/SnpSift.html">http://snpeff.sourceforge.net/SnpSift.html</a></p></li>
<li><p><strong>IGV</strong>: visualization tool for interactive exploration <a class="reference external" href="http://software.broadinstitute.org/software/igv/">http://software.broadinstitute.org/software/igv/</a></p></li>
<li><p><strong>SVDetect</strong>: a tool to detect genomic structural variations <a class="reference external" href="http://svdetect.sourceforge.net/Site/Home.html">http://svdetect.sourceforge.net/Site/Home.html</a></p></li>
<li><p><strong>GenomeSTRiP</strong>: A suite of tools for discovering and genotyping structural variations using sequencing data <a class="reference external" href="http://software.broadinstitute.org/software/genomestrip/">http://software.broadinstitute.org/software/genomestrip/</a></p></li>
<li><p><strong>BreakDancer</strong>: SV detection from paired end reads mapping <a class="reference external" href="https://github.com/genome/breakdancer">https://github.com/genome/breakdancer</a></p></li>
<li><p><strong>pIndel</strong>: Detect breakpoints of large deletions, medium sized insertions, inversions, and tandem duplications <a class="reference external" href="https://github.com/genome/pindel">https://github.com/genome/pindel</a></p></li>
<li><p><strong>VarScan</strong>: Variant calling and somatic mutation/CNV detection <a class="reference external" href="https://github.com/dkoboldt/varscan">https://github.com/dkoboldt/varscan</a></p></li>
<li><p><strong>VEP</strong>: Variant Effect Predictor <a class="reference external" href="https://www.ensembl.org/info/docs/tools/vep/index.html">https://www.ensembl.org/info/docs/tools/vep/index.html</a></p></li>
<li><p><strong>Probablistic2020</strong>: Simulates somatic mutations, and calls statistically significant oncogenes and tumor suppressor genes based on a randomization-based test  <a class="reference external" href="https://github.com/KarchinLab/probabilistic2020">https://github.com/KarchinLab/probabilistic2020</a></p></li>
<li><p><strong>2020plus</strong>: Classifies genes as an oncogene, tumor suppressor gene, or as a non-driver gene by using Random Forests <a class="reference external" href="https://github.com/KarchinLab/2020plus">https://github.com/KarchinLab/2020plus</a></p></li>
<li><p><strong>vtools</strong>: variant tools is a software tool for the manipulation, annotation, selection, simulation, and analysis of variants in the context of next-gen sequencing analysis. <a class="reference external" href="https://vatlab.github.io/vat-docs/">https://vatlab.github.io/vat-docs/</a></p></li>
<li><p><strong>vt</strong>: A variant tool set that discovers short variants from Next Generation Sequencing data. <a class="reference external" href="https://genome.sph.umich.edu/wiki/Vt">https://genome.sph.umich.edu/wiki/Vt</a> and <a class="reference external" href="https://github.com/atks/vt">https://github.com/atks/vt</a></p></li>
<li><p><strong>CNVnator</strong>: a tool for CNV discovery and genotyping from depth-of-coverage by mapped reads. <a class="reference external" href="https://github.com/abyzovlab/CNVnator">https://github.com/abyzovlab/CNVnator</a></p></li>
<li><p><strong>CNVpytor</strong>: a tool for copy number variation detection and analysis from read depth and allele imbalance in whole-genome sequencing. <a class="reference external" href="https://github.com/abyzovlab/CNVpytor">https://github.com/abyzovlab/CNVpytor</a></p></li>
<li><p><strong>SvABA</strong>: Structural variation and indel detection by local assembly. <a class="reference external" href="https://github.com/walaj/svaba">https://github.com/walaj/svaba</a></p></li>
<li><p><strong>indelope</strong>: find indels and SVs too small for structural variant callers and too large for GATK. <a class="reference external" href="https://github.com/brentp/indelope">https://github.com/brentp/indelope</a></p></li>
<li><p><strong>peddy</strong>: peddy compares familial-relationships and sexes as reported in a PED/FAM file with those inferred from a VCF. <a class="reference external" href="https://github.com/brentp/peddy">https://github.com/brentp/peddy</a></p></li>
<li><p><strong>cyvcf2</strong>: cyvcf2 is a cython wrapper around htslib built for fast parsing of Variant Call Format (VCF) files. <a class="reference external" href="https://github.com/brentp/cyvcf2">https://github.com/brentp/cyvcf2</a></p></li>
<li><p><strong>GQT</strong>: Genotype Query Tools (GQT) is command line software and a C API for indexing and querying large-scale genotype data sets. <a class="reference external" href="https://github.com/ryanlayer/gqt">https://github.com/ryanlayer/gqt</a></p></li>
<li><p><strong>LOFTEE</strong>: Loss-Of-Function Transcript Effect Estimator. A VEP plugin to identify LoF (loss-of-function) variation. Assesses variants that are: Stop-gained, Splice site disrupting, and Frameshift variants. <a class="reference external" href="https://github.com/konradjk/loftee">https://github.com/konradjk/loftee</a></p></li>
<li><p><strong>PureCN</strong>: copy number calling and SNV classification using targeted short read sequencing <a class="reference external" href="https://bioconductor.org/packages/release/bioc/html/PureCN.html">https://bioconductor.org/packages/release/bioc/html/PureCN.html</a></p></li>
<li><p><strong>SVCaller</strong>: A structural variant caller. <a class="reference external" href="https://github.com/tomwhi/svcaller">https://github.com/tomwhi/svcaller</a></p></li>
<li><p><strong>SnakeMake</strong>: A workflow manager. <a class="reference external" href="http://snakemake.readthedocs.io/en/stable/index.html">http://snakemake.readthedocs.io/en/stable/index.html</a></p></li>
<li><p><strong>BWA</strong>: BWA is a software package for mapping low-divergent sequences against a large reference genome, such as the human genome. It consists of three algorithms: BWA-backtrack, BWA-SW and BWA-MEM. <a class="reference external" href="http://bio-bwa.sourceforge.net/">http://bio-bwa.sourceforge.net/</a></p></li>
<li><p><strong>wgsim</strong>: Wgsim is a small tool for simulating sequence reads from a reference genome. It is able to simulate diploid genomes with SNPs and insertion/deletion (INDEL) polymorphisms, and simulate reads with uniform substitution sequencing errors. <a class="reference external" href="https://github.com/lh3/wgsim">https://github.com/lh3/wgsim</a></p></li>
<li><p><strong>dwgsim</strong>: Whole genome simulation can be performed with dwgsim. dwgsim is based off of wgsim found in SAMtools. <a class="reference external" href="https://github.com/nh13/DWGSIM">https://github.com/nh13/DWGSIM</a></p></li>
<li><p><strong>THetA</strong>: Tumor Heterogeneity Analysis. This algorithm estimates tumor purity and clonal/subclonal copy number aberrations directly from high-throughput DNA sequencing data. <a class="reference external" href="https://github.com/raphael-group/THetA">https://github.com/raphael-group/THetA</a></p></li>
<li><p><strong>Skewer</strong>: Adapter trimming, similar to cutadapt. <a class="reference external" href="https://github.com/relipmoc/skewer">https://github.com/relipmoc/skewer</a></p></li>
<li><p><strong>Phylowgs</strong>: Application for inferring subclonal composition and evolution from whole-genome sequencing data. <a class="reference external" href="https://github.com/morrislab/phylowgs">https://github.com/morrislab/phylowgs</a></p></li>
<li><p><strong>superFreq</strong>: SuperFreq is an R package that analyses cancer exomes to track subclones. <a class="reference external" href="https://github.com/ChristofferFlensburg/superFreq">https://github.com/ChristofferFlensburg/superFreq</a></p></li>
<li><p><strong>readVCF-r</strong>: Read VCFs into R and annotatte them. <a class="reference external" href="https://bioconductor.org/packages/release/bioc/html/VariantAnnotation.html">https://bioconductor.org/packages/release/bioc/html/VariantAnnotation.html</a></p></li>
<li><p><strong>vcfr</strong>: Read VCFs into R. <a class="reference external" href="https://github.com/knausb/vcfR">https://github.com/knausb/vcfR</a></p></li>
<li><p><strong>msisensor</strong>: microsatellite instability detection using paired tumor-normal <a class="reference external" href="https://github.com/ding-lab/msisensor">https://github.com/ding-lab/msisensor</a></p></li>
<li><p><strong>MOSAIC</strong>: MicrOSAtellite Instability Classifier <a class="reference external" href="https://github.com/ronaldhause/mosaic">https://github.com/ronaldhause/mosaic</a></p></li>
<li><p><strong>MANTIS</strong>: Microsatellite Analysis for Normal-Tumor InStability <a class="reference external" href="https://github.com/OSU-SRLab/MANTIS">https://github.com/OSU-SRLab/MANTIS</a></p></li>
<li><p><strong>SBDB</strong>: A toolkit for constricting and querying structural variant databases <a class="reference external" href="https://github.com/J35P312/SVDB">https://github.com/J35P312/SVDB</a></p></li>
</ol>
</section>
</section>

        </article>
      </div>
      <footer>
        
        <div class="related-pages">
          <a class="next-page" href="README.html">
              <div class="page-info">
                <div class="context">
                  <span>Next</span>
                </div>
                <div class="title">Documentation Guidline</div>
              </div>
              <svg class="furo-related-icon"><use href="#svg-arrow-right"></use></svg>
            </a>
          <a class="prev-page" href="bioinfo_softwares.html">
              <svg class="furo-related-icon"><use href="#svg-arrow-right"></use></svg>
              <div class="page-info">
                <div class="context">
                  <span>Previous</span>
                </div>
                
                <div class="title">Tools and software</div>
                
              </div>
            </a>
        </div>
        <div class="bottom-of-page">
          <div class="left-details">
            <div class="copyright">
                Copyright &#169; 2023, Clinical Genomics
            </div>
            Made with <a href="https://www.sphinx-doc.org/">Sphinx</a> and <a class="muted-link" href="https://pradyunsg.me">@pradyunsg</a>'s
            
            <a href="https://github.com/pradyunsg/furo">Furo</a>
            
          </div>
          <div class="right-details">
            
          </div>
        </div>
        
      </footer>
    </div>
    <aside class="toc-drawer">
      
      
      <div class="toc-sticky toc-scroll">
        <div class="toc-title-container">
          <span class="toc-title">
            On this page
          </span>
        </div>
        <div class="toc-tree-container">
          <div class="toc-tree">
            <ul>
<li><a class="reference internal" href="#">References and other resources</a><ul>
<li><a class="reference internal" href="#sample-datasets">Sample datasets</a></li>
<li><a class="reference internal" href="#relevant-publications">Relevant publications</a></li>
<li><a class="reference internal" href="#methods-and-tools">Methods and tools</a></li>
</ul>
</li>
</ul>

          </div>
        </div>
      </div>
      
      
    </aside>
  </div>
</div><script data-url_root="./" id="documentation_options" src="_static/documentation_options.js"></script>
    <script src="_static/doctools.js"></script>
    <script src="_static/sphinx_highlight.js"></script>
    <script src="_static/scripts/furo.js"></script>
    </body>
</html>