========
Workflow
========

BALSAMIC ( **version** = 8.2.7 ) uses myriad of tools and softwares to analyze fastq files. This section covers why each
one is included: usage and parameters, and relevant external links.


ascatNgs
~~~~~~~~
:Source code: `GitHub <https://github.com/cancerit/ascatNgs>`
:Article: `PNAS <https://www.pnas.org/doi/full/10.1073/pnas.1009843107/>`
:Version: `4.5.0`

bcftools
~~~~~~~~
:Source code: `GitHub <https://github.com/samtools/bcftools>`
:Article: `Bioinformatics <https://pubmed.ncbi.nlm.nih.gov/21903627/>`
:Version: `>1.9`

bedtools
~~~~~~~~
:Source code: `GitHub <https://github.com/arq5x/bedtools2>`
:Article: `Bioinformatics <https://pubmed.ncbi.nlm.nih.gov/20110278/>`
:Version: `2.30.0`

bwa
~~~
:Source code: `GitHub <https://github.com/lh3/bwa>`
:Article: `Bioinformatics <https://arxiv.org/abs/1303.3997>`
:Version: `0.7.15`

cnvkit
~~~~~~
:Source code: `GitHub <https://github.com/etal/cnvkit>`
:Article: `PLOS Computational Biology <http://dx.doi.org/10.1371/journal.pcbi.1004873>`
:Version: `0.9.4`

csvkit
~~~~~~
:Source code: `GitHub <https://github.com/wireservice/csvkit>`
:Article: `-`
:Version: `1.0.4`

delly
~~~~~~~
:Source code: `GitHub <https://github.com/dellytools/delly>`
:Article: `Bioinformatics <https://academic.oup.com/bioinformatics/article/28/18/i333/245403>`
:Version: `0.8.7`

ensembl-vep
~~~~~~~~~~~
:Source code: `GitHub <https://github.com/Ensembl/ensembl-vep>`
:Article: `Genome Biology <https://genomebiology.biomedcentral.com/articles/10.1186/s13059-016-0974-4>`
:Version: `100.2`

fastp
~~~~~
:Source code: `GitHub <https://github.com/OpenGene/fastp>`
:Article: `Bioinformatics <https://doi.org/10.1093/bioinformatics/bty560>`
:Version: `0.20.1`

fastqc
~~~~~~
:Source code: `GitHub <https://github.com/s-andrews/FastQC>`
:Article: `Babraham <http://www.bioinformatics.babraham.ac.uk/projects/fastqc/>`
:Version: `0.11.9`

gatk
~~~~
:Source code: `Github <https://github.com/broadinstitute/gatk>`
:Article: `Current Protocols in Bioinformatics <https://pubmed.ncbi.nlm.nih.gov/25431634/>`
:Version: `3.8`

manta
~~~~~
:Source code: `GitHub <https://github.com/Illumina/manta>`
:Article: `Bioinformatics <https://doi.org/10.1093/bioinformatics/btv710>`
:Version: `1.6.0`

multiqc
~~~~~~~
:Source code: `GitHub <https://github.com/ewels/MultiQC>`
:Article: `Bioinformatics <http://dx.doi.org/10.1093/bioinformatics/btw354>`
:Version: `1.11`

mosdepth
~~~~~~~~
:Source code: `GitHub <https://github.com/brentp/mosdepth>`
:Article: `Bioinformatics <https://academic.oup.com/bioinformatics/article/34/5/867/4583630?login=true>`
:Version: `0.2.9`

picard
~~~~~~
:Source code: `GitHub <https://github.com/broadinstitute/picard>`
:Article: `-`
:Version: `2.25.0`

sambamba
~~~~~~~~
:Source code: `GitHub <https://github.com/biod/sambamba>`
:Article: `Bioinformatics <https://pubmed.ncbi.nlm.nih.gov/25697820/>`
:Version: `0.6.6`

samtools
~~~~~~~~
:Source code: `GitHub <https://github.com/samtools/samtools>`
:Article: `Bioinformatics <https://pubmed.ncbi.nlm.nih.gov/19505943/>`
:Version: `1.12`

sentieon-tools
~~~~~~~~~~~~~~
:Source code: `Commercial Tool <https://www.sentieon.com/>`
:Article: `Bioinformatics <https://www.biorxiv.org/content/10.1101/115717v2>`
:Version: `202010.02`

tabix
~~~~~
:Source code: `GitHub <https://github.com/samtools/tabix>`
:Article: `Bioinformatics <https://academic.oup.com/bioinformatics/article/27/5/718/262743>`
:Version: `0.2.6`

vardict
~~~~~~~
:Source code: `GitHub <https://github.com/AstraZeneca-NGS/VarDict>`
:Article: `Nucleic Acid Research <https://pubmed.ncbi.nlm.nih.gov/27060149/>`
:Version: `2019.06.04`

vcfanno
~~~~~~~
:Source code: `GitHub <https://github.com/brentp/vcfanno>`
:Article: `Genome Biology <https://genomebiology.biomedcentral.com/articles/10.1186/s13059-016-0973-5/>_`
:Version: `0.3.3`

vcf2cytosure
~~~~~~~~~~~~~
:Source code: `GitHub <https://github.com/NBISweden/vcf2cytosure>`
:Article: `Bioinformatics <https://academic.oup.com/bioinformatics/article/28/18/i333/245403>`
:Version: `0.7.1`