======
Workflow
======

BALSAMIC ( **version** = 4.2.2 ) uses myriad of tools and softwares to analyze fastq files. This section covers why each
one is included: usage and parameters, and relevant external links.


bcftools
~~~~~~
:Source code: `GitHub <https://github.com/samtools/bcftools>`_
:Article: `Bioinformatics <https://pubmed.ncbi.nlm.nih.gov/21903627/>`_
:Version: `1.9.0`_

bedtools
~~~~~~
:Source code: `GitHub <https://github.com/arq5x/bedtools2>`_
:Article: `Bioinformatics <https://pubmed.ncbi.nlm.nih.gov/20110278/>`_
:Version: `2.28.0`_

bwa
~~~~~~
:Source code: `GitHub <https://github.com/lh3/bwa>`_
:Article: `Bioinformatics <https://arxiv.org/abs/1303.3997>`_
:Version: `0.7.15`_

cnvkit
~~~~~~
:Source code: `GitHub <https://github.com/etal/cnvkit>`_
:Article: `PLOS Computational Biology <http://dx.doi.org/10.1371/journal.pcbi.1004873>`_
:Version: `0.9.4a0`_

csvkit
~~~~~~
:Source code: `GitHub <https://github.com/wireservice/csvkit>`_
:Article: `-`_
:Version: `1.0.4`_

ensembl-vep
~~~~~~
:Source code: `GitHub <https://github.com/Ensembl/ensembl-vep>`_
:Article: `Genome Biology <https://genomebiology.biomedcentral.com/articles/10.1186/s13059-016-0974-4>`_
:Version: `99.1`_

fastp
~~~~~~
:Source code: `GitHub <https://github.com/OpenGene/fastp>`_
:Article: `Bioinformatics <https://doi.org/10.1093/bioinformatics/bty560>`_
:Version: `0.20.0`_

fastqc
~~~~~~
:Source code: `GitHub <https://github.com/s-andrews/FastQC>`_
:Article: `Babraham <http://www.bioinformatics.babraham.ac.uk/projects/fastqc/>`_
:Version: `0.11.5`_

gatk
~~~~~~
:Source code: `Github <https://github.com/broadinstitute/gatk>`_
:Article: `Current Protocols in Bioinformatics <https://pubmed.ncbi.nlm.nih.gov/25431634/>`_
:Version: `3.8`_

manta
~~~~~~
:Source code: `GitHub <https://github.com/Illumina/manta>`_
:Article: `Bioinformatics <https://doi.org/10.1093/bioinformatics/btv710>`_
:Version: `1.3.0`_

multiqc
~~~~~~
:Source code: `GitHub <https://github.com/ewels/MultiQC>`_
:Article: `Bioinformatics <http://dx.doi.org/10.1093/bioinformatics/btw354>`_
:Version: `1.7`_

picard
~~~~~~
:Source code: `GitHub <https://github.com/broadinstitute/picard>`_
:Article: `-`_
:Version: `2.17.0`_

sambamba
~~~~~~
:Source code: `GitHub <https://github.com/biod/sambamba>`_
:Article: `Bioinformatics <https://pubmed.ncbi.nlm.nih.gov/25697820/>`_
:Version: `0.6.6`_

samtools
~~~~~~
:Source code: `GitHub <https://github.com/samtools/samtools>`_
:Article: `Bioinformatics <https://pubmed.ncbi.nlm.nih.gov/19505943/>`_
:Version: `1.6`_

strelka
~~~~~~
:Source code: `GitHub <https://github.com/Illumina/strelka>`_
:Article: `Nature Methods <https://www.nature.com/articles/s41592-018-0051-x/>`_
:Version: `2.8.4`_

tabix
~~~~~~
:Source code: `GitHub <https://github.com/samtools/tabix>`_
:Article: `Bioinformatics <https://academic.oup.com/bioinformatics/article/27/5/718/262743>`_
:Version: `0.2.5`_

vardict
~~~~~~
:Source code: `GitHub <https://github.com/AstraZeneca-NGS/VarDict>`_
:Article: `Nucleic Acid Research <https://pubmed.ncbi.nlm.nih.gov/27060149/>`_
:Version: `2019.06.04`_

